From fe5794ee4ceb67bd7c5673a84c53fca24ad4b93c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 17 Dec 2018 23:31:31 +0800
Subject: [PATCH 6/8] pds: Remove unlikely in resched_cpu_if_curr_is().

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 6cf05e101c69..4f1b27baad10 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -582,7 +582,7 @@ static void resched_cpu_if_curr_is(int cpu, int priority)
 	if (set_nr_if_polling(rq->idle)) {
 		trace_sched_wake_idle_without_ipi(cpu);
 	} else {
-		if (unlikely(!do_raw_spin_trylock(&rq->lock)))
+		if (!do_raw_spin_trylock(&rq->lock))
 			goto out;
 		spin_acquire(&rq->lock.dep_map, SINGLE_DEPTH_NESTING, 1, _RET_IP_);
 
-- 
2.20.1.2.gb21ebb671b

