From dab684ca6fc50aad4e574561970ba13429dd1b51 Mon Sep 17 00:00:00 2001
From: Janne Grunau <j@jannau.net>
Date: Sun, 30 Aug 2026 10:47:29 +0200
Subject: [PATCH 27/47] fixup! HID: magicmouse: add support for Macbook
 trackpads

---
 drivers/hid/hid-magicmouse.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/hid/hid-magicmouse.c b/drivers/hid/hid-magicmouse.c
index 88a5f71d1..552ee3453 100644
--- a/drivers/hid/hid-magicmouse.c
+++ b/drivers/hid/hid-magicmouse.c
@@ -539,8 +539,8 @@ static int __magicmouse_raw_event_usb(struct hid_device *hdev,
 			return 0;
 		}
 
-		__magicmouse_raw_event(hdev, report, data + 2, data[1], true);
-		__magicmouse_raw_event(hdev, report, data + 2 + data[1],
+		__magicmouse_raw_event_usb(hdev, report, data + 2, data[1], true);
+		__magicmouse_raw_event_usb(hdev, report, data + 2 + data[1],
 			size - 2 - data[1], true);
 		return 0;
 	default:
-- 
2.55.0

