From 34eec75232dae8d9c6e8c2f79cde19d3987d7d60 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 30 Jun 2021 23:27:04 +0200
Subject: [PATCH 5/7] kbuild: add proper bool for MODULE_COMPRESS_ZSTD_ULTRA

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index d192b318e..66bbe0b1e 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2268,7 +2268,7 @@ endchoice
 
 config MODULE_COMPRESS_ZSTD_ULTRA
 	depends on MODULE_COMPRESS_ZSTD
-	bool "ZSTD"
+	bool "Enable ZSTD ultra compression"
 	help
 	  Compress modules with ZSTD using the highest possible compression.
 
-- 
2.32.0.349.gdaab8a564f

