From 4667a6134662b62801ae2bbeb86b74ed46b8d616 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 7 Aug 2020 23:09:43 +0800
Subject: [PATCH 029/265] Project-C v5.8-r0

---
 kernel/sched/bmq_imp.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq_imp.h b/kernel/sched/bmq_imp.h
index cb0fc0688a89..e581fbf3528b 100644
--- a/kernel/sched/bmq_imp.h
+++ b/kernel/sched/bmq_imp.h
@@ -1,4 +1,4 @@
-#define ALT_SCHED_VERSION_MSG "sched/bmq: BMQ CPU Scheduler 5.7-r3 by Alfred Chen.\n"
+#define ALT_SCHED_VERSION_MSG "sched/bmq: BMQ CPU Scheduler 5.8-r0 by Alfred Chen.\n"
 
 static inline void sched_queue_init(struct rq *rq)
 {
-- 
2.35.1.677.gabf474a5dd

