From 6774bd092fd660a88d302a2e4c008fb60b710f38 Mon Sep 17 00:00:00 2001
From: Alexandre Frade <kernel@xanmod.org>
Date: Sun, 29 May 2022 00:57:40 +0000
Subject: [PATCH 12/13] XANMOD: scripts/setlocalversion: remove "+" tag for git
 repo short version

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 scripts/setlocalversion | 1 -
 1 file changed, 1 deletion(-)

diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index 3d3babac8..7c0ccffd0 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -76,7 +76,6 @@ scm_version()
 		# If only the short version is requested, don't bother
 		# running further git commands
 		if $short; then
-			echo "+"
 			return
 		fi
 		# If we are past the tagged commit, we pretty print it.
-- 
2.40.0.71.g950264636c

