From ca18fbd89f83061a0adb30a3b5e7b053ccb5d186 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 24 Apr 2020 17:47:20 +0200
Subject: [PATCH 19/26] MAINTAINERS: add BFQ-dev

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 MAINTAINERS | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index 5a5332b3591d..a981268288aa 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -3074,6 +3074,13 @@ S:	Maintained
 F:	block/bfq-*
 F:	Documentation/block/bfq-iosched.rst
 
+BFQ-dev I/O SCHEDULER
+M:	Piotr Gorski <lucjan.lucjanov@gmail.com>
+T:	git git://github.com/sirlucjan/bfq-mq-lucjan.git
+S:	Maintained
+F:	block/bfq-*
+F:	Documentation/block/bfq-iosched.rst
+
 BFS FILE SYSTEM
 M:	"Tigran A. Aivazian" <aivazian.tigran@gmail.com>
 S:	Maintained
-- 
2.26.2.533.gb34789c0b0

