From 8b2ff8a0126994c46e901663993d3ac4cdabf44d Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 22 Aug 2022 14:09:58 +0200
Subject: [PATCH 39/58] arch/x86: improve KBUILD_CFLAG

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 arch/x86/Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/x86/Makefile b/arch/x86/Makefile
index 7854685c5..41d480623 100644
--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -67,7 +67,8 @@ export BITS
 #
 #    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53383
 #
-KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx
+KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -mno-avx2 \
+           -mno-avx512f -mpopcnt -O3 $(call cc-option, -fno-tree-vectorize)
 
 ifeq ($(CONFIG_X86_KERNEL_IBT),y)
 #
-- 
2.37.1.377.g679aad9e82

