From 7f0e6e8639972765d3e1d50e1965dcf24f35fd71 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 6 Sep 2022 11:05:01 +0200
Subject: [PATCH] cachyos-5.19: mm/swap: Disable swap-in readahead

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 mm/swap.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/mm/swap.c b/mm/swap.c
index 034bb2487..5b61f430d 100644
--- a/mm/swap.c
+++ b/mm/swap.c
@@ -1130,6 +1130,10 @@ EXPORT_SYMBOL(pagevec_lookup_range_tag);
  */
 void __init swap_setup(void)
 {
+#ifdef CONFIG_CACHY
+	/* Only swap-in pages requested, avoid readahead */
+	page_cluster = 0;
+#else
 	unsigned long megs = totalram_pages() >> (20 - PAGE_SHIFT);
 
 	/* Use a smaller cluster for small-memory machines */
@@ -1141,4 +1145,5 @@ void __init swap_setup(void)
 	 * Right now other parts of the system means that we
 	 * _really_ don't want to cluster much more
 	 */
+#endif
 }
-- 
2.37.1.377.g679aad9e82

