From 6ad99f0f0af1a9c1d51ae0e116f58f0b072c2ccf Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 24 Jun 2019 07:50:11 +0200
Subject: [PATCH 02/26] block, bfq, DEBUG: log inject limit reset

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index edd0cb21290f..594e4e8525da 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -2006,6 +2006,8 @@ static void bfq_reset_inject_limit(struct bfq_data *bfqd,
 		bfqq->inject_limit = 1;
 
 	bfqq->decrease_time_jif = jiffies;
+
+	bfq_log_bfqq(bfqd, bfqq, "");
 }
 
 static void bfq_add_request(struct request *rq)
-- 
2.23.0.162.gf1d4a28250

