From 68e85461a01d768b47c2fefe9ac7c89099404543 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 30 May 2019 15:41:57 +0200
Subject: [PATCH 07/28] block, bfq, DEBUG: add BUG_ONs on waker detection

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 50c784f81290..fb35bd8c8b06 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -4821,6 +4821,8 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 							async_bfqq),
 				     bfq_bfqq_budget_left(async_bfqq)
 				);
+		BFQ_BUG_ON(bfqq->waker_bfqq == bfqq);
+
 		/*
 		 * The next three mutually-exclusive ifs decide
 		 * whether to try injection, and choose the queue to
-- 
2.23.0.385.gbc12974a89

