From 53483d8ccc1c38cdfa6f233058b7e1cc8e42e0fd Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Tue, 4 Jun 2019 12:48:56 +0200
Subject: [PATCH 09/29] block, bfq, DEBUG: check that injection never chooses
 injected queue

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index dcb78a71ed60..2c4be9830b8d 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -4902,6 +4902,10 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 		    icq_to_bic(async_bfqq->next_rq->elv.icq) == bfqq->bic &&
 		    bfq_serv_to_charge(async_bfqq->next_rq, async_bfqq) <=
 		    bfq_bfqq_budget_left(async_bfqq)) {
+			bfq_log_bfqq(bfqd, bfqq,
+				     "choosing directly the async queue %d",
+				     bfqq->bic->bfqq[0]->pid);
+			BUG_ON(bfqq->bic->bfqq[0] == bfqq);
 			bfqq = bfqq->bic->bfqq[0];
 			bfq_log_bfqq(bfqd, bfqq,
 				     "chosen directly this async queue");
@@ -4915,17 +4919,22 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 			bfq_log_bfqq(bfqd, bfqq,
 				     "choosing directly the waker queue %d",
 				     bfqq->waker_bfqq->pid);
+			BUG_ON(bfqq->waker_bfqq == bfqq);
 			bfqq = bfqq->waker_bfqq;
 			bfq_log_bfqq(bfqd, bfqq,
 				     "chosen directly this waker queue");
 		} else if (!idling_boosts_thr_without_issues(bfqd, bfqq) &&
 			 (bfqq->wr_coeff == 1 || bfqd->wr_busy_queues > 1 ||
 			  !bfq_bfqq_has_short_ttime(bfqq))) {
+			struct bfq_queue *new_bfqq;
+
 			bfq_log_bfqq(bfqd, bfqq,
 				     "looking inject wr_busy %d long_tt %d",
 				     bfqd->wr_busy_queues,
 				     !bfq_bfqq_has_short_ttime(bfqq));
-			bfqq = bfq_choose_bfqq_for_injection(bfqd);
+			new_bfqq = bfq_choose_bfqq_for_injection(bfqd);
+			BUG_ON(new_bfqq == bfqq);
+			bfqq = new_bfqq;
 		} else {
 			bfqq = NULL;
 		}
-- 
2.23.0.385.gbc12974a89

