From cf2b801f6ce246c1c7337e0450799aebb90dea2c Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Sat, 22 Jun 2019 10:40:32 +0200
Subject: [PATCH 12/28] block, bfq, DEBUG: log per-queue rq completion times

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 block/bfq-iosched.h | 4 ++++
 2 files changed, 8 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 116681ff0b3f..b0edcf8e103c 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5226,6 +5226,7 @@ static struct request *__bfq_dispatch_request(struct blk_mq_hw_ctx *hctx)
 	       bfq_bfqq_wait_request(bfqq));
 
 	rq = bfq_dispatch_rq_from_bfqq(bfqd, bfqq);
+	bfqq->last_dispatch = ktime_get_ns();
 
 	BFQ_BUG_ON(bfqq->entity.budget < bfqq->entity.service);
 
@@ -6208,6 +6209,9 @@ static void bfq_completed_request(struct bfq_queue *bfqq, struct bfq_data *bfqd)
 
 	now_ns = ktime_get_ns();
 
+	bfq_log_bfqq(bfqd, bfqq, "rq completion time: %llu us",
+		     div_u64(now_ns - bfqq->last_dispatch, NSEC_PER_USEC));
+
 	bfqq->ttime.last_end_request = now_ns;
 
 	/*
diff --git a/block/bfq-iosched.h b/block/bfq-iosched.h
index 06628cc1da4f..698666ecb4e3 100644
--- a/block/bfq-iosched.h
+++ b/block/bfq-iosched.h
@@ -375,6 +375,10 @@ struct bfq_queue {
 	 * the woken queues when this queue exits.
 	 */
 	struct hlist_head woken_list;
+
+	/* time of last rq dispatch (ns) */
+	u64 last_dispatch;
+
 };
 
 /**
-- 
2.23.0.385.gbc12974a89

