From c0196f4e59f7eeffa8248269bd991d836237314b Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Sat, 22 Jun 2019 10:39:21 +0200
Subject: [PATCH 13/28] block, bfq, DEBUG: log TOTALLY_SEEKY

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index b0edcf8e103c..b6853885afd0 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5749,11 +5749,13 @@ bfq_update_io_seektime(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 {
 	bfqq->seek_history <<= 1;
 	bfqq->seek_history |= BFQ_RQ_SEEKY(bfqd, bfqq->last_request_pos, rq);
-	bfq_log_bfqq(bfqd, bfqq, "rq %p, distant %d, small %d, hist %x (%u)",
+	bfq_log_bfqq(bfqd, bfqq,
+		     "rq %p, distant %d, small %d, hist %x (%u), tot_seeky %d",
 		     rq, get_sdist(bfqq->last_request_pos, rq) > BFQQ_SEEK_THR,
 		     blk_rq_sectors(rq) < BFQQ_SECT_THR_NONROT,
 		     bfqq->seek_history,
-		     hweight32(bfqq->seek_history));
+		     hweight32(bfqq->seek_history),
+		     BFQQ_TOTALLY_SEEKY(bfqq));
 
 	if (bfqq->wr_coeff > 1 &&
 	    bfqq->wr_cur_max_time == bfqd->bfq_wr_rt_max_time &&
-- 
2.23.0.385.gbc12974a89

