From bfce0dae4dc51a43769ac79999dc4716db20e904 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 8 Oct 2019 15:02:17 +0200
Subject: [PATCH 29/29] block: Set bfq-dev instead of bfq-mq

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 include/linux/blkdev.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/blkdev.h b/include/linux/blkdev.h
index bc17d1e26d7b..8c34a0fb51ba 100644
--- a/include/linux/blkdev.h
+++ b/include/linux/blkdev.h
@@ -109,7 +109,7 @@ typedef __u32 __bitwise req_flags_t;
 #define RQF_MQ_POLL_SLEPT	((__force req_flags_t)(1 << 20))
 /* ->timeout has been called, don't expire again */
 #define RQF_TIMED_OUT		((__force req_flags_t)(1 << 21))
-/* DEBUG: rq in bfq-mq dispatch list */
+/* DEBUG: rq in bfq-dev dispatch list */
 #define RQF_DISP_LIST   ((__force req_flags_t)(1 << 22))
 /* DEBUG: rq had get_rq_private executed on it */
 #define RQF_GOT ((__force req_flags_t)(1 << 23))
-- 
2.23.0.385.gbc12974a89

