From a6ed4c54bfd2239d0086c3d60ee0ef8907af5ccf Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 14 May 2019 09:24:48 +0800
Subject: [PATCH 1/6] bmq: Use find_first_bit() in rq_first_bmq_task().

---
 kernel/sched/bmq.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 119ddc130c3e..a9291a7505c6 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -258,7 +258,11 @@ __rq_next_bmq_task(const struct rq *rq, unsigned long offset)
 
 static inline struct task_struct *rq_first_bmq_task(struct rq *rq)
 {
-	return __rq_next_bmq_task(rq, 0UL);
+	unsigned long idx = find_first_bit(rq->queue.bitmap, bmq_BITS);
+	const struct list_head *head = &rq->queue.heads[idx];
+
+	BUG_ON(list_empty(head));
+	return list_first_entry(head, struct task_struct, bmq_node);
 }
 
 static inline struct task_struct *
-- 
2.22.0.rc3

