From be81e26d852ebd77f94d44c5a872fbced58f25c7 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 6 Jun 2019 14:57:23 +0800
Subject: [PATCH 6/6] Tag BMQ v0.96

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 70296f9e126f..13d1192e6074 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.95 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.96 by Alfred Chen.\n");
 }
 
 /**
-- 
2.22.0.rc3

