From 1438ace877115e53daca7b55a2a20b83228b9634 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 19 Oct 2022 18:34:43 +0200
Subject: [PATCH 30/30] lrng-6.0: merge changes from dev tree

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 drivers/char/lrng/Kconfig | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/drivers/char/lrng/Kconfig b/drivers/char/lrng/Kconfig
index 9d74117bc..08ab921a1 100644
--- a/drivers/char/lrng/Kconfig
+++ b/drivers/char/lrng/Kconfig
@@ -10,6 +10,12 @@ config RANDOM_DEFAULT_IMPL
 	  The default random number generator as provided with
 	  drivers/char/random.c is selected with this option.
 
+config LRNG_AUTO_SELECTED
+	bool
+	default y if !RANDOM_DEFAULT_IMPL
+	default n if RANDOM_DEFAULT_IMPL
+	select LRNG
+
 config LRNG
 	bool "Linux Random Number Generator"
 	default n
-- 
2.38.0.rc1.6.g4fd6c5e444

