From b233315b61984d22d0ce9a08b5be67b272c23187 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sat, 20 Feb 2021 08:37:18 +0100
Subject: [PATCH 4/6] mm-5.11: simplify file_is_low

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 2032af8d8..f9fcde30d 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2442,9 +2442,8 @@ static void get_scan_count(struct lruvec *lruvec, struct scan_control *sc,
 
 #if defined(CONFIG_UNEVICTABLE_FILE)
 		if (file && scan) {
-			unsigned long low_scan_granularity = SWAP_CLUSTER_MAX >> sc->priority;
-			if (sc->file_is_low && scan > low_scan_granularity)
-				scan = low_scan_granularity;
+			if (sc->file_is_low)
+				scan = min(scan, SWAP_CLUSTER_MAX >> sc->priority);
 			else if (sc->file_is_min)
 				scan = 0;
 		}
-- 
2.31.0.97.g1424303384

