From 5e703bedf51de10a62a3289a61126a2ed6b99211 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 6 Mar 2019 11:13:21 +0800
Subject: [PATCH 05/43] Tag  BMQ v0.88

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index de75483806a3..c5901f7af93d 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -75,7 +75,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.87 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.88 by Alfred Chen.\n");
 }
 
 #define SCHED_TIMESLICE_NS	MS_TO_NS(CONFIG_SCHED_TIMESLICE)
-- 
2.21.0.777.g83232e3864

