From 9956fed71cf711ad50fb993da2e4e257f51c6a29 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 7 Mar 2019 11:01:18 +0800
Subject: [PATCH 08/43] bmq: Remove unused wakexxx fields in BMQ.

---
 include/linux/sched.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 865c686cb5bd..b21376df579a 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -616,11 +616,11 @@ struct task_struct {
 	/* Current CPU: */
 	unsigned int			cpu;
 #endif
+#ifndef CONFIG_SCHED_BMQ
 	unsigned int			wakee_flips;
 	unsigned long			wakee_flip_decay_ts;
 	struct task_struct		*last_wakee;
 
-#ifndef CONFIG_SCHED_BMQ
 	/*
 	 * recent_used_cpu is initially set as the last CPU used by a task
 	 * that wakes affine another task. Waker/wakee relationships can
@@ -629,8 +629,8 @@ struct task_struct {
 	 * used CPU that may be idle.
 	 */
 	int				recent_used_cpu;
-#endif /* CONFIG_SCHED_BMQ */
 	int				wake_cpu;
+#endif /* !CONFIG_SCHED_BMQ */
 #endif
 	int				on_rq;
 
-- 
2.21.0.777.g83232e3864

