From 26e61b426c58e65249f0113161216e1a1fd7b973 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 29 Apr 2019 15:29:47 +0800
Subject: [PATCH 34/43] Tag  BMQ v0.93

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index e56362710c0f..f12c6ea22e19 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.92 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.93 by Alfred Chen.\n");
 }
 
 /**
-- 
2.21.0.777.g83232e3864

