From f0f567abc692354af3f55fabe5072c53d1ec7187 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 20 Apr 2019 15:48:56 +0800
Subject: [PATCH 37/43] bmq: [Sync] 8dabe7245bbc y2038: syscalls: rename y2038
 compat syscalls

---
 kernel/sched/bmq.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index b1a7f1c1c377..1d7c4a9b986d 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -4885,9 +4885,8 @@ SYSCALL_DEFINE2(sched_rr_get_interval, pid_t, pid,
 }
 
 #ifdef CONFIG_COMPAT_32BIT_TIME
-COMPAT_SYSCALL_DEFINE2(sched_rr_get_interval,
-		       compat_pid_t, pid,
-		       struct old_timespec32 __user *, interval)
+SYSCALL_DEFINE2(sched_rr_get_interval_time32, pid_t, pid,
+		struct old_timespec32 __user *, interval)
 {
 	struct timespec64 t;
 	int retval = sched_rr_get_interval(pid, &t);
-- 
2.21.0.777.g83232e3864

