From bd62ae848f9ab8f9091b14bd45a09f11104d7466 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 23 Jan 2023 20:12:57 -0500
Subject: [PATCH 48/51] fixup! lib: implement context capture support for page
 and slab allocators

---
 lib/alloc_tag.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/alloc_tag.c b/lib/alloc_tag.c
index d6877a212..49f6b4d25 100644
--- a/lib/alloc_tag.c
+++ b/lib/alloc_tag.c
@@ -8,6 +8,7 @@
 #include <linux/page_ext.h>
 #include <linux/sched/clock.h>
 #include <linux/seq_buf.h>
+#include <linux/stackdepot.h>
 #include <linux/uaccess.h>
 
 #define STACK_BUF_SIZE 1024
-- 
2.39.0.rc2.1.gbd5df96b79

