From 7cb4c21609e7341ef1b0031d512d41cb01dd6bfb Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 23 Jan 2023 21:55:54 -0500
Subject: [PATCH 51/51] arm64: Fix circular header dependency

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 arch/arm64/include/asm/spectre.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm64/include/asm/spectre.h b/arch/arm64/include/asm/spectre.h
index db7b371b3..2129a28ec 100644
--- a/arch/arm64/include/asm/spectre.h
+++ b/arch/arm64/include/asm/spectre.h
@@ -14,7 +14,7 @@
 
 #ifndef __ASSEMBLY__
 
-#include <linux/percpu.h>
+#include <linux/percpu-defs.h>
 
 #include <asm/cpufeature.h>
 #include <asm/virt.h>
-- 
2.39.0.rc2.1.gbd5df96b79

