From 95b1609254300f87dd3e02a9b2d1e7720ead770f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 16 Feb 2021 01:08:58 +0100
Subject: [PATCH] Documentation: Fix: sphinx 3.5 causes compilation errors

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Documentation/conf.py | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/Documentation/conf.py b/Documentation/conf.py
index ed2b43ec7..10391e719 100644
--- a/Documentation/conf.py
+++ b/Documentation/conf.py
@@ -264,11 +264,9 @@ except ImportError:
 
 html_static_path = ['sphinx-static']
 
-html_context = {
-    'css_files': [
-        '_static/theme_overrides.css',
-    ],
-}
+html_css_files = [
+    '_static/theme_overrides.css',
+]
 
 # Add any extra paths that contain custom files (such as robots.txt or
 # .htaccess) here, relative to this directory. These files are copied
-- 
2.30.1.457.gf011795891

