From fe4ee6f160969d04d8f6f5ae2349a475f65ebe9c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 20 Mar 2019 13:33:14 +0800
Subject: [PATCH 2/7] bmq: treat NORMAL task as dl for lower ttwu latency.

---
 include/linux/sched/deadline.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/sched/deadline.h b/include/linux/sched/deadline.h
index 9da015d7bac7..b66abbae96ca 100644
--- a/include/linux/sched/deadline.h
+++ b/include/linux/sched/deadline.h
@@ -11,7 +11,7 @@ static inline int dl_prio(int prio)
 
 static inline int dl_task(struct task_struct *p)
 {
-	return 0;
+	return (SCHED_NORMAL == p->policy);
 }
 #else
 
-- 
2.21.0.196.g041f5ea1cf

