From 96d584e2346c6ec278a6bf72e9421a8564e9ef65 Mon Sep 17 00:00:00 2001
From: "J. R. Okajima" <hooanon05g@gmail.com>
Date: Thu, 11 Jul 2019 22:51:28 +0900
Subject: [PATCH 1/2] aufs: version 5.2

Signed-off-by: J. R. Okajima <hooanon05g@gmail.com>
---
 include/uapi/linux/aufs_type.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/uapi/linux/aufs_type.h b/include/uapi/linux/aufs_type.h
index 0e3e2b175..e00e914b7 100644
--- a/include/uapi/linux/aufs_type.h
+++ b/include/uapi/linux/aufs_type.h
@@ -40,7 +40,7 @@
 
 #include <linux/limits.h>
 
-#define AUFS_VERSION	"5.x-rcN-20190610"
+#define AUFS_VERSION	"5.2"
 
 /* todo? move this to linux-2.6.19/include/magic.h */
 #define AUFS_SUPER_MAGIC	('a' << 24 | 'u' << 16 | 'f' << 8 | 's')
-- 
2.22.0.214.g8dca754b1e


From 2b87be412a9934a674f872b910d3e6b3c984fbef Mon Sep 17 00:00:00 2001
From: "J. R. Okajima" <hooanon05g@gmail.com>
Date: Fri, 12 Jul 2019 22:29:09 +0900
Subject: [PATCH 2/2] aufs5.2 20190715

Signed-off-by: J. R. Okajima <hooanon05g@gmail.com>
---
 include/uapi/linux/aufs_type.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/uapi/linux/aufs_type.h b/include/uapi/linux/aufs_type.h
index e00e914b7..4e16fdc20 100644
--- a/include/uapi/linux/aufs_type.h
+++ b/include/uapi/linux/aufs_type.h
@@ -40,7 +40,7 @@
 
 #include <linux/limits.h>
 
-#define AUFS_VERSION	"5.2"
+#define AUFS_VERSION	"5.2-20190715"
 
 /* todo? move this to linux-2.6.19/include/magic.h */
 #define AUFS_SUPER_MAGIC	('a' << 24 | 'u' << 16 | 'f' << 8 | 's')
-- 
2.22.0.214.g8dca754b1e

