From bd2d487322b1669ccebdc2ad9bade8a5e428e4f7 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 10 Nov 2018 11:21:49 +0100
Subject: [PATCH 4/4] LL: Add .ll version

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 26c92f892..6479a03be 100644
--- a/Makefile
+++ b/Makefile
@@ -11,6 +11,10 @@ NAME = Shy Crocodile
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLNAME = linux-lucjan
+LLVERSION = .ll16
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
-- 
2.21.0.777.g83232e3864

