From 2da31e46d68edf53dd766dfe2e65480ca1541a8c Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 22 Sep 2020 13:41:10 +0200
Subject: [PATCH 21/36] bfq-dev: update version to v13

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index fbc24caa2..ce6af3b3a 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7684,7 +7684,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ-dev I/O-scheduler: v12";
+	char msg[60] = "BFQ-dev I/O-scheduler: v13";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.30.0.155.g66e871b664

