From fa0ded6124354740f382f0bccc5672d8029149f4 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 2 Dec 2019 15:05:30 +0100
Subject: [PATCH 06/19] Makefile: add -bfq version suffix

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 4d0711f54047..4e02f4ad8014 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 6
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION = -bfq
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.26.2.266.ge870325ee8

