From fe509287aa24169668cc4dd753b954c73dffa68f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 21 Sep 2020 17:26:14 +0200
Subject: [PATCH 2/2] init, Kconfig: Restore original PDS description

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index b59b0e60a62a..14223ca767e9 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -797,7 +797,9 @@ config SCHED_PDS
 	bool "PDS CPU scheduler"
 	help
 	  The Priority and Deadline based Skip list multiple queue CPU
-	  Scheduler.
+	  scheduler for excellent interactivity and responsiveness on the
+	  desktop and solid scalability on normal hardware and commodity
+	  servers.
 
 endchoice
 
-- 
2.30.1.457.gf011795891

