From 64c139894fb8f521c75e8f28acb2d59163e6c393 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 15 Jan 2019 16:18:57 +0100
Subject: [PATCH 1/2] elevator: set default scheduler to bfq-mq for blk-mq

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/elevator.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/block/elevator.c b/block/elevator.c
index 8fdcd64ae12e..fccde9ac3bc5 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -948,8 +948,8 @@ int elevator_switch_mq(struct request_queue *q,
 }
 
 /*
- * For blk-mq devices, we default to using mq-deadline, if available, for single
- * queue devices.  If deadline isn't available OR we have multiple queues,
+ * For blk-mq devices, we default to using bfq-mq, if available, for single
+ * queue devices.  If bfq-mq isn't available OR we have multiple queues,
  * default to "none".
  */
 int elevator_init_mq(struct request_queue *q)
@@ -968,7 +968,7 @@ int elevator_init_mq(struct request_queue *q)
 	if (unlikely(q->elevator))
 		goto out_unlock;
 
-	e = elevator_get(q, "mq-deadline", false);
+	e = elevator_get(q, "bfq-mq", false);
 	if (!e)
 		goto out_unlock;
 
-- 
2.20.1.390.gb5101f9297

