From ce04399bdf37db678e5537b1552855225d24d559 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 10 Nov 2018 11:21:49 +0100
Subject: [PATCH 4/4] LL: Add .ll version

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 3e4868a64..0e392a01d 100644
--- a/Makefile
+++ b/Makefile
@@ -11,6 +11,10 @@ NAME = Bobtail Squid
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLNAME = linux-lucjan
+LLVERSION = .ll38
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
-- 
2.22.0.214.g8dca754b1e

