From 3689467b60d2d313d23f5fa310f4cc8d786be396 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 21 Feb 2023 09:21:06 +0100
Subject: [PATCH 4/4] Kconfig, zstd: Rename COMP_VAL to COMPRESSION_LEVEL

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig         | 2 +-
 scripts/Makefile.lib | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 12a698ff9..1342c107f 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -351,7 +351,7 @@ endchoice
 menu "ZSTD compression options"
 	depends on KERNEL_ZSTD
 
-config ZSTD_COMP_VAL
+config ZSTD_COMPRESSION_LEVEL
 	int "Compression level (1-22)"
 	range 1 22
 	default "22"
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index f45507b22..993e4578c 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -557,8 +557,8 @@ quiet_cmd_xzmisc = XZMISC  $@
 # decompression is used, like initramfs decompression, zstd22 should likely not
 # be used because it would require zstd to allocate a 128 MB buffer.
 
-ifdef CONFIG_ZSTD_COMP_VAL
-zstd_comp_val	:= $(CONFIG_ZSTD_COMP_VAL)
+ifdef CONFIG_ZSTD_COMPRESSION_LEVEL
+zstd_comp_val	:= $(CONFIG_ZSTD_COMPRESSION_LEVEL)
 ifeq ($(shell test $(zstd_comp_val) -gt 19; echo $$?),0)
 zstd_comp_val	+= --ultra
 endif
-- 
2.39.2.501.gd9d677b2d8

