From c3eb44ee94c92cfa8c8f87587ae4e91b16ccd03b Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <lkml@antheas.dev>
Date: Sun, 28 Sep 2025 17:05:22 +0200
Subject: [PATCH 104/112] lower sleep interval to avoid waking up too long

Signed-off-by: Antheas Kapenekakis <lkml@antheas.dev>
---
 drivers/platform/x86/amd/pmc/pmc.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/platform/x86/amd/pmc/pmc.c b/drivers/platform/x86/amd/pmc/pmc.c
index cae3fcafd..79870f371 100644
--- a/drivers/platform/x86/amd/pmc/pmc.c
+++ b/drivers/platform/x86/amd/pmc/pmc.c
@@ -637,7 +637,7 @@ static void amd_pmc_s2idle_check(void)
 
 	/* Avoid triggering OVP */
 	if (!get_metrics_table(pdev, &table) && table.s0i3_last_entry_status)
-		msleep(2500);
+		msleep(500);
 
 	/* Dump the IdleMask before we add to the STB */
 	amd_pmc_idlemask_read(pdev, pdev->dev, NULL);
-- 
2.52.0

