From 6dbbb3c736f66ecca645ff776a88ac26706a4af4 Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Thu, 9 Feb 2023 03:28:48 +0000
Subject: [PATCH 51/54] fixup! mm: enable page allocation tagging

Instrument compaction_alloc().

Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 mm/compaction.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/mm/compaction.c b/mm/compaction.c
index 8238e8338..158ef2b28 100644
--- a/mm/compaction.c
+++ b/mm/compaction.c
@@ -1673,7 +1673,7 @@ static void isolate_freepages(struct compact_control *cc)
  * This is a migrate-callback that "allocates" freepages by taking pages
  * from the isolated freelists in the block we are migrating to.
  */
-static struct page *compaction_alloc(struct page *migratepage,
+static struct page *_compaction_alloc(struct page *migratepage,
 					unsigned long data)
 {
 	struct compact_control *cc = (struct compact_control *)data;
@@ -1693,6 +1693,13 @@ static struct page *compaction_alloc(struct page *migratepage,
 	return freepage;
 }
 
+static struct page *compaction_alloc(struct page *migratepage,
+				     unsigned long data)
+{
+	return alloc_hooks(_compaction_alloc(migratepage, data),
+			   struct page *, NULL);
+}
+
 /*
  * This is a migrate-callback that "frees" freepages back to the isolated
  * freelist.  All pages on the freelist are from the same zone, so there is no
-- 
2.39.1.418.g7876265d61

