From d7e99525e69a3953873768bb5c8a8ff1cfeb950c Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 30 May 2019 15:31:04 +0200
Subject: [PATCH 03/20] block, bfq, DEBUG: log think-time state change

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 594e4e8525da..33248e06d010 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5742,6 +5742,9 @@ static void bfq_update_has_short_ttime(struct bfq_data *bfqd,
 
 	state_changed = has_short_ttime != bfq_bfqq_has_short_ttime(bfqq);
 
+	bfq_log_bfqq(bfqd, bfqq, "has_short_ttime %d, changed %d",
+		     has_short_ttime, state_changed);
+
 	if (has_short_ttime)
 		bfq_mark_bfqq_has_short_ttime(bfqq);
 	else
-- 
2.23.0.162.gf1d4a28250

