From 428e3e8e264e569d8b2e5d2fd9c469c2fab561a1 Mon Sep 17 00:00:00 2001
From: Eric Biggers <ebiggers@google.com>
Date: Fri, 18 Oct 2024 15:37:18 -0700
Subject: [PATCH 46/73] crc32test: stop disabling interrupts

Signed-off-by: Eric Biggers <ebiggers@google.com>
---
 lib/crc32test.c | 12 ------------
 1 file changed, 12 deletions(-)

diff --git a/lib/crc32test.c b/lib/crc32test.c
index f073f19b5..3dd1be8fd 100644
--- a/lib/crc32test.c
+++ b/lib/crc32test.c
@@ -667,7 +667,6 @@ static int __init crc32c_test(void)
 	int errors = 0;
 	int bytes = 0;
 	u64 nsec;
-	unsigned long flags;
 
 	/* keep static to prevent cache warming code from
 	 * getting eliminated by the compiler */
@@ -681,9 +680,6 @@ static int __init crc32c_test(void)
 		    test[i].start, test[i].length);
 	}
 
-	/* reduce OS noise */
-	local_irq_save(flags);
-
 	nsec = ktime_get_ns();
 	for (i = 0; i < ARRAY_SIZE(test); i++) {
 		if (test[i].crc32c_le != __crc32c_le(test[i].crc, test_buf +
@@ -692,8 +688,6 @@ static int __init crc32c_test(void)
 	}
 	nsec = ktime_get_ns() - nsec;
 
-	local_irq_restore(flags);
-
 	pr_info("crc32c: CRC_LE_BITS = %d\n", CRC_LE_BITS);
 
 	if (errors)
@@ -747,7 +741,6 @@ static int __init crc32_test(void)
 	int errors = 0;
 	int bytes = 0;
 	u64 nsec;
-	unsigned long flags;
 
 	/* keep static to prevent cache warming code from
 	 * getting eliminated by the compiler */
@@ -764,9 +757,6 @@ static int __init crc32_test(void)
 		    test[i].start, test[i].length);
 	}
 
-	/* reduce OS noise */
-	local_irq_save(flags);
-
 	nsec = ktime_get_ns();
 	for (i = 0; i < ARRAY_SIZE(test); i++) {
 		if (test[i].crc_le != crc32_le(test[i].crc, test_buf +
@@ -779,8 +769,6 @@ static int __init crc32_test(void)
 	}
 	nsec = ktime_get_ns() - nsec;
 
-	local_irq_restore(flags);
-
 	pr_info("crc32: CRC_LE_BITS = %d, CRC_BE BITS = %d\n",
 		 CRC_LE_BITS, CRC_BE_BITS);
 
-- 
2.47.0

