From ed395a9600baa38ad2beb2e968cf622c74a653b6 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Fri, 20 Jan 2023 11:17:56 -0500
Subject: [PATCH 45/45] fixup! mm: percpu: Alloc tagging

---
 include/linux/percpu.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/percpu.h b/include/linux/percpu.h
index 84c41fd32..482deb07e 100644
--- a/include/linux/percpu.h
+++ b/include/linux/percpu.h
@@ -9,6 +9,7 @@
 #include <linux/cpumask.h>
 #include <linux/pfn.h>
 #include <linux/init.h>
+#include <linux/sched.h>
 
 #include <asm/percpu.h>
 
-- 
2.39.0.rc2.1.gbd5df96b79

