From be2448fdaaa1a8d5318ec47ea798eb2c7c7137fa Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 12 Dec 2024 11:36:54 +0100
Subject: [PATCH 1/2] ovpn-dco-6.12: introduce NLA_POLICY_MAX_LEN macro

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 include/net/netlink.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/net/netlink.h b/include/net/netlink.h
index db6af2072..2dc671c97 100644
--- a/include/net/netlink.h
+++ b/include/net/netlink.h
@@ -469,6 +469,7 @@ struct nla_policy {
 	.max = _len						\
 }
 #define NLA_POLICY_MIN_LEN(_len)	NLA_POLICY_MIN(NLA_BINARY, _len)
+#define NLA_POLICY_MAX_LEN(_len)	NLA_POLICY_MAX(NLA_BINARY, _len)
 
 /**
  * struct nl_info - netlink source information
-- 
2.47.1

