From c43c62efa59dc6a25f34b5f7986e6abc5c893fce Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Thu, 17 Aug 2023 13:53:54 +0200
Subject: [PATCH] Cachy: Allow -O3

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile     | 3 +++
 init/Kconfig | 6 ++++++
 2 files changed, 9 insertions(+)

diff --git a/Makefile b/Makefile
index b9c5792c7..d45645f58 100644
--- a/Makefile
+++ b/Makefile
@@ -929,6 +929,9 @@ KBUILD_CFLAGS	+= -fno-delete-null-pointer-checks
 ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE
 KBUILD_CFLAGS += -O2
 KBUILD_RUSTFLAGS += -Copt-level=2
+else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
+KBUILD_CFLAGS += -O3
+KBUILD_RUSTFLAGS += -Copt-level=3
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os
 KBUILD_RUSTFLAGS += -Copt-level=s
diff --git a/init/Kconfig b/init/Kconfig
index 5230d4879..75dbc7540 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1613,6 +1613,12 @@ config CC_OPTIMIZE_FOR_PERFORMANCE
 	  with the "-O2" compiler flag for best performance and most
 	  helpful compile-time warnings.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE_O3
+	bool "Optimize more for performance (-O3)"
+	help
+	  Choosing this option will pass "-O3" to your compiler to optimize
+	  the kernel yet more for performance.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
 	help
-- 
2.54.0

