From f2bfded54d47f4a11c1cff2eaf6b422d119c2750 Mon Sep 17 00:00:00 2001
From: Venkata Prasad Potturu <venkataprasad.potturu@amd.com>
Date: Mon, 5 Jan 2026 16:59:45 +0530
Subject: [PATCH 004/188] ASoC: SOF: Change IPC timed out to 2000ms
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Venkata Prasad Potturu <venkataprasad.potturu@amd.com>
Signed-off-by: Nícolas F. R. A. Prado <nfraprado@collabora.com>
---
 sound/soc/sof/core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/soc/sof/core.c b/sound/soc/sof/core.c
index 2d394389c..1bc288f4d 100644
--- a/sound/soc/sof/core.c
+++ b/sound/soc/sof/core.c
@@ -61,7 +61,7 @@ MODULE_PARM_DESC(boot_timeout,
 #endif
 
 /* SOF defaults if not provided by the platform in ms */
-#define TIMEOUT_DEFAULT_IPC_MS  500
+#define TIMEOUT_DEFAULT_IPC_MS  2000
 #define TIMEOUT_DEFAULT_BOOT_MS 2000
 
 /**
-- 
2.55.0

