From e045c3f3beff40572a7741b7198c9ae99885bc6c Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sat, 13 Jan 2018 22:07:32 +0100
Subject: [PATCH] fixes-4.18: unmask CHECKPOINT_RESTORE

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 041f3a022122..2aec66b3d458 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1349,7 +1349,7 @@ config MEMBARRIER
 	  If unsure, say Y.
 
 config CHECKPOINT_RESTORE
-	bool "Checkpoint/restore support" if EXPERT
+	bool "Checkpoint/restore support"
 	select PROC_CHILDREN
 	default n
 	help
-- 
2.18.0.547.g1d89318c4

