From 5a6848e91dac06c6bd48f4f043dd485c6b6050db Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sat, 13 Jan 2018 22:07:32 +0100
Subject: [PATCH] fixes-4.17: unmask CHECKPOINT_RESTORE

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 18b151f0ddc1..cefe12e474f7 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1325,7 +1325,7 @@ config MEMBARRIER
 	  If unsure, say Y.
 
 config CHECKPOINT_RESTORE
-	bool "Checkpoint/restore support" if EXPERT
+	bool "Checkpoint/restore support"
 	select PROC_CHILDREN
 	default n
 	help
-- 
2.18.0.rc0

