From d69f4354e8de0c90fbf07c523f68111f66472ea6 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 19 Oct 2021 13:47:47 +0200
Subject: [PATCH 3/4] LL: Disable RDB by default

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index fbb1aa393..694c4c00d 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -877,14 +877,14 @@ config CACULE_SCHED
 
 config CACULE_RDB
 	bool "RDB (Response Driven Balancer)"
-	default y
+	default n
 	depends on CACULE_SCHED
 	help
 	  This is an experimental load balancer for CacULE. It is a lightweight
 	  load balancer which is a replacement of CFS load balancer. It migrates
 	  tasks based on their interactivity scores.
 
-	  If unsure, say Y here.
+	  If unsure, say N here.
 
 config RDB_INTERVAL
 	int "RDB load balancer interval"
-- 
2.34.1.75.gabe6bb3905

