From 9b8b64d49698ee6ea355f73a8e5f5ddca54b196a Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sat, 19 Apr 2025 11:40:56 +0200
Subject: [PATCH 24/42] ovpn-dco-6.14: undo 69c7be1b903fca

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/net/ovpn/main.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/drivers/net/ovpn/main.c b/drivers/net/ovpn/main.c
index 0acb0934c..0785cd706 100644
--- a/drivers/net/ovpn/main.c
+++ b/drivers/net/ovpn/main.c
@@ -169,12 +169,11 @@ static void ovpn_setup(struct net_device *dev)
 	SET_NETDEV_DEVTYPE(dev, &ovpn_type);
 }
 
-static int ovpn_newlink(struct net_device *dev,
-			struct rtnl_newlink_params *params,
+static int ovpn_newlink(struct net *net, struct net_device *dev,
+			struct nlattr *tb[], struct nlattr *data[],
 			struct netlink_ext_ack *extack)
 {
 	struct ovpn_priv *ovpn = netdev_priv(dev);
-	struct nlattr **data = params->data;
 	enum ovpn_mode mode = OVPN_MODE_P2P;
 
 	if (data && data[IFLA_OVPN_MODE]) {
-- 
2.49.0

