From ade90498015caeda98a0c517419b389da4c71521 Mon Sep 17 00:00:00 2001
From: Alexandre Frade <kernel@xanmod.org>
Date: Tue, 14 Dec 2021 16:53:15 +0000
Subject: [PATCH 3/3] wait: Add EXPORT_SYMBOL for __wake_up_pollfree

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 kernel/sched/wait.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/wait.c b/kernel/sched/wait.c
index eca3810..1034b23 100644
--- a/kernel/sched/wait.c
+++ b/kernel/sched/wait.c
@@ -244,6 +244,7 @@ void __wake_up_pollfree(struct wait_queue_head *wq_head)
 	/* POLLFREE must have cleared the queue. */
 	WARN_ON_ONCE(waitqueue_active(wq_head));
 }
+EXPORT_SYMBOL(__wake_up_pollfree);
 
 /*
  * Note: we use "set_current_state()" _after_ the wait-queue add,
-- 
2.34.1.75.gabe6bb3905

