From eee5b213149b6a078f3ddf43cb6e547054a9f874 Mon Sep 17 00:00:00 2001
From: bakedpotato191 <vladislavp@pm.me>
Date: Fri, 3 Dec 2021 21:19:28 +0100
Subject: [PATCH 2/2] Add module_version macro

---
 drivers/media/v4l2-core/v4l2loopback.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/media/v4l2-core/v4l2loopback.c b/drivers/media/v4l2-core/v4l2loopback.c
index 73defda..b31b646 100644
--- a/drivers/media/v4l2-core/v4l2loopback.c
+++ b/drivers/media/v4l2-core/v4l2loopback.c
@@ -61,6 +61,7 @@ MODULE_AUTHOR("Vasily Levin, "
 	      "Stefan Diewald,"
 	      "Anton Novikov"
 	      "et al.");
+MODULE_VERSION("0.12.5");
 MODULE_LICENSE("GPL");
 
 /*
-- 
2.34.1.75.gabe6bb3905

