From bd1f5405b67723abbf9e40afbc2c87c1c9b94a80 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 18 Mar 2019 13:30:52 +0800
Subject: [PATCH 6/6] Tag  BMQ v0.90

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 1a035da944d3..a393df2be45f 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.89 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.90 by Alfred Chen.\n");
 }
 
 /**
-- 
2.21.0.155.ge902e9bcae

