From ee694d08363167d1dad25d578ababf01db7f1eb4 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 16 Jul 2018 22:14:20 +0800
Subject: [PATCH] pds: Fix possibe kernel warning in pds_load_balance().

---
 kernel/sched/pds.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 7cde0d7f58a3..66f6b1198024 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -452,9 +452,9 @@ __update_sched_rq_queued_masks(struct rq *rq, const int cpu,
 
 static inline void update_sched_rq_queued_masks_normal(struct rq *rq)
 {
-	struct task_struct *p = rq->curr;
+	struct task_struct *p = rq_first_queued_task(rq);
 
-	if (p->prio == NORMAL_PRIO && rq_first_queued_task(rq) == p) {
+	if (p != NULL && p->prio == NORMAL_PRIO) {
 		int level = task_running_policy_level(p, rq);
 		int last_level = rq->queued_level;
 
-- 
2.18.0.129.ge3331758f1

