From a27f5a688e092b2ce550f98091a9b0dd7a963ee8 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 19 Dec 2019 11:02:38 +0800
Subject: [PATCH 2/2] Tag BMQ v5.4-r1

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 90b6b8ac7acf..5c5603249b78 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r0 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r1 by Alfred Chen.\n");
 }
 
 /**
-- 
2.24.1.590.gb02fd2acca

