From e6fb493070387f0f1ea726f1dc7f4cbf1fc5e025 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Sat, 24 Feb 2024 16:59:53 -0500
Subject: [PATCH 40/40] codetag.idx

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 include/linux/codetag.h | 11 ++++++-----
 lib/codetag.c           |  4 ++++
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/include/linux/codetag.h b/include/linux/codetag.h
index 6aa8cf22d..b5a58d82c 100644
--- a/include/linux/codetag.h
+++ b/include/linux/codetag.h
@@ -19,11 +19,12 @@ struct module;
  * an array of these.
  */
 struct codetag {
-	unsigned int flags; /* used in later patches */
-	unsigned int lineno;
-	const char *modname;
-	const char *function;
-	const char *filename;
+	u16		idx;
+	u16		flags; /* used in later patches */
+	u32		lineno;
+	const char	*modname;
+	const char	*function;
+	const char	*filename;
 } __aligned(8);
 
 union codetag_ref {
diff --git a/lib/codetag.c b/lib/codetag.c
index 5890cecab..8c47d3d4a 100644
--- a/lib/codetag.c
+++ b/lib/codetag.c
@@ -244,6 +244,10 @@ static int __codetag_module_init(struct codetag_type *cttype, struct module *mod
 
 	rcu_assign_pointer(cttype->mods_rcu, new_mods);
 	kfree_rcu(old_mods, rcu);
+
+	for (unsigned i = 0; i < cmod.nr; i++)
+		__idx_to_codetag(cttype, &cmod, cmod.idx + i)->idx = cmod.idx + i;
+
 	return 0;
 }
 
-- 
2.43.0.232.ge79552d197

