From f707dbaa78b57cb102cf0e18dffd86147f6aa214 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 25 Jun 2022 00:22:05 +0200
Subject: [PATCH 34/36] lrng-5.18: merge changes from dev tree

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 drivers/char/lrng/lrng_switch.c | 42 +++++++++------------------------
 1 file changed, 11 insertions(+), 31 deletions(-)

diff --git a/drivers/char/lrng/lrng_switch.c b/drivers/char/lrng/lrng_switch.c
index c5870aefd..5ffc0286a 100644
--- a/drivers/char/lrng/lrng_switch.c
+++ b/drivers/char/lrng/lrng_switch.c
@@ -174,32 +174,6 @@ static int lrng_switch(const void *cb,
 	return 0;
 }
 
-/* Reset LRNG such that all existing entropy is gone */
-static const struct lrng_drng_cb *lrng_drng_cb = NULL;
-static void lrng_set_drng_cb_workfn(struct work_struct *work)
-{
-	if (!lrng_drng_cb)
-		return;
-
-	/*
-	 * Wait until the LRNG becomes operational. The reason is that
-	 * an SP800-90A DRBG shall be fully initialized with full entropy.
-	 * This is only present when the LRNG is fully initialized.
-	 */
-	lrng_drng_sleep_while_nonoperational(0);
-	lrng_switch(lrng_drng_cb, lrng_drng_switch);
-
-	/* The swtich may imply new entropy due to larger DRNG sec strength. */
-	if (!lrng_switch(lrng_drng_cb, lrng_drng_switch))
-		lrng_es_add_entropy();
-
-	lrng_drng_cb = NULL;
-
-	mutex_unlock(&lrng_crypto_cb_update);
-}
-
-static DECLARE_WORK(lrng_set_drng_cb_work, lrng_set_drng_cb_workfn);
-
 /*
  * lrng_set_drng_cb - Register new cryptographic callback functions for DRNG
  * The registering implies that all old DRNG states are replaced with new
@@ -215,6 +189,7 @@ static DECLARE_WORK(lrng_set_drng_cb_work, lrng_set_drng_cb_workfn);
 int lrng_set_drng_cb(const struct lrng_drng_cb *drng_cb)
 {
 	struct lrng_drng *lrng_drng_init = lrng_drng_init_instance();
+	int ret;
 
 	if (!IS_ENABLED(CONFIG_LRNG_SWITCH_DRNG))
 		return -EOPNOTSUPP;
@@ -236,13 +211,18 @@ int lrng_set_drng_cb(const struct lrng_drng_cb *drng_cb)
 	if ((drng_cb != lrng_default_drng_cb) &&
 	    (lrng_drng_init->drng_cb != lrng_default_drng_cb)) {
 		pr_warn("disallow setting new DRNG callbacks, unload the old callbacks first!\n");
-		mutex_unlock(&lrng_crypto_cb_update);
-		return -EINVAL;
+		ret = -EINVAL;
+		goto out;
 	}
 
-	lrng_drng_cb = drng_cb;
-	schedule_work(&lrng_set_drng_cb_work);
-	return 0;
+	ret = lrng_switch(drng_cb, lrng_drng_switch);
+	/* The swtich may imply new entropy due to larger DRNG sec strength. */
+	if (!ret)
+		lrng_es_add_entropy();
+
+out:
+	mutex_unlock(&lrng_crypto_cb_update);
+	return ret;
 }
 EXPORT_SYMBOL(lrng_set_drng_cb);
 
-- 
2.37.0.rc0.15.g3b9a5a33c2

