From 701213ef4a9d9175f9e6b6dea7073fee08a056b3 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 23 Nov 2021 19:11:26 +0100
Subject: [PATCH 34/35] mm/damon/dbgfs: fix missing damon_dbgfs_lock

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 mm/damon/dbgfs.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mm/damon/dbgfs.c b/mm/damon/dbgfs.c
index 670fbcdcb..881ccb254 100644
--- a/mm/damon/dbgfs.c
+++ b/mm/damon/dbgfs.c
@@ -894,6 +894,7 @@ static ssize_t dbgfs_monitor_on_write(struct file *file,
 		for (i = 0; i < dbgfs_nr_ctxs; i++) {
 			if (damon_targets_empty(dbgfs_ctxs[i])) {
 				kfree(kbuf);
+				mutex_unlock(&damon_dbgfs_lock);
 				return -EINVAL;
 			}
 		}
-- 
2.34.1.75.gabe6bb3905

