From 1867eebf36fbc6014e6bbbc0eba0aa6a01a46e39 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Tue, 11 Jun 2024 22:57:37 +0200
Subject: [PATCH 20/23] cpufreq: amd-pstate: remove forgotten dev printk

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/cpufreq/amd-pstate.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/cpufreq/amd-pstate.c b/drivers/cpufreq/amd-pstate.c
index ffc11aa70..bb5e447f1 100644
--- a/drivers/cpufreq/amd-pstate.c
+++ b/drivers/cpufreq/amd-pstate.c
@@ -1996,7 +1996,6 @@ static int __init amd_pstate_init(void)
 	 * command line choice will override the kconfig option
 	 */
 	if (cppc_state == AMD_PSTATE_UNDEFINED) {
-		pr_err("pyuan cppc_state == AMD_PSTATE_UNDEFINED \n");
 		/* Disable on the following configs by default:
 		 * 1. Undefined platforms
 		 * 2. Server platforms
-- 
2.45.1.145.g83f1add914

