From 802fdc3e2b59920b01ccb75dfad22b51f1469944 Mon Sep 17 00:00:00 2001
From: Eric Biggers <ebiggers@google.com>
Date: Fri, 18 Oct 2024 15:37:18 -0700
Subject: [PATCH 44/72] crc32test: add code that regenerates the test cases

Signed-off-by: Eric Biggers <ebiggers@google.com>
---
 lib/crc32test.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/lib/crc32test.c b/lib/crc32test.c
index 3dd1be8fd..03cf5c1f2 100644
--- a/lib/crc32test.c
+++ b/lib/crc32test.c
@@ -817,6 +817,19 @@ static int __init crc32_combine_test(void)
 	return 0;
 }
 
+static void crc32test_regenerate(void)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(test); i++) {
+		pr_info("{0x%08x, 0x%08x, 0x%08x, 0x%08x, 0x%08x, 0x%08x},\n",
+			test[i].crc, test[i].start, test[i].length,
+			crc32_le_base(test[i].crc, test_buf + test[i].start, test[i].length),
+			crc32_be_base(test[i].crc, test_buf + test[i].start, test[i].length),
+			__crc32c_le_base(test[i].crc, test_buf + test[i].start, test[i].length));
+	}
+}
+
 static int __init crc32test_init(void)
 {
 	crc32_test();
@@ -825,6 +838,9 @@ static int __init crc32test_init(void)
 	crc32_combine_test();
 	crc32c_combine_test();
 
+	if (0)
+		crc32test_regenerate();
+
 	return 0;
 }
 
-- 
2.47.1

