From 69a565a3a769c0208ac4fda693a825291e6f3da0 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Tue, 24 Sep 2024 19:28:33 +0200
Subject: [PATCH 2/3] Cachy: Allow x86-64-v4 Optimization

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 arch/x86/Kconfig.cpu | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig.cpu b/arch/x86/Kconfig.cpu
index 521dfeaf9..46130801c 100644
--- a/arch/x86/Kconfig.cpu
+++ b/arch/x86/Kconfig.cpu
@@ -625,7 +625,7 @@ config X86_GENERIC
 
 config X86_64_VERSION
 	int "x86-64 compiler ISA level"
-	range 1 3
+	range 1 4
 	depends on (CC_IS_GCC && GCC_VERSION > 110000) || (CC_IS_CLANG && CLANG_VERSION >= 120000)
 	depends on X86_64 && GENERIC_CPU
 	help
-- 
2.47.1

