From ca7035c587ff92b0e55e8de8e755bcc5ec504d08 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 23 Sep 2019 09:58:37 +0800
Subject: [PATCH 2/3] bmq: Fix task_sched_prio()

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index e753e34886a3..2ed179a833f4 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -212,7 +212,7 @@ static inline void update_sched_rq_watermark(struct rq *rq)
 
 static inline int task_sched_prio(struct task_struct *p)
 {
-	return (p->prio < MAX_RT_PRIO)? 0:p->prio - MAX_RT_PRIO + p->boost_prio;
+	return (p->prio < MAX_RT_PRIO)? 0:p->prio - MAX_RT_PRIO + p->boost_prio + 1;
 }
 
 static inline void bmq_init(struct bmq *q)
-- 
2.23.0.385.gbc12974a89

