From 9ddb4375f24f90ba897b0edb4e29f370c1ae2ce9 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Sun, 1 Dec 2024 03:41:35 +0800
Subject: [PATCH 1/2] CACHY: Use BFQ for SQ devices and mq-deadline for MQ
 devices

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 block/elevator.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/block/elevator.c b/block/elevator.c
index ab22542e6..ec45c8446 100644
--- a/block/elevator.c
+++ b/block/elevator.c
@@ -727,14 +727,22 @@ void elevator_set_default(struct request_queue *q)
 	if (q->tag_set->flags & BLK_MQ_F_NO_SCHED_BY_DEFAULT)
 		return;
 
+#if defined(CONFIG_CACHY) && defined(CONFIG_IOSCHED_BFQ)
 	/*
-	 * For single queue devices, default to using mq-deadline. If we
-	 * have multiple queues or mq-deadline is not available, default
-	 * to "none".
+	 * For single queue devices, default to using bfq. If we
+	 * have multiple queues and CONFIG_CACHY is enabled, default to using
+	 * mq-deadline. Otherwise, fallback to "none".
 	 */
+	if (q->nr_hw_queues == 1 || blk_mq_is_shared_tags(q->tag_set->flags))
+		ctx.name = "bfq";
+
+	if (elevator_find_get(ctx.name))
+		err = elevator_change(q, &ctx);
+#else
 	if (elevator_find_get(ctx.name) && (q->nr_hw_queues == 1 ||
 			 blk_mq_is_shared_tags(q->tag_set->flags)))
 		err = elevator_change(q, &ctx);
+#endif /* CONFIG_CACHY */
 	if (err < 0)
 		pr_warn("\"%s\" elevator initialization, failed %d, "
 			"falling back to \"none\"\n", ctx.name, err);
-- 
2.49.0

