From a1a6a47bf355c2a4d9f17fdb313306cb7a935bc4 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@redhat.com>
Date: Fri, 26 Jun 2020 09:32:08 +0200
Subject: [PATCH 21/21] fsgsbase-5.7: remove erroneous if expression bit

Signed-off-by: Oleksandr Natalenko <oleksandr@redhat.com>
---
 tools/testing/selftests/x86/fsgsbase_restore.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/testing/selftests/x86/fsgsbase_restore.c b/tools/testing/selftests/x86/fsgsbase_restore.c
index 70502a708..6fffadc51 100644
--- a/tools/testing/selftests/x86/fsgsbase_restore.c
+++ b/tools/testing/selftests/x86/fsgsbase_restore.c
@@ -70,7 +70,7 @@ static void init_seg(void)
 		.seg_not_present = 0,
 		.useable         = 0
 	};
-	if (false && syscall(SYS_modify_ldt, 1, &desc, sizeof(desc)) == 0) {
+	if (syscall(SYS_modify_ldt, 1, &desc, sizeof(desc)) == 0) {
 		printf("\tusing LDT slot 0\n");
 		asm volatile ("mov %0, %" SEG :: "rm" ((unsigned short)0x7));
 	} else {
-- 
2.27.0.112.g101b3204f3

