From b1d96b731ddc7d4569ecc8715b8b99d3c6d9822d Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Sat, 31 Dec 2022 09:10:48 +0100
Subject: [PATCH 59/59] add missing commit for vma-6.1 tree

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 mm/memory.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/mm/memory.c b/mm/memory.c
index b5f6941e0..5752bb52f 100644
--- a/mm/memory.c
+++ b/mm/memory.c
@@ -5264,6 +5264,19 @@ struct vm_area_struct *lock_vma_under_rcu(struct mm_struct *mm,
 	if (!vma)
 		goto inval;
 
+	/* A known issue to be investigated */
+	if (WARN_ON(vma <= (struct vm_area_struct *)0xFF)) {
+		trace_printk("RECEIVED VMA 0x%lx for addr 0x%lx\n",
+			(unsigned long)vma, address);
+		printk("GOT AN INVALID VMA 0x%lx for address 0x%lx\n",
+			(unsigned long)vma, address);
+#ifdef CONFIG_DEBUG_MAPLE_TREE
+		pr_err("maple state: %p %u\n", mas.node, mas.offset);
+		mt_dump(&mm->mm_mt);
+#endif
+		goto inval;
+	}
+
 	/* Only anonymous vmas are supported for now */
 	if (!vma_is_anonymous(vma))
 		goto inval;
-- 
2.39.0.rc2.1.gbd5df96b79

