From 433ba94df726c63f137b58dea546756da466f93f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 10 Nov 2018 11:21:49 +0100
Subject: [PATCH 7/7] LL: Add .ll version

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 5033c0577..5be0e2c45 100644
--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,10 @@ NAME = Superb Owl
 $(if $(filter __%, $(MAKECMDGOALS)), \
 	$(error targets prefixed with '__' are only for internal use))
 
+LLNAME = linux-lucjan
+LLVERSION = .ll123
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := __all
 __all:
-- 
2.36.1.74.g277cf0bc36

