From d180bb4338b5bda4ce1974354a64f276a1d67aa6 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 4 Nov 2024 14:51:18 +0100
Subject: [PATCH 27/27] appletbdrm: Fix compilation

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 drivers/gpu/drm/tiny/appletbdrm.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/tiny/appletbdrm.c b/drivers/gpu/drm/tiny/appletbdrm.c
index b9440ce00..7a74c8ad3 100644
--- a/drivers/gpu/drm/tiny/appletbdrm.c
+++ b/drivers/gpu/drm/tiny/appletbdrm.c
@@ -7,7 +7,7 @@
 
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
 
-#include <asm/unaligned.h>
+#include <linux/unaligned.h>
 
 #include <linux/usb.h>
 #include <linux/module.h>
-- 
2.47.0

