From 8783b6a5567de098adfd95c5d18378099309d094 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Sun, 9 Aug 2020 19:26:55 +0200
Subject: [PATCH 37/55] block, bfq, DEBUG: avoid double del of same task from
 task list

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index e83dac306..0c725a40c 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5733,7 +5733,8 @@ static void bfq_exit_bfqq(struct bfq_data *bfqd, struct bfq_queue *bfqq)
 
 	bfq_put_cooperator(bfqq);
 
-	hlist_del_init(&current->task_list_node);
+	if (!hlist_unhashed(&current->task_list_node))
+		hlist_del_init(&current->task_list_node);
 	bfq_release_process_ref(bfqd, bfqq);
 }
 
-- 
2.30.0.155.g66e871b664

