From 03259b7ceeacce95073f6c7c144f0b1a83539f7f Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Fri, 13 Dec 2024 11:39:24 +0100
Subject: [PATCH 4/4] kbuild: Fix polly config call

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 3c1408671..66e24b3fe 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -253,7 +253,7 @@ config BUILD_SALT
 
 config POLLY_CLANG
 	bool "Use Clang Polly optimizations"
-	depends on CC_IS_CLANG && $(cc-option,-mllvm -polly)
+	depends on CC_IS_CLANG && $(cc-option,-mllvm -polly -fplugin=LLVMPolly.so
 	depends on !COMPILE_TEST
 	help
 	  This option enables Clang's polyhedral loop optimizer known as
-- 
2.47.1

