From 98e7663fe0d7798be0ef7235d0c7ecbc0c38ce82 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 4 Sep 2021 08:59:59 +0000
Subject: [PATCH 206/288] sched/alt: Optimize parameter to __normal_prio().

---
 kernel/sched/alt_core.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index bdff1c65e2e9..688a6696749c 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -1140,10 +1140,10 @@ static inline void hrtick_rq_init(struct rq *rq)
 }
 #endif	/* CONFIG_SCHED_HRTICK */
 
-static inline int __normal_prio(int policy, int rt_prio, int nice)
+static inline int __normal_prio(int policy, int rt_prio, int static_prio)
 {
 	return rt_policy(policy) ? (MAX_RT_PRIO - 1 - rt_prio) :
-		NICE_TO_PRIO(nice) + MAX_PRIORITY_ADJ;
+		static_prio + MAX_PRIORITY_ADJ;
 }
 
 /*
@@ -1155,7 +1155,7 @@ static inline int __normal_prio(int policy, int rt_prio, int nice)
  */
 static inline int normal_prio(struct task_struct *p)
 {
-	return __normal_prio(p->policy, p->rt_priority, PRIO_TO_NICE(p->static_prio));
+	return __normal_prio(p->policy, p->rt_priority, p->static_prio);
 }
 
 /*
@@ -5079,7 +5079,7 @@ static int __sched_setscheduler(struct task_struct *p,
 
 	p->sched_reset_on_fork = reset_on_fork;
 
-	newprio = __normal_prio(policy, attr->sched_priority, attr->sched_nice);
+	newprio = __normal_prio(policy, attr->sched_priority, NICE_TO_PRIO(attr->sched_nice));
 	if (pi) {
 		/*
 		 * Take priority boosted tasks into account. If the new
-- 
2.37.0.rc0.15.g3b9a5a33c2

