From d42c21cac754a08350aa2a735357a417d1263f0f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 4 Apr 2022 12:03:13 +0200
Subject: [PATCH 266/288] prjc: remove duplicated struct __call_single_nod

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 include/linux/sched.h | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 1d9da0d79eb3..7a4bf3a0db5a 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -758,8 +758,9 @@ struct task_struct {
 #if defined(CONFIG_SMP) || defined(CONFIG_SCHED_ALT)
 	int				on_cpu;
 #endif
-#if defined(CONFIG_SMP) && !defined(CONFIG_SCHED_ALT)
-	struct __call_single_node	wake_entry;
+
+#ifdef CONFIG_SMP
+#ifndef CONFIG_SCHED_ALT
 	unsigned int			wakee_flips;
 	unsigned long			wakee_flip_decay_ts;
 	struct task_struct		*last_wakee;
@@ -773,6 +774,7 @@ struct task_struct {
 	 */
 	int				recent_used_cpu;
 	int				wake_cpu;
+#endif /* !CONFIG_SCHED_ALT */
 #endif
 	int				on_rq;
 
-- 
2.37.0.rc0.15.g3b9a5a33c2

