From f0976d7b844d2e93ea2bc9586b16a970bfbeb6fc Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 30 Aug 2022 23:24:41 +0200
Subject: [PATCH] arch/x86: improve KBUILD_CFLAG

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile          | 3 +++
 arch/x86/Makefile | 3 ++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 952d35406..931279ea2 100644
--- a/Makefile
+++ b/Makefile
@@ -987,6 +987,9 @@ ifdef CONFIG_CC_IS_GCC
 KBUILD_CFLAGS += $(call cc-ifversion, -ge, 0901, -Wno-alloc-size-larger-than)
 endif
 
+# disable GCC vectorization on trees
+KBUILD_CFLAGS	+= $(call cc-option, -fno-tree-vectorize)
+
 # disable invalid "can't wrap" optimizations for signed / pointers
 KBUILD_CFLAGS	+= -fno-strict-overflow
 
diff --git a/arch/x86/Makefile b/arch/x86/Makefile
index bafbd905e..aaee018a8 100644
--- a/arch/x86/Makefile
+++ b/arch/x86/Makefile
@@ -67,7 +67,8 @@ export BITS
 #
 #    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=53383
 #
-KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx
+KBUILD_CFLAGS += -mno-sse -mno-mmx -mno-sse2 -mno-3dnow -mno-avx -mno-avx2 \
+		   -mno-avx512f -O3
 
 ifeq ($(CONFIG_X86_KERNEL_IBT),y)
 #
-- 
2.37.1.377.g679aad9e82

