From 7430a9e9a78e574c3d61d4f1631f1dfb5f37c2a5 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 4 Sep 2019 05:27:03 +0800
Subject: [PATCH 07/81] bmq: [Sync] ucamp

BMQ doesn't support ucamp.
---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index 36d4c429a41f..d35e21682fe8 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -721,6 +721,7 @@ menu "Scheduler features"
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
+	depends on !SCHED_BMQ
 	help
 	  This feature enables the scheduler to track the clamped utilization
 	  of each CPU based on RUNNABLE tasks scheduled on that CPU.
-- 
2.26.0.51.ga7d14a4428

