From b10dd6ae5d46592be198179ce330e8220f72e5df Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 24 Dec 2019 14:43:10 +0800
Subject: [PATCH 56/81] bmq: Change SCHED_BMQ Kconfig location.

---
 init/Kconfig | 37 +++++++++++++++++--------------------
 1 file changed, 17 insertions(+), 20 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index bbecaf7450f1..338e5dd9f2f7 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -74,26 +74,6 @@ config THREAD_INFO_IN_TASK
 
 menu "General setup"
 
-config SCHED_BMQ
-	bool "BMQ CPU scheduler"
-	help
-	  The BitMap Queue CPU scheduler for excellent interactivity and
-	  responsiveness on the desktop and solid scalability on normal
-	  hardware and commodity servers.
-
-	  Currently incompatible with the Group CPU scheduler, and RCU TORTURE
-          TEST so these options are disabled.
-
-          Say Y here.
-	default y
-
-config SCHED_TIMESLICE
-	int "Scheduler Task time slice"
-	depends on SCHED_BMQ
-	help
-	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
-	default 4
-
 config BROKEN
 	bool
 
@@ -718,6 +698,23 @@ config GENERIC_SCHED_CLOCK
 
 menu "Scheduler features"
 
+config SCHED_BMQ
+	bool "BMQ CPU scheduler"
+	help
+	  The BitMap Queue CPU scheduler for excellent interactivity and
+	  responsiveness on the desktop and solid scalability on normal
+	  hardware and commodity servers.
+
+          Say Y here.
+	default y
+
+config SCHED_TIMESLICE
+	int "Scheduler Task time slice"
+	depends on SCHED_BMQ
+	help
+	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
+	default 4
+
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
-- 
2.26.0.51.ga7d14a4428

