From 8ce96e2e1fe18154973355ca1a8bd1d3732e861a Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 22 Jan 2020 16:01:48 +0800
Subject: [PATCH 64/81] bmq: Rework sched_init_topology_cpumask().

---
 kernel/sched/bmq.c | 31 +++++++++++--------------------
 1 file changed, 11 insertions(+), 20 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index d60bd88bec7c..189865fa0597 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -5605,42 +5605,33 @@ static void sched_init_topology_cpumask(void)
 	for_each_online_cpu(cpu) {
 		chk = &(per_cpu(sched_cpu_affinity_masks, cpu)[0]);
 
+		cpumask_complement(chk, cpumask_of(cpu));
 #ifdef CONFIG_SCHED_SMT
-		cpumask_setall(chk);
-		cpumask_clear_cpu(cpu, chk);
-		if (cpumask_and(chk, chk, topology_sibling_cpumask(cpu))) {
-			printk(KERN_INFO "bmq: cpu #%d affinity check mask - smt 0x%08lx",
+		if (cpumask_and(chk, chk, topology_sibling_cpumask(cpu)))
+			printk(KERN_INFO "bmq: cpu #%d affinity mask - smt 0x%08lx",
 			       cpu, (chk++)->bits[0]);
-		}
 		cpumask_complement(chk, topology_sibling_cpumask(cpu));
-#else
-		cpumask_clear_cpu(cpu, chk);
 #endif
+		/* Set up sd_llc_id per CPU */
+		per_cpu(sd_llc_id, cpu) =
 #ifdef CONFIG_SCHED_MC
+			cpumask_first(cpu_coregroup_mask(cpu));
+
 		if (cpumask_and(chk, chk, cpu_coregroup_mask(cpu)))
-			printk(KERN_INFO "bmq: cpu #%d affinity check mask - coregroup 0x%08lx",
+			printk(KERN_INFO "bmq: cpu #%d affinity mask - coregroup 0x%08lx",
 			       cpu, (chk++)->bits[0]);
 		cpumask_complement(chk, cpu_coregroup_mask(cpu));
-
-		/**
-		 * Set up sd_llc_id per CPU
-		 */
-		per_cpu(sd_llc_id, cpu) =
-			cpumask_first(cpu_coregroup_mask(cpu));
 #else
-		per_cpu(sd_llc_id, cpu) =
 			cpumask_first(topology_core_cpumask(cpu));
+#endif
 
-		cpumask_setall(chk);
-		cpumask_clear_cpu(cpu, chk);
-#endif /* NOT CONFIG_SCHED_MC */
 		if (cpumask_and(chk, chk, topology_core_cpumask(cpu)))
-			printk(KERN_INFO "bmq: cpu #%d affinity check mask - core 0x%08lx",
+			printk(KERN_INFO "bmq: cpu #%d affinity mask - core 0x%08lx",
 			       cpu, (chk++)->bits[0]);
 		cpumask_complement(chk, topology_core_cpumask(cpu));
 
 		if (cpumask_and(chk, chk, cpu_online_mask))
-			printk(KERN_INFO "bmq: cpu #%d affinity check mask - others 0x%08lx",
+			printk(KERN_INFO "bmq: cpu #%d affinity mask - others 0x%08lx",
 			       cpu, (chk++)->bits[0]);
 
 		per_cpu(sched_cpu_affinity_end_mask, cpu) = chk;
-- 
2.26.0.51.ga7d14a4428

