From 08d6513ceed5dba954ed55f2eff96b7b63fa3371 Mon Sep 17 00:00:00 2001
From: Hamad Al Marri <hamad.s.almarri@gmail.com>
Date: Mon, 13 Dec 2021 13:12:37 +0300
Subject: [PATCH 25/29] push all even if having 1 task

---
 kernel/sched/bs_nohz.h | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/kernel/sched/bs_nohz.h b/kernel/sched/bs_nohz.h
index 017e29278..7dee1be32 100644
--- a/kernel/sched/bs_nohz.h
+++ b/kernel/sched/bs_nohz.h
@@ -706,10 +706,6 @@ void push_to_grq(struct rq *rq)
 	if (!cfs_rq->head)
 		return;
 
-	/* no need to push a single task and take it again */
-	if (cfs_rq->h_nr_running == 1 && !grq->cfs.head)
-		return;
-
 	rq_lock_irqsave(rq, &rf);
 	update_rq_clock(rq);
 
-- 
2.34.1.75.gabe6bb3905

