From b39ac639831816b7beec3ff8fd90b180d4e4bbbe Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 8 Aug 2022 22:15:07 +0200
Subject: [PATCH 15/15] Add Hz min_ttl_ms for thrashing prevention

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 mm/vmscan.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 7096ff783..8d510f9d1 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -4434,7 +4434,7 @@ static bool age_lruvec(struct lruvec *lruvec, struct scan_control *sc, unsigned
 }
 
 /* to protect the working set of the last N jiffies */
-static unsigned long lru_gen_min_ttl __read_mostly;
+static unsigned long lru_gen_min_ttl __read_mostly = HZ;
 
 static void lru_gen_age_node(struct pglist_data *pgdat, struct scan_control *sc)
 {
-- 
2.37.1.377.g679aad9e82

