From 8c494e528352ae2e369a73ea695599993ddedc5b Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 9 Mar 2019 03:19:01 +0100
Subject: [PATCH] bfq-sq, bfq-mq: Avoid double definition of bfq_pid_to_str

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-mq.h | 2 +-
 block/bfq.h    | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/block/bfq-mq.h b/block/bfq-mq.h
index f158b1dca92c..b2e1615218ab 100644
--- a/block/bfq-mq.h
+++ b/block/bfq-mq.h
@@ -779,7 +779,7 @@ BFQ_BFQQ_FNS(softrt_update);
 #undef BFQ_BFQQ_FNS
 
 /* Logging facilities. */
-void bfq_pid_to_str(int pid, char *str, int len)
+static inline void bfq_pid_to_str(int pid, char *str, int len)
 {
 	if (pid != -1)
 		snprintf(str, len, "%d", pid);
diff --git a/block/bfq.h b/block/bfq.h
index dcf783deb540..0c7c669c7f50 100644
--- a/block/bfq.h
+++ b/block/bfq.h
@@ -762,7 +762,7 @@ BFQ_BFQQ_FNS(softrt_update);
 #undef BFQ_BFQQ_FNS
 
 /* Logging facilities. */
-void bfq_pid_to_str(int pid, char *str, int len)
+static inline void bfq_pid_to_str(int pid, char *str, int len)
 {
 	if (pid != -1)
 		snprintf(str, len, "%d", pid);
-- 
2.21.0.4.g36eb1cb9cf

