From edaa9bc93eba192c7ff6cb497b15f9a13e5fb715 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Tue, 24 Dec 2019 05:31:17 +0100
Subject: [PATCH 20/20] Arch Linux kernel v5.4.6-arch3

---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 20ec7c20279e..bca093aebe6f 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 4
 SUBLEVEL = 6
-EXTRAVERSION =
+EXTRAVERSION = -arch3
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.24.1.591.g12029dc57d

