From d28dddb5642fa1fb877e639985cf21bd2165ca99 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 14 Jan 2021 10:54:35 +0100
Subject: [PATCH 20/35] block, bfq, DEBUG: check tot_busy and log asymmetric
 scenario

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 2db4056bd..aac6a73f7 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -4035,6 +4035,8 @@ static bool idling_needed_for_service_guarantees(struct bfq_data *bfqd,
 	int tot_busy_queues = bfq_tot_busy_queues(bfqd);
 	bool asymmetric_scenario;
 
+	BFQ_BUG_ON(tot_busy_queues < 1);
+
 	/* No point in idling for bfqq if it won't get requests any longer */
 	if (unlikely(!bfqq_process_refs(bfqq)))
 		return false;
@@ -4051,7 +4053,7 @@ static bool idling_needed_for_service_guarantees(struct bfq_data *bfqd,
 		     "wr_coeff %d wr_busy %d busy %d asymmetric %d",
 		     bfqq->wr_coeff,
 		     bfqd->wr_busy_queues,
-		     bfq_tot_busy_queues(bfqd),
+		     tot_busy_queues,
 		     asymmetric_scenario);
 
 	return asymmetric_scenario;
-- 
2.31.1.362.g311531c9de

