From 933e6cf2e9df283617e8531ef4f8e890b49afdfc Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Fri, 8 May 2020 09:47:20 +0200
Subject: [PATCH 27/35] block, bfq, DEBUG: add detailed log on denied dispatch
 for I/O plugging

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 470d30cf0..9385ccb79 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5270,6 +5270,8 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 					bfq_get_first_task_pid(new_bfqq));
 			bfqq = new_bfqq;
 		} else {
+			bfq_log_bfqq(bfqd, bfqq, "idling with %d busy queues",
+				     bfq_tot_busy_queues(bfqd));
 			bfqq = NULL;
 		}
 		goto keep_queue;
-- 
2.31.1.362.g311531c9de

