From 5726c515a7013f2cf759c23ae2001dfa829201d4 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Sat, 1 May 2021 12:24:53 +0200
Subject: [PATCH 33/35] block, bfq, DEBUG: add BUG_ON on circular stable merge

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index a7b5e4ffe..979b413f5 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -3059,10 +3059,12 @@ bfq_setup_cooperator(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 				if (!new_bfqq)
 					bfq_log_bfqq(bfqd, bfqq,
 					     "merging already occurred");
-				else
+				else {
+					BFQ_BUG_ON(stable_merge_bfqq == bfqq);
 					bfq_log_bfqq(bfqd, bfqq,
 						     "late stable merging with %p",
 						     new_bfqq);
+				}
 
 				bic->stably_merged = true;
 				if (new_bfqq && new_bfqq->bic)
-- 
2.31.1.362.g311531c9de

