From 139dd4c3b0d33037fc3537c6dff26ff59d37b947 Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 14 Nov 2022 17:48:24 +0100
Subject: [PATCH 47/47] gcc-lto: pie-fix

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 Makefile               | 2 +-
 include/linux/export.h | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/Makefile b/Makefile
index 5059181a8..50899754a 100644
--- a/Makefile
+++ b/Makefile
@@ -504,7 +504,7 @@ OBJDUMP		= $(LLVM_PREFIX)llvm-objdump$(LLVM_SUFFIX)
 READELF		= $(LLVM_PREFIX)llvm-readelf$(LLVM_SUFFIX)
 STRIP		= $(LLVM_PREFIX)llvm-strip$(LLVM_SUFFIX)
 else
-CC		= $(CROSS_COMPILE)gcc
+CC		= $(CROSS_COMPILE)gcc -fno-pie -no-pie
 LD		= $(CROSS_COMPILE)ld
 AR		= $(CROSS_COMPILE)ar
 NM		= $(CROSS_COMPILE)nm
diff --git a/include/linux/export.h b/include/linux/export.h
index 3cb5f8532..0dbac53e4 100644
--- a/include/linux/export.h
+++ b/include/linux/export.h
@@ -120,7 +120,7 @@ struct kernel_symbol {
  * discarded in the final link stage.
  */
 #define __ksym_marker(sym)	\
-	static int __ksym_marker_##sym[0] __section(".discard.ksym") __used
+	int __ksym_marker_##sym[0] __section(".discard.ksym") __used
 
 #define __EXPORT_SYMBOL(sym, sec, ns)					\
 	__ksym_marker(sym);						\
-- 
2.39.0.rc2.1.gbd5df96b79

