From 501dc6623e23376d69a25bc20db42c8b215d2dce Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@redhat.com>
Date: Wed, 17 Jun 2020 10:07:18 +0200
Subject: [PATCH 12/21] mm-5.7: fix proactive compaction build with THP
 disabled

Signed-off-by: Oleksandr Natalenko <oleksandr@redhat.com>
---
 mm/compaction.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/compaction.c b/mm/compaction.c
index b684af7..93e703c 100644
--- a/mm/compaction.c
+++ b/mm/compaction.c
@@ -60,7 +60,7 @@ static const int HPAGE_FRAG_CHECK_INTERVAL_MSEC = 500;
  * calculates external fragmentation, which is used as
  * the "fragmentation score" of a node/zone.
  */
-#if defined HPAGE_PMD_ORDER
+#if defined CONFIG_TRANSPARENT_HUGEPAGE
 #define COMPACTION_HPAGE_ORDER	HPAGE_PMD_ORDER
 #elif defined HUGETLB_PAGE_ORDER
 #define COMPACTION_HPAGE_ORDER	HUGETLB_PAGE_ORDER
-- 
2.28.0.rc0

