From f4636fb1d858b7fa1c17018c93f6646ba7eac831 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 26 Mar 2021 13:46:22 +0800
Subject: [PATCH 133/148] sched/alt: Remove unnecessary CONFIG_SMP macros
 usage.

---
 kernel/sched/alt_core.c | 6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 02610d086d00..ae37764ced1e 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -3816,15 +3816,13 @@ static inline int take_other_rq_tasks(struct rq *rq, int cpu)
 
 			if ((nr_migrated = migrate_pending_tasks(src_rq, rq, cpu))) {
 				src_rq->nr_running -= nr_migrated;
-#ifdef CONFIG_SMP
 				if (src_rq->nr_running < 2)
 					cpumask_clear_cpu(i, &sched_rq_pending_mask);
-#endif
+
 				rq->nr_running += nr_migrated;
-#ifdef CONFIG_SMP
 				if (rq->nr_running > 1)
 					cpumask_set_cpu(cpu, &sched_rq_pending_mask);
-#endif
+
 				update_sched_rq_watermark(rq);
 				cpufreq_update_util(rq, 0);
 
-- 
2.32.0.93.g670b81a890

