From 38bd5427e33cb711356589c3de380db06f288377 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 14 Jan 2021 11:54:08 +0100
Subject: [PATCH 19/55] block, bfq, DEBUG: log I/O-intensity parameters

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index d19d7bb96..fd48fec86 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -2058,6 +2058,10 @@ static void bfq_update_io_intensity(struct bfq_queue *bfqq, u64 now_ns)
 	else
 		bfq_mark_bfqq_IO_bound(bfqq);
 
+	bfq_log_bfqq(bfqq->bfqd, bfqq, "tot %llu idle %llu bound %d",
+		     tot_io_time, bfqq->tot_idle_time,
+		     bfq_bfqq_IO_bound(bfqq));
+
 	/*
 	 * Keep an observation window of at most 200 ms in the past
 	 * from now.
-- 
2.30.0.155.g66e871b664

