From 77b44605d85ff8cc698cc2a4b524043b7b26c1dc Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 4 Sep 2018 15:51:37 +0800
Subject: [PATCH 07/15] Tag PDS 0.98y

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 07f750efee06..c95ce51215b4 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -90,7 +90,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98x by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98y by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.18.0.748.gfa03cdc39

