From 9783640eb5aab20ff0643368ab030115c2ea5099 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 5 Sep 2018 11:00:31 +0800
Subject: [PATCH 09/13] pds: Don't balance on an idle task.

---
 kernel/sched/pds.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 2db97cdc2e38..c57ffc2dc39f 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2956,7 +2956,8 @@ static __latent_entropy void pds_run_rebalance(struct softirq_action *h)
 
 	raw_spin_lock_irqsave(&this_rq->lock, flags);
 	curr = this_rq->curr;
-	if (cpumask_and(&tmp, &curr->cpus_allowed, &sched_cpu_sg_idle_mask)) {
+	if (!is_idle_task(curr) &&
+	    cpumask_and(&tmp, &curr->cpus_allowed, &sched_cpu_sg_idle_mask)) {
 		int active_balance = 0;
 
 		if (likely(!this_rq->active_balance)) {
-- 
2.19.0

