From e598a67932b1e979df9c79d784f6b9f57e4de04a Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 11 Sep 2018 20:04:35 +0800
Subject: [PATCH 10/12] pds: Improve idle task SMT_NICE handling in ttwu.

---
 kernel/sched/pds.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index c57ffc2dc39f..4bad95c112e2 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -1603,17 +1603,19 @@ task_preemptible_rq_idle(struct task_struct *p, cpumask_t *chk_mask)
 	if (cpumask_and(&tmp, chk_mask, &sched_cpu_sg_idle_mask))
 		return best_mask_cpu(task_cpu(p), &tmp);
 #endif
-	if (cpumask_and(&tmp, chk_mask, &sched_rq_queued_masks[SCHED_RQ_EMPTY])) {
+
 #ifdef CONFIG_SMT_NICE
-		/* Only ttwu on cpu which is not smt supressed */
+	/* Only ttwu on cpu which is not smt supressed */
+	if (cpumask_andnot(&tmp, chk_mask, &sched_smt_supressed_mask)) {
 		cpumask_t t;
-		if (cpumask_andnot(&t, &tmp,  &sched_smt_supressed_mask))
-			best_mask_cpu(task_cpu(p), &t);
-#else
+		if (cpumask_and(&t, &tmp, &sched_rq_queued_masks[SCHED_RQ_EMPTY]))
+			return best_mask_cpu(task_cpu(p), &t);
 		return best_mask_cpu(task_cpu(p), &tmp);
-#endif /* !CONFIG_SMT_NICE */
 	}
+#endif
 
+	if (cpumask_and(&tmp, chk_mask, &sched_rq_queued_masks[SCHED_RQ_EMPTY]))
+		return best_mask_cpu(task_cpu(p), &tmp);
 	return best_mask_cpu(task_cpu(p), chk_mask);
 }
 
-- 
2.19.0

