From e8a54f05bcce178f4b4a713755edf40d0bd4097e Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 17 Sep 2018 16:01:10 +0800
Subject: [PATCH 12/13] Tag PDS 0.98z

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 7bf296f4f111..f15d1a7c5d81 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -90,7 +90,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98y by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98z by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.19.0

