From ee54f8e1539875b9d2b3fef0916cd3a67f96143f Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 26 Sep 2018 13:58:30 +0800
Subject: [PATCH 15/15] Tag PDS 0.99a

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 88202ed42123..1872b1f9f8d4 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -88,7 +88,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.98z by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99a by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.19.0.221.g150f307af

