From 4020af2c92f9f588e07684654608f4a65ba4a31e Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 19 Jan 2020 09:02:51 +0800
Subject: [PATCH 063/112] Tag BMQ v5.5-r0

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 8cebdc571f7b..d60bd88bec7c 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -69,7 +69,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r2 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.5-r0 by Alfred Chen.\n");
 }
 
 /**
-- 
2.27.0.112.g101b3204f3

