From 9f15e43fe1a7a38d43c02ef0ec785de86f1cea3b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 30 Mar 2020 10:30:42 +0800
Subject: [PATCH 082/112] bmq: Fix compile warning on 32bit system.

---
 kernel/sched/bmq.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 58657044d58c..e4c1561e00e6 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -5554,10 +5554,10 @@ static void sched_init_topology_cpumask(void)
 			       cpu, (chk++)->bits[0]);
 
 		per_cpu(sched_cpu_affinity_end_mask, cpu) = chk;
-		printk(KERN_INFO "bmq: cpu#%d llc_id = %d, llc_mask idx = %ld\n",
+		printk(KERN_INFO "bmq: cpu#%d llc_id = %d, llc_mask idx = %d\n",
 		       cpu, per_cpu(sd_llc_id, cpu),
-		       per_cpu(sched_cpu_llc_mask, cpu) -
-		       &(per_cpu(sched_cpu_affinity_masks, cpu)[0]));
+		       (int) (per_cpu(sched_cpu_llc_mask, cpu) -
+			      &(per_cpu(sched_cpu_affinity_masks, cpu)[0])));
 	}
 }
 #endif
-- 
2.27.0.112.g101b3204f3

