From d6b06173bd08c4adcc9a4f34ccb2220c994a155c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 14 May 2020 17:50:49 +0800
Subject: [PATCH 100/112] bmq: [Sync] b3212fe2bc06 sched/swait: Prepare usage
 in completions

---
 kernel/sched/bmq_sched.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/sched/bmq_sched.h b/kernel/sched/bmq_sched.h
index fca42b270620..1449c4da7088 100644
--- a/kernel/sched/bmq_sched.h
+++ b/kernel/sched/bmq_sched.h
@@ -507,4 +507,8 @@ static inline int sched_numa_find_closest(const struct cpumask *cpus, int cpu)
 	return nr_cpu_ids;
 }
 #endif
+
+void swake_up_all_locked(struct swait_queue_head *q);
+void __prepare_to_swait(struct swait_queue_head *q, struct swait_queue *wait);
+
 #endif /* BMQ_SCHED_H */
-- 
2.27.0.112.g101b3204f3

