From 36867777c4d512f7e08f99f2b18039ee39b816f5 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 3 Jun 2020 06:14:55 +0800
Subject: [PATCH 107/112] bmq: Fix compile error in psi.c.

---
 kernel/sched/bmq_sched.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/kernel/sched/bmq_sched.h b/kernel/sched/bmq_sched.h
index 59b3c43c7d9f..c0fae307869f 100644
--- a/kernel/sched/bmq_sched.h
+++ b/kernel/sched/bmq_sched.h
@@ -329,6 +329,11 @@ this_rq_lock_irq(struct rq_flags *rf)
 	return rq;
 }
 
+static inline int task_current(struct rq *rq, struct task_struct *p)
+{
+	return rq->curr == p;
+}
+
 static inline bool task_running(struct task_struct *p)
 {
 	return p->on_cpu;
-- 
2.27.0.112.g101b3204f3

