From bf3b60c8edbaf1a75bcff4793ea01b95e53b5b37 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 6 Mar 2019 10:25:40 +0800
Subject: [PATCH] pds: update_task_priodl() for all task when timeslice
 expired.

---
 kernel/sched/pds.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 20c5749e3bb6..7e6c0cbed07b 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -458,11 +458,10 @@ static inline u64 static_deadline_diff(int static_prio)
 static inline void time_slice_expired(struct task_struct *p, struct rq *rq)
 {
 	p->time_slice = timeslice();
-
-	if (p->prio >= NORMAL_PRIO) {
+	if (p->prio >= NORMAL_PRIO)
 		p->deadline = rq->clock + task_deadline_diff(p);
-		update_task_priodl(p);
-	}
+
+	update_task_priodl(p);
 }
 
 static inline struct task_struct *rq_first_queued_task(struct rq *rq)
-- 
2.21.0

