From 26ed57e2b346bf9e914943c63e9a41822e6990e6 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 11 Dec 2023 14:12:02 -0500
Subject: [PATCH 74/74] XXX what do about thread_info, asm/current

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 include/linux/sched.h | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index 34af5333d..c3c08dcf3 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -9,8 +9,10 @@
 
 #include <uapi/linux/sched.h>
 
-#include <asm/current.h>
+//#include <asm/current.h>
+//#include <asm/thread_info.h>
 
+#include <linux/cache.h>
 #include <linux/pid_types.h>
 #include <linux/sem_types.h>
 #include <linux/shm.h>
@@ -18,9 +20,10 @@
 #include <linux/mutex_types.h>
 #include <linux/plist_types.h>
 #include <linux/hrtimer_types.h>
+#include <linux/timer_types.h>
 #include <linux/seccomp_types.h>
 #include <linux/nodemask_types.h>
-#include <linux/rcupdate.h>
+//#include <linux/rcupdate.h>
 #include <linux/refcount_types.h>
 #include <linux/resource.h>
 #include <linux/latencytop.h>
@@ -58,6 +61,7 @@ struct nsproxy;
 struct perf_event_context;
 struct pid_namespace;
 struct pipe_inode_info;
+struct pt_regs;
 struct rcu_node;
 struct reclaim_state;
 struct robust_list_head;
-- 
2.43.0.rc2

