From 5148e4e563e17507b3c81710a76329549ed3ca05 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 4 Mar 2019 14:35:18 +0800
Subject: [PATCH 02/64] bmq: Deboost task for sched_yield_type == 1.

---
 kernel/sched/bmq.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 39a7cdfc9e0f..b6465a8f1088 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -84,8 +84,7 @@ static inline void print_scheduler_version(void)
 /**
  * sched_yield_type - Choose what sort of yield sched_yield will perform.
  * 0: No yield.
- * 1: Yield only to better priority/deadline tasks. (default)
- * 2: Expire timeslice and recalculate deadline.
+ * 1: Deboost and requeue task. (default)
  */
 int sched_yield_type __read_mostly = 1;
 
@@ -4630,8 +4629,8 @@ static void do_sched_yield(void)
 
 	rq = this_rq_lock_irq(&rf);
 
-	if (sched_yield_type > 1) {
-		current->time_slice = SCHED_TIMESLICE_NS;
+	if (rt_task(current)) {
+		current->boost_prio = MAX_PRIORITY_ADJ;
 		requeue_task(current, rq);
 	}
 	schedstat_inc(rq->yld_count);
-- 
2.22.0.214.g8dca754b1e

