From 6bff8a46882cf5051f9bf1f6c3b75a227225707b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 11 Mar 2019 14:08:36 +0800
Subject: [PATCH 12/64] Tag  BMQ v0.89

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 0d3a73dabb45..b0e233a73542 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.88 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.89 by Alfred Chen.\n");
 }
 
 /**
-- 
2.22.0.214.g8dca754b1e

