From df3905afa63140a722be97418825b302e9b1586f Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 10 Apr 2019 14:08:41 +0800
Subject: [PATCH 29/64] Tag  BMQ v0.92

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index de3ea4926f3f..defd7350ace0 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.91 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.92 by Alfred Chen.\n");
 }
 
 /**
-- 
2.22.0.214.g8dca754b1e

