From db409b2fdc020c918eb011c1a23d09a96faed066 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 20 Apr 2019 16:25:44 +0800
Subject: [PATCH 41/64] bmq: [Sync] 5e1f0f098b46 mm, compaction: capture a page
 under direct compaction

---
 kernel/sched/bmq.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 5bc8d7a8f920..476e05970b12 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -1845,6 +1845,10 @@ int sched_fork(unsigned long __maybe_unused clone_flags, struct task_struct *p)
 	p->stime =
 	p->sched_time = 0;
 
+#ifdef CONFIG_COMPACTION
+	p->capture_control = NULL;
+#endif
+
 	/*
 	 * We mark the process as NEW here. This guarantees that
 	 * nobody will actually run it, and a signal or other external
-- 
2.22.0.214.g8dca754b1e

