From 854a2ddc5d18566bdcc83e4c0f3f9a1efa157ffa Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 5 May 2019 11:16:44 +0800
Subject: [PATCH 45/64] bmq: Fix rq missing on arm arch.

---
 kernel/sched/bmq.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index da767e230019..87275e62b466 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -177,8 +177,13 @@ static inline void update_sched_rq_watermark(struct rq *rq)
 		return;
 
 	cpu = cpu_of(rq);
+#ifdef CONFIG_X86
 	if (!cpumask_andnot(&sched_rq_watermark[last_wm],
 			    &sched_rq_watermark[last_wm], cpumask_of(cpu)))
+#else
+	cpumask_clear_cpu(cpu, &sched_rq_watermark[last_wm]);
+	if (cpumask_empty(&sched_rq_watermark[last_wm]))
+#endif
 		clear_bit(last_wm, sched_rq_watermark_bitmap);
 	cpumask_set_cpu(cpu, &sched_rq_watermark[watermark]);
 	set_bit(watermark, sched_rq_watermark_bitmap);
-- 
2.22.0.214.g8dca754b1e

