From 2370cf9d1518b318eee4ef95dfbb0b2da895a114 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 21 Jun 2019 13:48:25 +0800
Subject: [PATCH 60/64] bmq: [Sync] 1b174a2cb67a sched/core: Remove
 ttwu_activate()

---
 kernel/sched/bmq.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index c48970c5c331..1b5361a27be5 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -1439,11 +1439,6 @@ ttwu_stat(struct task_struct *p, int cpu, int wake_flags)
 	__schedstat_inc(rq->ttwu_count);
 }
 
-static inline void ttwu_activate(struct task_struct *p, struct rq *rq)
-{
-	activate_task(p, rq);
-}
-
 /*
  * Mark the task runnable and perform wakeup-preemption.
  */
@@ -1462,7 +1457,7 @@ ttwu_do_activate(struct rq *rq, struct task_struct *p, int wake_flags)
 		rq->nr_uninterruptible--;
 #endif
 
-	ttwu_activate(p, rq);
+	activate_task(p, rq);
 	ttwu_do_wakeup(rq, p, 0);
 }
 
-- 
2.22.0.214.g8dca754b1e

