From 8bf4d9a6f634f5b444241cece29bc224573b9077 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 21 Jun 2019 14:05:27 +0800
Subject: [PATCH 61/64] bmq: [Sync] bee9853932e9 sched/core: Fix typo in
 comment

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 1b5361a27be5..aaf093c3fb90 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -1005,7 +1005,7 @@ static inline bool is_per_cpu_kthread(struct task_struct *p)
 }
 
 /*
- * Per-CPU kthreads are allowed to run on !actie && online CPUs, see
+ * Per-CPU kthreads are allowed to run on !active && online CPUs, see
  * __set_cpus_allowed_ptr() and select_fallback_rq().
  */
 static inline bool is_cpu_allowed(struct task_struct *p, int cpu)
-- 
2.22.0.214.g8dca754b1e

