From 43d3b94848af177f262c772112bc1665889485c0 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 24 Jun 2019 23:44:39 +0800
Subject: [PATCH 63/64] bmq: Fix compilation error when PSI disabled and CGROUP
 enabled.

---
 kernel/sched/bmq_sched.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/bmq_sched.h b/kernel/sched/bmq_sched.h
index 3dbe34797530..d9ea1288f467 100644
--- a/kernel/sched/bmq_sched.h
+++ b/kernel/sched/bmq_sched.h
@@ -21,6 +21,7 @@
 
 #include <uapi/linux/sched/types.h>
 
+#include <linux/cgroup.h>
 #include <linux/cpufreq.h>
 #include <linux/cpuidle.h>
 #include <linux/cpuset.h>
-- 
2.22.0.214.g8dca754b1e

