From 54506e345a9c97989971aac0860cad86e5efec53 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 2 Jul 2019 22:42:31 +0800
Subject: [PATCH 64/64] Tag BMQ v0.97

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index aaf093c3fb90..269fcb1f653a 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.96 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.97 by Alfred Chen.\n");
 }
 
 /**
-- 
2.22.0.214.g8dca754b1e

