From d69d055ed8f45f5f2a4e0e4ddef177f53f886597 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 29 Dec 2020 01:23:01 +0100
Subject: [PATCH 25/34] bfq-dev: update version to v14

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index bff060db4..b9ea55666 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7684,7 +7684,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ-dev I/O-scheduler: v13";
+	char msg[60] = "BFQ-dev I/O-scheduler: v14";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.30.1.457.gf011795891

