From 080f13e6c0ccf58a54f98890218c5586691df278 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 22 Jan 2021 14:47:52 +0100
Subject: [PATCH 34/34] port: bfq: don't duplicate code for different paths

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-wf2q.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/block/bfq-wf2q.c b/block/bfq-wf2q.c
index 1076bb98e..2ce01c5ea 100644
--- a/block/bfq-wf2q.c
+++ b/block/bfq-wf2q.c
@@ -154,8 +154,7 @@ static bool bfq_update_next_in_service(struct bfq_sched_data *sd,
 
 	sd->next_in_service = next_in_service;
 
-	if (!next_in_service)
-		return parent_sched_may_change;
+	return parent_sched_may_change;
 
 	bfqq = bfq_entity_to_bfqq(next_in_service);
 	if (bfqq)
@@ -171,7 +170,6 @@ static bool bfq_update_next_in_service(struct bfq_sched_data *sd,
 			     "chosen this entity");
 	}
 #endif
-	return parent_sched_may_change;
 }
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
-- 
2.30.1.457.gf011795891

