From e6de91a5ea4e2bac5921aa89c561443298438c4d Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 10 Aug 2026 10:17:32 +0700
Subject: [PATCH 51/51] Fix audio t2bce

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/staging/t2bce/t2bce_audio/audio.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/t2bce/t2bce_audio/audio.c b/drivers/staging/t2bce/t2bce_audio/audio.c
index 60f552c06..98c0fef39 100644
--- a/drivers/staging/t2bce/t2bce_audio/audio.c
+++ b/drivers/staging/t2bce/t2bce_audio/audio.c
@@ -435,6 +435,7 @@ static void t2audio_init_dev(struct t2audio_device *a, t2audio_device_id_t dev_i
         dev_err(a->dev, "Failed to get device uid for device %llx\n", dev_id);
         goto fail;
     }
+    uid[uid_len] = '\0';
     pr_debug("t2bce_audio: Remote device %llx %.*s\n", dev_id, (int) uid_len, uid);
 
     sdev->a = a;
-- 
2.55.0

