From 3cd529db7c7df646c8f280bbfb94bd8026a664de Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Tue, 12 Mar 2019 18:19:35 +0100
Subject: [PATCH] fixes-5.0: merge BFQ improvements, v3

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 block/bfq-iosched.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/block/bfq-iosched.h b/block/bfq-iosched.h
index 1e34cce59ba7..60c148728cc5 100644
--- a/block/bfq-iosched.h
+++ b/block/bfq-iosched.h
@@ -1027,7 +1027,7 @@ void bfq_add_bfqq_busy(struct bfq_data *bfqd, struct bfq_queue *bfqq);
 /* --------------- end of interface of B-WF2Q+ ---------------- */
 
 /* Logging facilities. */
-static void bfq_pid_to_str(int pid, char *str, int len)
+static inline void bfq_pid_to_str(int pid, char *str, int len)
 {
 	if (pid != -1)
 		snprintf(str, len, "%d", pid);
@@ -1059,7 +1059,7 @@ struct bfq_group *bfqq_group(struct bfq_queue *bfqq);
 	bfq_pid_to_str((bfqq)->pid, pid_str, MAX_PID_STR_LENGTH);	\
 	blk_add_trace_msg((bfqd)->queue, "bfq%s%c " fmt, pid_str,	\
 			bfq_bfqq_sync((bfqq)) ? 'S' : 'A',		\
-				##args)		\
+				##args);	\
 } while (0)
 #define bfq_log_bfqg(bfqd, bfqg, fmt, args...)		do {} while (0)
 
-- 
2.21.0.155.ge902e9bcae

