From 42e29d29d23c076d2e7b82578c4eb9ca17a0ccd6 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 10 Nov 2018 11:21:49 +0100
Subject: [PATCH 7/7] LL: Add .ll version

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index 08510230b..d80c1c1df 100644
--- a/Makefile
+++ b/Makefile
@@ -14,6 +14,10 @@ NAME = Gobble Gobble
 $(if $(filter __%, $(MAKECMDGOALS)), \
 	$(error targets prefixed with '__' are only for internal use))
 
+LLNAME = linux-lucjan
+LLVERSION = .ll34
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := __all
 __all:
-- 
2.34.1.75.gabe6bb3905

