From 7908956d1b836cb3a6a2e36b94e0a6e767464b94 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 14 Jan 2021 11:54:08 +0100
Subject: [PATCH 21/34] block, bfq, DEBUG: log I/O-intensity parameters

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index aac6a73f7..6ed3588fd 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -2083,6 +2083,10 @@ static void bfq_update_io_intensity(struct bfq_queue *bfqq, u64 now_ns)
 	else
 		bfq_mark_bfqq_IO_bound(bfqq);
 
+	bfq_log_bfqq(bfqq->bfqd, bfqq, "tot %llu idle %llu bound %d",
+		     tot_io_time, bfqq->tot_idle_time,
+		     bfq_bfqq_IO_bound(bfqq));
+
 	/*
 	 * Keep an observation window of at most 200 ms in the past
 	 * from now.
-- 
2.31.1.362.g311531c9de

