From 66d9c52d55d854975db10c5441e7bb227ef65b51 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 18 Jun 2020 10:50:16 +0200
Subject: [PATCH 28/34] block, bfq, DEBUG: improve detailed log on denied
 dispatch

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 9385ccb79..fa45b71a6 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5268,9 +5268,15 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 				bfq_log_bfqq(bfqd, bfqq,
 					"chosen the queue %d for injection",
 					bfq_get_first_task_pid(new_bfqq));
+			else
+				bfq_log_bfqq(bfqd, bfqq,
+			"idling with %d busy queues, as no injector found",
+					     bfq_tot_busy_queues(bfqd));
+
 			bfqq = new_bfqq;
 		} else {
-			bfq_log_bfqq(bfqd, bfqq, "idling with %d busy queues",
+			bfq_log_bfqq(bfqd, bfqq,
+			"idling with %d busy queues, intentionally",
 				     bfq_tot_busy_queues(bfqd));
 			bfqq = NULL;
 		}
-- 
2.31.1.362.g311531c9de

