From 8647b9d02c8afe59fdadf23d45e834b3f242542c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 5 May 2019 11:16:44 +0800
Subject: [PATCH 2/7] bmq: Fix rq missing on arm arch.

---
 kernel/sched/bmq.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 3298456cf209..f86f78d7b0c2 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -177,8 +177,12 @@ static inline void update_sched_rq_watermark(struct rq *rq)
 		return;
 
 	cpu = cpu_of(rq);
-	if (!cpumask_andnot(&sched_rq_watermark[last_wm],
-			    &sched_rq_watermark[last_wm], cpumask_of(cpu)))
+#ifdef CONFIG_X86
+	__cpumask_clear_cpu(cpu, &sched_rq_watermark[last_wm]);
+#else
+	cpumask_clear_cpu(cpu, &sched_rq_watermark[last_wm]);
+#endif
+	if (cpumask_empty(&sched_rq_watermark[last_wm]))
 		clear_bit(last_wm, sched_rq_watermark_bitmap);
 	cpumask_set_cpu(cpu, &sched_rq_watermark[watermark]);
 	set_bit(watermark, sched_rq_watermark_bitmap);
-- 
2.22.0.rc1

