From 382cce8548daa72ddd6ad7ba3a1262578e109f38 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 14 May 2019 09:38:18 +0800
Subject: [PATCH 7/7] Tag BMQ v0.95

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index ef94fafb182e..1d584e69718f 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.94 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.95 by Alfred Chen.\n");
 }
 
 /**
-- 
2.22.0.rc1

