From 7f55c388c3b6ccf480ebbdc58b1a8889901fa793 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 15 Feb 2021 11:44:39 +0100
Subject: [PATCH 2/4] mm-5.11: inequality should not exclude lower threshold

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index d763f5fd0..14e1dd828 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2805,7 +2805,7 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 		sc->file_is_low = K(file) < sysctl_unevictable_file_kbytes_low &&
 				  K(file) > sysctl_unevictable_file_kbytes_min;
 
-		sc->file_is_min = K(file) < sysctl_unevictable_file_kbytes_min;
+		sc->file_is_min = K(file) <= sysctl_unevictable_file_kbytes_min;
 #endif
 	}
 
-- 
2.30.1.457.gf011795891

