diff --git a/block/blk-mq.c b/block/blk-mq.c
index d74c89c624c4..9ba4f40421b5 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2948,8 +2948,8 @@ int blk_mq_update_nr_requests(struct request_queue *q, unsigned int nr)
 		}
 		if (ret)
 			break;
-		if (q->elevator && q->elevator->type->ops.depth_updated)
-			q->elevator->type->ops.depth_updated(hctx);
+		if (q->elevator && q->elevator->type->ops.mq.depth_updated)
+			q->elevator->type->ops.mq.depth_updated(hctx);
 	}
 
 	if (!ret)
