From 365dcfb21aa68bde1ceca343558e5fae7eb10724 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 21 Dec 2020 11:23:51 +0100
Subject: [PATCH 2/3] mm-5.10: fix unevictable nits

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/Kconfig  | 2 +-
 mm/vmscan.c | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/mm/Kconfig b/mm/Kconfig
index 310111a0d..0b1047994 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -105,7 +105,7 @@ config UNEVICTABLE_INACTIVEFILE
 	  See also: Documentation/admin-guide/sysctl/vm.rst
 
 config UNEVICTABLE_INACTIVEFILE_KBYTES
-	int "Set default value for vm.unevictable_inactivefile_kbytes"
+	int "Default value for vm.unevictable_inactivefile_kbytes"
 	depends on UNEVICTABLE_INACTIVEFILE
 	default "0"
 
diff --git a/mm/vmscan.c b/mm/vmscan.c
index ccd6839c2..7bb5ae75e 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -167,11 +167,11 @@ struct scan_control {
 #endif
 
 #if defined(CONFIG_UNEVICTABLE_ACTIVEFILE)
-extern unsigned int sysctl_unevictable_activefile_kbytes;
+extern unsigned long sysctl_unevictable_activefile_kbytes;
 #endif
 
 #if defined(CONFIG_UNEVICTABLE_INACTIVEFILE)
-extern unsigned int sysctl_unevictable_inactivefile_kbytes;
+extern unsigned long sysctl_unevictable_inactivefile_kbytes;
 #endif
 
 /*
-- 
2.30.0.rc0

