From bdab16099584e0c2e9dca411c9ae75ab13711d0b Mon Sep 17 00:00:00 2001
From: Redecorating <69827514+Redecorating@users.noreply.github.com>
Date: Fri, 24 Dec 2021 18:12:25 +1100
Subject: [PATCH 03/32] add modalias to apple-bce

---
 drivers/staging/apple-bce/apple_bce.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/staging/apple-bce/apple_bce.c b/drivers/staging/apple-bce/apple_bce.c
index ad89632df..5e2f2f3b9 100644
--- a/drivers/staging/apple-bce/apple_bce.c
+++ b/drivers/staging/apple-bce/apple_bce.c
@@ -439,5 +439,6 @@ MODULE_LICENSE("GPL");
 MODULE_AUTHOR("MrARM");
 MODULE_DESCRIPTION("Apple BCE Driver");
 MODULE_VERSION("0.01");
+MODULE_ALIAS("pci:v0000106Bd00001801sv*sd*bc*sc*i*");
 module_init(apple_bce_module_init);
 module_exit(apple_bce_module_exit);
-- 
2.45.1.145.g83f1add914

