diff -Narup a/Makefile b/Makefile
--- a/Makefile	2018-11-05 00:37:52.000000000 +0100
+++ b/Makefile	2018-11-09 20:35:57.742780679 +0100
@@ -15,6 +15,10 @@ NAME = "People's Front"
 PHONY := _all
 _all:
 
+LLVERSION = .ll1
+LLNAME = linux-lucjan
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # Do not use make's built-in rules and variables
 # (this increases performance and avoids hard-to-debug behaviour)
 MAKEFLAGS += -rR
