diff -Narup a/Makefile b/Makefile
--- a/Makefile	2018-12-17 00:46:55.000000000 +0100
+++ b/Makefile	2018-12-23 21:12:21.554949701 +0100
@@ -11,6 +11,10 @@ NAME = Shy Crocodile
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLNAME = linux-lucjan
+LLVERSION = .ll6
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
