From 0fe2831c1bc24e030b2223d4ae79625dd6b8ea33 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 24 Dec 2019 14:43:10 +0800
Subject: [PATCH 1/4] bmq: Change SCHED_BMQ Kconfig location.

---
 init/Kconfig | 37 +++++++++++++++++--------------------
 1 file changed, 17 insertions(+), 20 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index c135142a2c85..674d2e8b8d36 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -73,26 +73,6 @@ config THREAD_INFO_IN_TASK
 
 menu "General setup"
 
-config SCHED_BMQ
-	bool "BMQ CPU scheduler"
-	help
-	  The BitMap Queue CPU scheduler for excellent interactivity and
-	  responsiveness on the desktop and solid scalability on normal
-	  hardware and commodity servers.
-
-	  Currently incompatible with the Group CPU scheduler, and RCU TORTURE
-          TEST so these options are disabled.
-
-          Say Y here.
-	default y
-
-config SCHED_TIMESLICE
-	int "Scheduler Task time slice"
-	depends on SCHED_BMQ
-	help
-	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
-	default 4
-
 config BROKEN
 	bool
 
@@ -737,6 +717,23 @@ config GENERIC_SCHED_CLOCK
 
 menu "Scheduler features"
 
+config SCHED_BMQ
+	bool "BMQ CPU scheduler"
+	help
+	  The BitMap Queue CPU scheduler for excellent interactivity and
+	  responsiveness on the desktop and solid scalability on normal
+	  hardware and commodity servers.
+
+          Say Y here.
+	default y
+
+config SCHED_TIMESLICE
+	int "Scheduler Task time slice"
+	depends on SCHED_BMQ
+	help
+	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
+	default 4
+
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
-- 
2.25.0.2.g232378479e.dirty


From fcf7aea5a0914353889e27fe5a41aa57b471e570 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 16 Jan 2020 11:03:50 +0800
Subject: [PATCH 2/4] bmq: Update recommend SCHED_TIMESLICE

Recommend 2 in ms for PREEMPT system and 4ms for others.
---
 init/Kconfig | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 674d2e8b8d36..f9faeb82f677 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -732,7 +732,8 @@ config SCHED_TIMESLICE
 	depends on SCHED_BMQ
 	help
 	  Time slice in ms for BMQ CPU scheduler, default 4 ms.
-	default 4
+	default 2 if PREEMPT
+	default 4 if !PREEMPT
 
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
-- 
2.25.0.2.g232378479e.dirty


From bcdf6f6786d9878a4ecc1bec261a07ed53f52680 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 19 Jan 2020 08:54:56 +0800
Subject: [PATCH 3/4] bmq: Update yield_type documentation.

---
 Documentation/admin-guide/sysctl/kernel.rst | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/Documentation/admin-guide/sysctl/kernel.rst b/Documentation/admin-guide/sysctl/kernel.rst
index 2fb0b416d04b..97ea247cc43a 100644
--- a/Documentation/admin-guide/sysctl/kernel.rst
+++ b/Documentation/admin-guide/sysctl/kernel.rst
@@ -1184,5 +1184,5 @@ BMQ CPU scheduler only. This determines what type of yield calls to
 sched_yield will perform.
 
   0 - No yield.
-  1 - Yield only to better priority/deadline tasks. (default)
-  2 - Expire timeslice and recalculate deadline.
+  1 - Deboost and requeue task. (default)
+  2 - Set run queue skip task.
-- 
2.25.0.2.g232378479e.dirty


From eac9c85e232d0a9bd6d43bfd2f66daee6a132d31 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 19 Jan 2020 08:37:33 +0800
Subject: [PATCH 4/4] Tag BMQ v5.4-r2

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 5c5603249b78..42a2a5b3d172 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r1 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r2 by Alfred Chen.\n");
 }
 
 /**
-- 
2.25.0.2.g232378479e.dirty

