From 852a1406a9e5329068aa89e65089d8bf16448903 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 4 Jan 2021 11:30:51 +0100
Subject: [PATCH 2/2] v4l2loopback-5.10: confine v4l2loopback_cleanup_module

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/media/v4l2-core/v4l2loopback.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/media/v4l2-core/v4l2loopback.c b/drivers/media/v4l2-core/v4l2loopback.c
index 030bc5cf90b0..74af58d558c5 100644
--- a/drivers/media/v4l2-core/v4l2loopback.c
+++ b/drivers/media/v4l2-core/v4l2loopback.c
@@ -2863,6 +2863,7 @@ static int __init v4l2loopback_init_module(void)
 	return err;
 }
 
+#ifdef MODULE
 static void v4l2loopback_cleanup_module(void)
 {
 	MARK();
@@ -2872,6 +2873,7 @@ static void v4l2loopback_cleanup_module(void)
 	misc_deregister(&v4l2loopback_misc);
 	dprintk("module removed\n");
 }
+#endif
 
 MODULE_ALIAS_MISCDEV(MISC_DYNAMIC_MINOR);
 MODULE_ALIAS("devname:v4l2loopback");
-- 
2.30.0

