From 02a22ba013fb1a8e25e791c413b294f91d72db9d Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 5 Mar 2019 15:45:53 +0800
Subject: [PATCH 04/12] bmq: BATCH/IDLE tasks always deboost when deactivated.

---
 kernel/sched/bmq.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index e1684a47f931..de75483806a3 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -3150,7 +3150,9 @@ static void __sched notrace __schedule(bool preempt)
 			prev->state = TASK_RUNNING;
 		} else {
 			prev->ts_deboost |= ts_over_run(rq);
-			if (prev->ts_deboost)
+			if ((SCHED_NORMAL == prev->policy && prev->ts_deboost) ||
+			    SCHED_BATCH == prev->policy ||
+			    SCHED_IDLE == prev->policy)
 				deboost_task(prev, 1);
 			deactivate_task(prev, rq);
 
-- 
2.21.0.155.ge902e9bcae

