From d7a79b67212c5bbb46cf2cf3abf97bd62b0b3378 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 7 Mar 2019 11:01:18 +0800
Subject: [PATCH 08/12] bmq: Remove unused wakexxx fields in BMQ.

---
 include/linux/sched.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched.h b/include/linux/sched.h
index cd7b77d954ba..23a3d91aa128 100644
--- a/include/linux/sched.h
+++ b/include/linux/sched.h
@@ -623,11 +623,11 @@ struct task_struct {
 	/* Current CPU: */
 	unsigned int			cpu;
 #endif
+#ifndef CONFIG_SCHED_BMQ
 	unsigned int			wakee_flips;
 	unsigned long			wakee_flip_decay_ts;
 	struct task_struct		*last_wakee;
 
-#ifndef CONFIG_SCHED_BMQ
 	/*
 	 * recent_used_cpu is initially set as the last CPU used by a task
 	 * that wakes affine another task. Waker/wakee relationships can
@@ -636,8 +636,8 @@ struct task_struct {
 	 * used CPU that may be idle.
 	 */
 	int				recent_used_cpu;
-#endif /* CONFIG_SCHED_BMQ */
 	int				wake_cpu;
+#endif /* !CONFIG_SCHED_BMQ */
 #endif
 	int				on_rq;
 
-- 
2.21.0.155.ge902e9bcae

