From 292ae9e1019d508ceb33176518d8bc34ef881e6d Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Tue, 10 Feb 2026 18:12:20 +0800
Subject: [PATCH 2/4] r8125: Enable default feature set as in the OOT Makefile

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/net/ethernet/realtek/r8125.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/net/ethernet/realtek/r8125.h b/drivers/net/ethernet/realtek/r8125.h
index e6948108d..57b2b9487 100755
--- a/drivers/net/ethernet/realtek/r8125.h
+++ b/drivers/net/ethernet/realtek/r8125.h
@@ -35,6 +35,13 @@
 #ifndef __R8125_H
 #define __R8125_H
 
+#define CONFIG_SOC_LAN
+#define CONFIG_ASPM
+#define ENABLE_S5WOL
+#define ENABLE_EEE
+#define ENABLE_TX_NO_CLOSE
+#define ENABLE_GIGA_LITE
+
 //#include <linux/pci.h>
 #include <linux/ethtool.h>
 #include <linux/interrupt.h>
-- 
2.53.0

