From 19f59e790e15426250443fd1c87492f2f3277f94 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 16 Mar 2020 19:13:12 +0100
Subject: [PATCH 08/18] block, bfq, DEBUG: remove useless BUG_ON in
 bfq_pd_offline

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-cgroup.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/block/bfq-cgroup.c b/block/bfq-cgroup.c
index 82667af01829..af7dfc3a605c 100644
--- a/block/bfq-cgroup.c
+++ b/block/bfq-cgroup.c
@@ -943,7 +943,6 @@ static void bfq_pd_offline(struct blkg_policy_data *pd)
 	 * deactivating the group itself.
 	 */
 	for (i = 0; i < BFQ_IOPRIO_CLASSES; i++) {
-		BFQ_BUG_ON(!bfqg->sched_data.service_tree);
 		st = bfqg->sched_data.service_tree + i;
 
 		/*
-- 
2.26.0.51.ga7d14a4428

