From 5b405d644d394d9d0dca628a91d4cb9184103c68 Mon Sep 17 00:00:00 2001
From: ckolivas <kernel@kolivas.org>
Date: Mon, 17 Aug 2026 21:57:38 +1000
Subject: [PATCH 13/15] Trivial documentation for sched cache.

---
 kernel/Kconfig.MuQSS | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/Kconfig.MuQSS b/kernel/Kconfig.MuQSS
index 3dad0e814..6b0ec452d 100644
--- a/kernel/Kconfig.MuQSS
+++ b/kernel/Kconfig.MuQSS
@@ -205,6 +205,7 @@ config MUQSS_IOTIME
 #   init/Kconfig                              UCLAMP_TASK
 #                                             SCHED_PROXY_EXEC
 #                                             NUMA_BALANCING
+#                                             SCHED_CACHE
 #                                             SCHED_MM_CID
 #                                             CGROUP_CPUACCT
 #                                             SCHED_AUTOGROUP
-- 
2.55.0

