From e627700c646b1fcd664ac383b173889a7b85a953 Mon Sep 17 00:00:00 2001
From: ckolivas <kernel@kolivas.org>
Date: Thu, 20 Aug 2026 06:55:51 +1000
Subject: [PATCH 14/15] Fix PSI irq time accounting build for default arch
 config.

---
 kernel/sched/MuQSS.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/MuQSS.h b/kernel/sched/MuQSS.h
index 0c461e650..7c1a90e7b 100644
--- a/kernel/sched/MuQSS.h
+++ b/kernel/sched/MuQSS.h
@@ -309,6 +309,7 @@ struct rq {
 
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 	u64 prev_irq_time;
+	u64 psi_irq_time;
 #endif /* CONFIG_IRQ_TIME_ACCOUNTING */
 #ifdef CONFIG_PARAVIRT
 	u64 prev_steal_time;
-- 
2.55.0

