From 9b38348ab57d6e62b5455272b378d5bde1ca0bcd Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 20 Apr 2021 23:41:15 +0200
Subject: [PATCH 117/117] block,bfq: improve bfq-lucjan branding

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 9b8103fd8..6be17a3d7 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7293,7 +7293,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ I/O-scheduler: BFQ-lucjan";
+	char msg[60] = "BFQ I/O-scheduler: BFQ-lucjan v5.11";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.31.1.305.gd1b10fc6d8

