From bca654d740a2999bb9822e850acb5e317df351ad Mon Sep 17 00:00:00 2001
From: Peter Jung <admin@ptr1337.dev>
Date: Mon, 16 Dec 2024 19:35:45 +0100
Subject: [PATCH 3/3] Cachy: Fix Zen5 and Clang CPU Optimization

Signed-off-by: Peter Jung <admin@ptr1337.dev>
---
 arch/x86/Kconfig.cpu | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig.cpu b/arch/x86/Kconfig.cpu
index 46130801c..c73ce84f2 100644
--- a/arch/x86/Kconfig.cpu
+++ b/arch/x86/Kconfig.cpu
@@ -274,7 +274,7 @@ config MZEN4
 
 config MZEN5
 	bool "AMD Zen 5"
-	depends on (CC_IS_GCC && GCC_VERSION > 140000) || (CC_IS_CLANG && CLANG_VERSION >= 191000)
+	depends on (CC_IS_GCC && GCC_VERSION > 140000) || (CC_IS_CLANG && CLANG_VERSION >= 190000)
 	help
 	  Select this for AMD Family 19h Zen 5 processors.
 
-- 
2.48.1

