From 80aa8ef7e7ffd20761cec291d30768ea868e1a73 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 13 Apr 2026 01:30:37 +0800
Subject: [PATCH 16/16] mglru: Add changes from v5

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 mm/vmscan.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index f073cd454..35ad87c04 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -4787,7 +4787,7 @@ static long get_nr_to_scan(struct lruvec *lruvec, struct scan_control *sc,
 	 * to keep reclaim moving forwards.
 	 */
 	nr_to_scan >>= sc->priority;
-	if (!nr_to_scan)
+	if (!nr_to_scan && sc->priority < DEF_PRIORITY)
 		nr_to_scan = min(evictable, SWAP_CLUSTER_MAX);
 
 	return nr_to_scan;
-- 
2.53.0

