From 79bb5670c0256d75eb56fa1f587fc6580ac1237c Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 28 Sep 2020 20:20:29 +0200
Subject: [PATCH 1/4] sched/alt: Export can_nice symbol for Android Binder IPC

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/alt_core.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 9576c57f82da..9109a24c21c9 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -5145,6 +5145,7 @@ int can_nice(const struct task_struct *p, const int nice)
 	return (nice_rlim <= task_rlimit(p, RLIMIT_NICE) ||
 		capable(CAP_SYS_NICE));
 }
+EXPORT_SYMBOL_GPL(can_nice);
 
 #ifdef __ARCH_WANT_SYS_NICE
 
-- 
2.33.1.711.g9d530dc002


From ad60d67092aae4ec27d2283f9c1e09fb7c466099 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 29 Apr 2021 02:53:36 +0200
Subject: [PATCH 2/4] sched/alt: Add MG-LRU changes through ifdef macro

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/alt_core.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 9109a24c21c9..cfc0ef116b44 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -3543,6 +3543,9 @@ context_switch(struct rq *rq, struct task_struct *prev,
 		 * finish_task_switch()'s mmdrop().
 		 */
 		switch_mm_irqs_off(prev->active_mm, next->mm, next);
+#ifdef CONFIG_LRU_GEN
+		lru_gen_switch_mm(prev->active_mm, next->mm);
+#endif
 
 		if (!prev->mm) {                        // from kernel
 			/* will mmdrop() in finish_task_switch(). */
@@ -6640,6 +6643,9 @@ void idle_task_exit(void)
 
 	if (mm != &init_mm) {
 		switch_mm(mm, &init_mm, current);
+#ifdef CONFIG_LRU_GEN
+		lru_gen_switch_mm(mm, &init_mm);
+#endif
 		finish_arch_post_lock_switch();
 	}
 
-- 
2.33.1.711.g9d530dc002


From b5a08413253862ac25fab9ea618195ec2bc06804 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 17 Sep 2020 01:38:47 +0200
Subject: [PATCH 3/4] init, Kconfig: set default value of SCHED_PDS

Prefer PDS to BMQ by default due to better CPU utilisation & higher throughput.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c8e82fcafb9e..c5e55b64c032 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -824,7 +824,7 @@ if SCHED_ALT
 
 choice
 	prompt "Alternative CPU Scheduler"
-	default SCHED_BMQ
+	default SCHED_PDS
 
 config SCHED_BMQ
 	bool "BMQ CPU scheduler"
-- 
2.33.1.711.g9d530dc002


From b7790d1111abadcea34a7cf812519419135516c7 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 21 Sep 2020 17:26:14 +0200
Subject: [PATCH 4/4] init, Kconfig: Restore original PDS description

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c5e55b64c032..2cf5301a4255 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -837,7 +837,9 @@ config SCHED_PDS
 	bool "PDS CPU scheduler"
 	help
 	  The Priority and Deadline based Skip list multiple queue CPU
-	  Scheduler.
+	  scheduler for excellent interactivity and responsiveness on the
+	  desktop and solid scalability on normal hardware and commodity
+	  servers.
 
 endchoice
 
-- 
2.33.1.711.g9d530dc002

