From 081f935aa48e4e9adab0b15350a2979c07051783 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 14 Jun 2021 15:46:03 +0200
Subject: [PATCH] prjc-5.12: fix compilation error: if (!sched_feat(UTIL_EST))

Bug report: https://gitlab.com/alfredchen/linux-prjc/-/issues/27

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/pelt.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pelt.h b/kernel/sched/pelt.h
index bc722a476..26a33f76b 100644
--- a/kernel/sched/pelt.h
+++ b/kernel/sched/pelt.h
@@ -44,6 +44,7 @@ static inline u32 get_pelt_divider(struct sched_avg *avg)
 	return LOAD_AVG_MAX - 1024 + avg->period_contrib;
 }
 
+#ifndef CONFIG_SCHED_ALT
 static inline void cfs_se_util_change(struct sched_avg *avg)
 {
 	unsigned int enqueued;
@@ -61,7 +62,6 @@ static inline void cfs_se_util_change(struct sched_avg *avg)
 	WRITE_ONCE(avg->util_est.enqueued, enqueued);
 }
 
-#ifndef CONFIG_SCHED_ALT
 /*
  * The clock_pelt scales the time to reflect the effective amount of
  * computation done during the running delta time but then sync back to
-- 
2.32.0

