From 48639c208e21465811bb3ee364cc1eae875520b6 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 28 Aug 2023 13:14:47 +0200
Subject: [PATCH 26/26] MAINTAINERS: add Stephan Mueller as a LRNG maintainer

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 MAINTAINERS | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index 28e20975c..1e6b564b0 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -12572,6 +12572,13 @@ S:	Supported
 B:	mailto:linux-next@vger.kernel.org and the appropriate development tree
 T:	git git://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/
 
+LINUX RANDOM NUMBER GENERATOR (LRNG) DRIVER
+M:	Stephan Mueller <smueller@chronox.de>
+S:	Maintained
+W:	https://www.chronox.de/lrng.html
+F:	drivers/char/lrng/
+F:	include/linux/lrng.h
+
 LIS3LV02D ACCELEROMETER DRIVER
 M:	Eric Piel <eric.piel@tremplin-utc.net>
 S:	Maintained
-- 
2.45.1.145.g83f1add914

