From 67b1817085a3f062a38df48fbd37a8fbb550112f Mon Sep 17 00:00:00 2001
From: Anthony Ruhier <aruhier@mailbox.org>
Date: Fri, 14 May 2021 22:02:11 +0200
Subject: [PATCH 11/33] fs/ntfs3: Fix unsupported flags by clang

---
 fs/ntfs3/Makefile | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/fs/ntfs3/Makefile b/fs/ntfs3/Makefile
index b06a06cc0..5adc54eba 100644
--- a/fs/ntfs3/Makefile
+++ b/fs/ntfs3/Makefile
@@ -4,7 +4,11 @@
 #
 
 # to check robot warnings
-ccflags-y += -Wunused-but-set-variable -Wold-style-declaration -Wint-to-pointer-cast
+ccflags-y += -Wint-to-pointer-cast
+condflags := \
+	$(call cc-option, -Wunused-but-set-variable) \
+	$(call cc-option, -Wold-style-declaration)
+ccflags-y += $(condflags)
 
 obj-$(CONFIG_NTFS3_FS) += ntfs3.o
 
@@ -31,4 +35,4 @@ ntfs3-$(CONFIG_NTFS3_LZX_XPRESS) += $(addprefix lib/,\
 		decompress_common.o \
 		lzx_decompress.o \
 		xpress_decompress.o \
-		)
\ No newline at end of file
+		)
-- 
2.33.0

