From dd4592517aa54db96721014efc758ed2bead3ab2 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 20 Jul 2017 17:46:46 +0200
Subject: [PATCH] Check presence on tree of every entity after every activation

---
 block/bfq-sched.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/block/bfq-sched.c b/block/bfq-sched.c
index b6eb258..ad200f3 100644
--- a/block/bfq-sched.c
+++ b/block/bfq-sched.c
@@ -1284,6 +1284,7 @@ static void bfq_activate_requeue_entity(struct bfq_entity *entity,
 					bool requeue)
 {
 	struct bfq_sched_data *sd;
+	struct bfq_entity *old_entity = entity;
 
 	for_each_entity(entity) {
 		BUG_ON(!entity);
@@ -1300,6 +1301,15 @@ static void bfq_activate_requeue_entity(struct bfq_entity *entity,
 		}
 		BUG_ON(!sd->next_in_service);
 	}
+
+	entity = old_entity;
+	for_each_entity(entity) {
+		struct bfq_service_tree *st = bfq_entity_service_tree(entity);
+
+		BUG_ON(!entity->tree);
+		BUG_ON(entity->tree == &st->idle);
+		BUG_ON(entity->tree != &st->active);
+	}
 }
 
 /**
-- 
2.10.0

