From 77ebd9287765a7993368345c1cac2706c682276c Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <lkml@antheas.dev>
Date: Thu, 17 Jul 2025 23:12:32 +0200
Subject: [PATCH 083/104] HID: hid-steam: keep evdev connected while steam is
 active

Signed-off-by: Antheas Kapenekakis <lkml@antheas.dev>
---
 drivers/hid/hid-steam.c | 18 +++++-------------
 1 file changed, 5 insertions(+), 13 deletions(-)

diff --git a/drivers/hid/hid-steam.c b/drivers/hid/hid-steam.c
index 197126d6e..09abb62f7 100644
--- a/drivers/hid/hid-steam.c
+++ b/drivers/hid/hid-steam.c
@@ -983,11 +983,11 @@ static int steam_register(struct steam_device *steam)
 	client_opened = steam->client_opened;
 	spin_unlock_irqrestore(&steam->lock, flags);
 
+	ret = steam_input_register(steam);
+	if (ret != 0)
+		goto steam_register_input_fail;
 	if (!client_opened) {
 		steam_set_lizard_mode(steam, lizard_mode);
-		ret = steam_input_register(steam);
-		if (ret != 0)
-			goto steam_register_input_fail;
 		ret = steam_sensors_register(steam);
 		if (ret != 0)
 			goto steam_register_sensors_fail;
@@ -1081,12 +1081,11 @@ static void steam_work_unregister_cb(struct work_struct *work)
 	spin_unlock_irqrestore(&steam->lock, flags);
 
 	if (connected) {
+		steam_input_register(steam);
 		if (opened) {
 			steam_sensors_unregister(steam);
-			steam_input_unregister(steam);
 		} else {
 			steam_set_lizard_mode(steam, lizard_mode);
-			steam_input_register(steam);
 			steam_sensors_register(steam);
 		}
 	}
@@ -1615,9 +1614,6 @@ static void steam_do_deck_input_event(struct steam_device *steam,
 		schedule_delayed_work(&steam->mode_switch, 45 * HZ / 100);
 	}
 
-	if (!steam->gamepad_mode && lizard_mode)
-		return;
-
 	lpad_touched = b10 & BIT(3);
 	rpad_touched = b10 & BIT(4);
 
@@ -1767,8 +1763,6 @@ static int steam_raw_event(struct hid_device *hdev,
 
 	switch (data[2]) {
 	case ID_CONTROLLER_STATE:
-		if (steam->client_opened)
-			return 0;
 		rcu_read_lock();
 		input = rcu_dereference(steam->input);
 		if (likely(input))
@@ -1776,14 +1770,12 @@ static int steam_raw_event(struct hid_device *hdev,
 		rcu_read_unlock();
 		break;
 	case ID_CONTROLLER_DECK_STATE:
-		if (steam->client_opened)
-			return 0;
 		rcu_read_lock();
 		input = rcu_dereference(steam->input);
 		if (likely(input))
 			steam_do_deck_input_event(steam, input, data);
 		sensors = rcu_dereference(steam->sensors);
-		if (likely(sensors))
+		if (!steam->client_opened && likely(sensors))
 			steam_do_deck_sensors_event(steam, sensors, data);
 		rcu_read_unlock();
 		break;
-- 
2.53.0

