From 728ef8fcf604d9c069b7e02850f827a71cde088b Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <lkml@antheas.dev>
Date: Sun, 21 Sep 2025 15:47:29 +0200
Subject: [PATCH 093/104] tmp: add MSI Center button event to msi-wmi

Signed-off-by: Antheas Kapenekakis <lkml@antheas.dev>
---
 drivers/platform/x86/msi-wmi.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/msi-wmi.c b/drivers/platform/x86/msi-wmi.c
index 4a7ac85c4..20de3173e 100644
--- a/drivers/platform/x86/msi-wmi.c
+++ b/drivers/platform/x86/msi-wmi.c
@@ -41,6 +41,7 @@ enum msi_scancodes {
 	MSI_KEY_MUTE,
 	/* MSI Wind keys */
 	WIND_KEY_TOUCHPAD	= 0x08,	/* Fn+F3 touchpad toggle */
+	WIND_KEY_MSI_CENTER	= 0x29,
 	WIND_KEY_BLUETOOTH	= 0x56,	/* Fn+F11 Bluetooth toggle */
 	WIND_KEY_CAMERA,		/* Fn+F6 webcam toggle */
 	WIND_KEY_WLAN		= 0x5f,	/* Fn+F11 Wi-Fi toggle */
@@ -68,6 +69,7 @@ static struct key_entry msi_wmi_keymap[] = {
 	/* These are MSI Wind keys that should be handled via WMI */
 	{ KE_KEY, WIND_KEY_TURBO,		{KEY_PROG1} },
 	{ KE_KEY, WIND_KEY_ECO,			{KEY_PROG2} },
+	{ KE_KEY, WIND_KEY_MSI_CENTER,		{KEY_F15}},
 
 	{ KE_END, 0 }
 };
@@ -173,14 +175,24 @@ static const struct backlight_ops msi_backlight_ops = {
 static void msi_wmi_notify(union acpi_object *obj, void *context)
 {
 	struct key_entry *key;
+	int eventcode;
+
+	if (obj) {
+		if (obj->type == ACPI_TYPE_INTEGER)
+			eventcode = obj->integer.value;
+		else if (obj->type == ACPI_TYPE_BUFFER && obj->buffer.length)
+			eventcode = obj->buffer.pointer[0];
+		else {
+			pr_info("Unknown event received\n");
+			return;
+		}
 
-	if (obj && obj->type == ACPI_TYPE_INTEGER) {
-		int eventcode = obj->integer.value;
 		pr_debug("Eventcode: 0x%x\n", eventcode);
 		key = sparse_keymap_entry_from_scancode(msi_wmi_input_dev,
 				eventcode);
 		if (!key) {
-			pr_info("Unknown key pressed - %x\n", eventcode);
+			pr_info("Unknown key pressed - %x (ACPI type: %d)\n",
+				eventcode, obj->type);
 			return;
 		}
 
-- 
2.53.0

