From 7b806cb9018a340e97bd8035b7c3bb10c84740c3 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 10 Feb 2021 19:21:30 +0100
Subject: [PATCH 6/6] sched/pds: Set pds-dev instead of pds

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/pds_imp.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds_imp.h b/kernel/sched/pds_imp.h
index 6de20895819c..fe732699daa4 100644
--- a/kernel/sched/pds_imp.h
+++ b/kernel/sched/pds_imp.h
@@ -87,7 +87,7 @@ void sched_internal_debug(void)
 }
 #endif
 
-#define ALT_SCHED_VERSION_MSG "sched/pds: PDS CPU Scheduler "ALT_SCHED_VERSION" by Alfred Chen.\n"
+#define ALT_SCHED_VERSION_MSG "sched/pds: PDS-dev CPU Scheduler (with bitmap allocator) "ALT_SCHED_VERSION" by Alfred Chen.\n"
 
 static const u64 user_prio2deadline[NICE_WIDTH] = {
 /* -20 */	  4194304,   4613734,   5075107,   5582617,   6140878,
-- 
2.32.0.93.g670b81a890

