From 923as3ks8a038a09e8a98237283aa922323q82a Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 27 Dec 2018 18:13:15 +0000
Subject: [PATCH 2/5] LL: block/Kconfig.iosched: Fix depends for BLK_CGROUP_IOLATENCY

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/Kconfig.iosched | 3 ++-
 1 file changed, 2 insertions(+), 1 deletions(-)

diff --git a/block/Kconfig.iosched b/block/Kconfig.iosched
index 378dfb268b7f..e4e60008e5c6 100644
--- a/block/Kconfig.iosched
+++ b/block/Kconfig.iosched
@@ -53,6 +53,7 @@ config IOSCHED_BFQ_SQ
 config BFQ_SQ_GROUP_IOSCHED
 	bool "BFQ-SQ hierarchical scheduling support"
 	depends on IOSCHED_BFQ_SQ && BLK_CGROUP
+	select BLK_CGROUP_IOLATENCY
 	default n
 	---help---
 
@@ -111,7 +112,6 @@ config MQ_IOSCHED_BFQ
 	select DM_MQ_DEFAULT
 	select SCSI_MQ_DEFAULT
 	depends on BLOCK
-	select BLK_CGROUP_IOLATENCY
 	default y
 	---help---
 	BFQ I/O scheduler for BLK-MQ. BFQ-MQ distributes bandwidth
@@ -123,6 +123,7 @@ config MQ_IOSCHED_BFQ
 config MQ_BFQ_GROUP_IOSCHED
 	bool "BFQ-MQ hierarchical scheduling support"
 	depends on MQ_IOSCHED_BFQ && BLK_CGROUP
+	select BLK_CGROUP_IOLATENCY
 	default n
 	---help---

-- 
2.20.1.390.gb5101f9297

