From c147864c93b79e8b5d0d0d709b21b5bb64bafa6e Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 3 Jan 2019 20:23:41 +0100
Subject: [PATCH 3/3] LL: scsi: Always use the blk-mq I/O path

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/Kconfig.iosched | 2 --
 1 file changed, 2 deletions(-)

diff --git a/block/Kconfig.iosched b/block/Kconfig.iosched
index e4e60008e5c6..40af9df7904d 100644
--- a/block/Kconfig.iosched
+++ b/block/Kconfig.iosched
@@ -109,8 +109,6 @@ config DEFAULT_IOSCHED
 
 config MQ_IOSCHED_BFQ
 	tristate "BFQ-MQ I/O Scheduler"
-	select DM_MQ_DEFAULT
-	select SCSI_MQ_DEFAULT
 	depends on BLOCK
 	default y
 	---help---
-- 
2.20.1.2.gb21ebb671b

