From 04a8c71eb2bf647410fc756ecd6db26d30891ff4 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Thu, 5 Jan 2023 03:01:06 -0500
Subject: [PATCH 96/97] fixup! lib: add support for allocation tagging

---
 include/linux/alloc_tag.h | 2 --
 1 file changed, 2 deletions(-)

diff --git a/include/linux/alloc_tag.h b/include/linux/alloc_tag.h
index 488270da5..157984a21 100644
--- a/include/linux/alloc_tag.h
+++ b/include/linux/alloc_tag.h
@@ -9,7 +9,6 @@
 #include <linux/codetag.h>
 #include <linux/container_of.h>
 #include <linux/lazy-percpu-counter.h>
-#include <linux/sched.h>
 #include <linux/static_key.h>
 
 /*
@@ -128,7 +127,6 @@ static inline void set_codetag_empty(union codetag_ref *ref) {}
 static inline void alloc_tag_sub(union codetag_ref *ref, size_t bytes) {}
 static inline void alloc_tag_add(union codetag_ref *ref, struct alloc_tag *tag,
 				 size_t bytes) {}
-#define alloc_tag_restore(_tag, _old)
 
 #endif
 
-- 
2.39.0.rc2.1.gbd5df96b79

