From 1e3bb186d0bcd58182bdfa1827bb4f9b95fbcfde Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Thu, 5 Jan 2023 03:01:18 -0500
Subject: [PATCH 97/97] fixup! mm: enable page allocation tagging

---
 include/linux/gfp.h | 1 +
 1 file changed, 1 insertion(+)

diff --git a/include/linux/gfp.h b/include/linux/gfp.h
index 6e32ed27b..2be027c6c 100644
--- a/include/linux/gfp.h
+++ b/include/linux/gfp.h
@@ -7,6 +7,7 @@
 #include <linux/mmzone.h>
 #include <linux/topology.h>
 #include <linux/pgalloc_tag.h>
+#include <linux/sched.h>
 
 struct vm_area_struct;
 
-- 
2.39.0.rc2.1.gbd5df96b79

