From 047a79a7def0d34bf81e84c589ccf1a029d1857c Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 25 Apr 2024 10:27:05 +0200
Subject: [PATCH 18/20] amd-pstate-6.9: update fixes and enhancements patchset
 to v12

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 include/linux/amd-pstate.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/amd-pstate.h b/include/linux/amd-pstate.h
index c5e41de65..8ba5dd4d3 100644
--- a/include/linux/amd-pstate.h
+++ b/include/linux/amd-pstate.h
@@ -49,8 +49,8 @@ struct amd_aperf_mperf {
  * @lowest_perf: the absolute lowest performance level of the processor
  * @prefcore_ranking: the preferred core ranking, the higher value indicates a higher
  * 		  priority.
- * @min_limit_perf: Cached value of the perf corresponding to policy->min
- * @max_limit_perf: Cached value of the perf corresponding to policy->max
+ * @min_limit_perf: Cached value of the performance corresponding to policy->min
+ * @max_limit_perf: Cached value of the performance corresponding to policy->max
  * @min_limit_freq: Cached value of policy->min (in khz)
  * @max_limit_freq: Cached value of policy->max (in khz)
  * @max_freq: the frequency (in khz) that mapped to highest_perf
-- 
2.44.0.325.g11c821f2f2

