From 0572067ed4405bc7f01912f71b7d81949de1e6f5 Mon Sep 17 00:00:00 2001
From: Tejun Heo <tj@kernel.org>
Date: Fri, 23 Feb 2024 09:47:06 -1000
Subject: [PATCH 141/161] scx: Include cpu_to_kick_if_idle in debug dump

---
 kernel/sched/ext.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/sched/ext.c b/kernel/sched/ext.c
index 47ec319bb..efb01a7cb 100644
--- a/kernel/sched/ext.c
+++ b/kernel/sched/ext.c
@@ -3541,6 +3541,9 @@ static void scx_dump_state(struct scx_exit_info *ei)
 		if (!cpumask_empty(rq->scx.cpus_to_kick))
 			seq_buf_printf(&s, "  cpus_to_kick   : %*pb\n",
 				       cpumask_pr_args(rq->scx.cpus_to_kick));
+		if (!cpumask_empty(rq->scx.cpus_to_kick_if_idle))
+			seq_buf_printf(&s, "  idle_to_kick   : %*pb\n",
+				       cpumask_pr_args(rq->scx.cpus_to_kick_if_idle));
 		if (!cpumask_empty(rq->scx.cpus_to_preempt))
 			seq_buf_printf(&s, "  cpus_to_preempt: %*pb\n",
 				       cpumask_pr_args(rq->scx.cpus_to_preempt));
-- 
2.43.0.232.ge79552d197

