From 6b1d2837e9d05db5f593af4d84eb59f4924aa095 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 18 Jan 2021 15:14:43 +0100
Subject: [PATCH 14/55] block, bfq, DEBUG: log new weight and ioprio

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 29c2c50db..b55a18f78 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5656,6 +5656,8 @@ static void bfq_set_next_ioprio_data(struct bfq_queue *bfqq,
 	}
 
 	bfqq->entity.new_weight = bfq_ioprio_to_weight(bfqq->new_ioprio);
+	bfq_log_bfqq(bfqd, bfqq, "new_ioprio %d new_weight %d",
+		     bfqq->new_ioprio, bfqq->entity.new_weight);
 	bfqq->entity.prio_changed = 1;
 }
 
-- 
2.30.0.155.g66e871b664

