From b249f705f3db45a91896643898c8cb984cd5b8c1 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 4 Sep 2019 05:27:03 +0800
Subject: [PATCH 07/63] bmq: [Sync] ucamp

BMQ doesn't support ucamp.
---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index d871b32d6875..b90c809d555d 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -720,6 +720,7 @@ menu "Scheduler features"
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
+	depends on !SCHED_BMQ
 	help
 	  This feature enables the scheduler to track the clamped utilization
 	  of each CPU based on RUNNABLE tasks scheduled on that CPU.
-- 
2.25.0.2.g232378479e.dirty

