From d3d7c506d9dd6d1633df21c7533a3dde5e9e8a5c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 3 Nov 2019 10:10:07 +0800
Subject: [PATCH 53/63] Tag BMQ v5.4-r0

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 206621482cfd..6c52f2d5b797 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.3-r2 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.4-r0 by Alfred Chen.\n");
 }
 
 /**
-- 
2.25.0.2.g232378479e.dirty

