From 3b48638c7d345d15a692a83e83718c2f983b066f Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sat, 19 Jan 2019 11:30:59 +0100
Subject: [PATCH] blk-mq: fix updating internal depth state for 4.20 series

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 block/blk-mq.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index d74c89c624c4..9ba4f40421b5 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2948,8 +2948,8 @@ int blk_mq_update_nr_requests(struct request_queue *q, unsigned int nr)
 		}
 		if (ret)
 			break;
-		if (q->elevator && q->elevator->type->ops.depth_updated)
-			q->elevator->type->ops.depth_updated(hctx);
+		if (q->elevator && q->elevator->type->ops.mq.depth_updated)
+			q->elevator->type->ops.mq.depth_updated(hctx);
 	}
 
 	if (!ret)
-- 
2.20.1.142.g77556354bb

