From 72da04f0d96e12bb20beaa33a980f7fd279110eb Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 8 Dec 2018 21:52:59 +0000
Subject: [PATCH 2/2] pds: Fix smt boot up crash.

---
 kernel/sched/pds.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index cddf591a1603..5a5ad448f23c 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -619,7 +619,7 @@ static inline void update_sched_rq_queued_masks(struct rq *rq)
 		return;
 
 #ifdef CONFIG_SCHED_SMT
-	if (~0 == per_cpu(sched_sibling_cpu, cpu))
+	if (cpu == per_cpu(sched_sibling_cpu, cpu))
 		return;
 
 	if (SCHED_RQ_EMPTY == last_level) {
@@ -6103,7 +6103,7 @@ void __init sched_init(void)
 		rq->queued_level = SCHED_RQ_EMPTY;
 		rq->pending_level = SCHED_RQ_EMPTY;
 #ifdef CONFIG_SCHED_SMT
-		per_cpu(sched_sibling_cpu, i) = ~0;
+		per_cpu(sched_sibling_cpu, i) = i;
 		rq->active_balance = 0;
 #endif
 #endif
-- 
2.20.0.rc2.7.g965798d1f2

