From 838b3cede8df6d9c9c5c5dc71067cfa8c9e320a8 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 8 Apr 2021 00:39:08 +0200
Subject: [PATCH 06/11] mm-5.11: be more concise about clean pages

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 6623af0d1..5e50353ff 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2816,14 +2816,14 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 			 * vbabka: ugh, never say never
 			 */
 			VM_WARN_ON_ONCE(1);
-		} else {
+			clean_file = ULONG_MAX;
+		} else
 			clean_file = reclaimable_file - dirty_file;
 
-			sc->file_is_low = K(clean_file) < sysctl_unevictable_file_kbytes_low &&
-				          K(clean_file) > sysctl_unevictable_file_kbytes_min;
+		sc->file_is_low = K(clean_file) < sysctl_unevictable_file_kbytes_low &&
+			          K(clean_file) > sysctl_unevictable_file_kbytes_min;
 
-			sc->file_is_min = K(clean_file) <= sysctl_unevictable_file_kbytes_min;
-		}
+		sc->file_is_min = K(clean_file) <= sysctl_unevictable_file_kbytes_min;
 #endif
 	}
 
-- 
2.31.0.97.g1424303384

