From 90dce1bdfdc91f11253febce25c7afde8e95fc6d Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 8 Apr 2021 08:08:50 +0200
Subject: [PATCH 07/11] mm-5.11: nits on branching

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 5e50353ff..267763fe2 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2809,13 +2809,13 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 		reclaimable_file = file + node_page_state(pgdat, NR_ISOLATED_FILE);
 		dirty_file = node_page_state(pgdat, NR_FILE_DIRTY);
 
-		if (reclaimable_file < dirty_file) {
+		if (unlikely(reclaimable_file < dirty_file)) {
 			/*
 			 * post-factum: so, nr_file_dirty can never exceed
 			 *              (nr_inactive_file+nr_active_file+nr_isolated_file)?
 			 * vbabka: ugh, never say never
 			 */
-			VM_WARN_ON_ONCE(1);
+			WARN_ON_ONCE(1);
 			clean_file = ULONG_MAX;
 		} else
 			clean_file = reclaimable_file - dirty_file;
-- 
2.31.0.97.g1424303384

