From 9aaa7d4fa23db7a886362f245336cdc7529e7a0c Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 8 Apr 2021 08:14:26 +0200
Subject: [PATCH 10/11] mm-5.11: explain non-atomicity

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index d7a4d80ff..3ac537721 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2810,6 +2810,10 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 			anon >> sc->priority;
 
 #if defined(CONFIG_UNEVICTABLE_FILE)
+		/*
+		 * node_page_state() sum can go out of sync since
+		 * all the values are not read at once
+		 */
 		if (unlikely(reclaimable_file < dirty_file))
 			clean_file = ULONG_MAX;
 		else
-- 
2.31.0.97.g1424303384

