From 26f289a07902d8805f4ee0a7226c2d08691a18c2 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Thu, 8 Apr 2021 16:40:18 +0200
Subject: [PATCH 11/11] mm-5.11: do not disable file pages protection under
 dirty pressure

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 mm/vmscan.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/mm/vmscan.c b/mm/vmscan.c
index 3ac537721..c8ebf5266 100644
--- a/mm/vmscan.c
+++ b/mm/vmscan.c
@@ -2815,7 +2815,11 @@ static void shrink_node(pg_data_t *pgdat, struct scan_control *sc)
 		 * all the values are not read at once
 		 */
 		if (unlikely(reclaimable_file < dirty_file))
-			clean_file = ULONG_MAX;
+			/*
+			 * in this case assume the system does not have
+			 * clean file pages anymore
+			 */
+			clean_file = 0;
 		else
 			clean_file = reclaimable_file - dirty_file;
 
-- 
2.31.0.97.g1424303384

