From b04808f8f0574b16fdff99ba24b7ad797739333f Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:22:12 -0400
Subject: [PATCH 001/104] make DEFAULT_MMAP_MIN_ADDR match LSM_MMAP_MIN_ADDR

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 mm/Kconfig | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/Kconfig b/mm/Kconfig
index 034d87953..fd932dac4 100644
--- a/mm/Kconfig
+++ b/mm/Kconfig
@@ -314,7 +314,8 @@ config KSM
 config DEFAULT_MMAP_MIN_ADDR
 	int "Low address space to protect from user allocation"
 	depends on MMU
-	default 4096
+	default 32768 if ARM || (ARM64 && COMPAT)
+	default 65536
 	help
 	  This is the portion of low virtual memory which should be protected
 	  from userspace allocation.  Keeping a user from writing to low pages
-- 
2.36.1.203.g1bcf4f6271

