From fe622b189f3f9a576406d63ce6136567b55e1811 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 18:11:31 -0400
Subject: [PATCH 015/104] stop hiding X86_16BIT behind EXPERT

---
 arch/x86/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 4bed3abf4..7b21b2e30 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1204,7 +1204,7 @@ config VM86
 	default X86_LEGACY_VM86
 
 config X86_16BIT
-	bool "Enable support for 16-bit segments" if EXPERT
+	bool "Enable support for 16-bit segments"
 	default y
 	depends on MODIFY_LDT_SYSCALL
 	help
-- 
2.36.1.203.g1bcf4f6271

