From 9519ccee36eca387039c57c3fcffdb6c83e88f90 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 18:16:16 -0400
Subject: [PATCH 018/103] disable MODIFY_LDT_SYSCALL by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 arch/x86/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 7d96b71ba..4eed2eeb5 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2414,7 +2414,6 @@ config CMDLINE_OVERRIDE
 
 config MODIFY_LDT_SYSCALL
 	bool "Enable the LDT (local descriptor table)"
-	default y
 	help
 	  Linux can allow user programs to install a per-process x86
 	  Local Descriptor Table (LDT) using the modify_ldt(2) system
-- 
2.36.1.203.g1bcf4f6271

