From 6dced769059959e2d6aee797bc49bacc8f38e8ba Mon Sep 17 00:00:00 2001
From: Bernhard40 <32568352+Bernhard40@users.noreply.github.com>
Date: Fri, 6 Oct 2017 10:24:10 +0000
Subject: [PATCH 021/104] disable AIO by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 29d95c88d..73bd128ec 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1642,7 +1642,6 @@ config SHMEM
 
 config AIO
 	bool "Enable AIO support"
-	default y
 	help
 	  This option enables POSIX asynchronous I/O which may by used
 	  by some high performance threaded applications. Disabling
-- 
2.36.1.203.g1bcf4f6271

