From 3acedda87cbcf907f59c1c24c1d2cd11615e6efd Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 28 May 2017 03:03:46 -0400
Subject: [PATCH 025/104] disable NFS_DEBUG by default

---
 fs/nfs/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/fs/nfs/Kconfig b/fs/nfs/Kconfig
index 14a72224b..080a8027c 100644
--- a/fs/nfs/Kconfig
+++ b/fs/nfs/Kconfig
@@ -195,7 +195,6 @@ config NFS_DEBUG
 	bool
 	depends on NFS_FS && SUNRPC_DEBUG
 	select CRC32
-	default y
 
 config NFS_DISABLE_UDP_SUPPORT
        bool "NFS: Disable NFS UDP protocol support"
-- 
2.36.1.203.g1bcf4f6271

