From 51a61311fb03d31453ddd3e605dac5161a9b0966 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 25 Feb 2018 02:13:48 -0500
Subject: [PATCH 033/104] enable SECURITY by default

---
 security/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/Kconfig b/security/Kconfig
index 6bbe80267..362da3e72 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -23,6 +23,7 @@ config SECURITY
 	bool "Enable different security models"
 	depends on SYSFS
 	depends on MULTIUSER
+	default y
 	help
 	  This allows you to choose different security modules to be
 	  configured into your kernel.
-- 
2.36.1.203.g1bcf4f6271

