From 13348b074135cd499bca99dcdf2a0925f5186e62 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Mon, 29 May 2017 06:17:59 -0400
Subject: [PATCH 034/104] enable SECURITY_YAMA by default

---
 security/yama/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/security/yama/Kconfig b/security/yama/Kconfig
index a81030412..b809050b2 100644
--- a/security/yama/Kconfig
+++ b/security/yama/Kconfig
@@ -2,7 +2,7 @@
 config SECURITY_YAMA
 	bool "Yama support"
 	depends on SECURITY
-	default n
+	default y
 	help
 	  This selects Yama, which extends DAC support with additional
 	  system-wide security settings beyond regular Linux discretionary
-- 
2.36.1.203.g1bcf4f6271

