From 4db05a290cfdfd2970938cfbcd246c70f2f4791f Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 25 Feb 2018 02:14:02 -0500
Subject: [PATCH 035/104] enable SECURITY_NETWORK by default

---
 security/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/Kconfig b/security/Kconfig
index 362da3e72..9eda95bba 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -49,6 +49,7 @@ config SECURITYFS
 config SECURITY_NETWORK
 	bool "Socket and Networking Security Hooks"
 	depends on SECURITY
+	default y
 	help
 	  This enables the socket and networking security hooks.
 	  If enabled, a security module can use these hooks to
-- 
2.36.1.203.g1bcf4f6271

