From 2e733f788a023a32598c18bc464f50b695ce61cf Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:35:53 +0100
Subject: [PATCH 042/104] stop hiding SYSFS_SYSCALL behind EXPERT

---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index e91653cba..408bb7247 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1496,7 +1496,7 @@ config SGETMASK_SYSCALL
 	  If unsure, leave the default option here.
 
 config SYSFS_SYSCALL
-	bool "Sysfs syscall support" if EXPERT
+	bool "Sysfs syscall support"
 	default y
 	help
 	  sys_sysfs is an obsolete system call no longer supported in libc.
-- 
2.36.1.203.g1bcf4f6271

