From c46e9187513154cfd99281319378571de2eab7bc Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:36:54 +0100
Subject: [PATCH 043/103] disable SYSFS_SYSCALL by default

---
 init/Kconfig | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index ef5f61598..68669520c 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1502,13 +1502,12 @@ config SGETMASK_SYSCALL
 
 config SYSFS_SYSCALL
 	bool "Sysfs syscall support"
-	default y
 	help
 	  sys_sysfs is an obsolete system call no longer supported in libc.
 	  Note that disabling this option is more secure but might break
 	  compatibility with some systems.
 
-	  If unsure say Y here.
+	  If unsure say N here.
 
 config FHANDLE
 	bool "open by fhandle syscalls" if EXPERT
-- 
2.36.1.203.g1bcf4f6271

