From ef784e821d1ae27c64b9ebd198f9da80353331c3 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Wed, 25 Aug 2021 22:24:10 +0200
Subject: [PATCH 046/103] kconfig: enable RANDOMIZE_KSTACK_OFFSET_DEFAULT by
 default

---
 arch/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/Kconfig b/arch/Kconfig
index 916a0e5e1..9112af46d 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -1190,6 +1190,7 @@ config RANDOMIZE_KSTACK_OFFSET
 config RANDOMIZE_KSTACK_OFFSET_DEFAULT
 	bool "Default state of kernel stack offset randomization"
 	depends on RANDOMIZE_KSTACK_OFFSET
+	default y
 	help
 	  Kernel stack offset randomization is controlled by kernel boot param
 	  "randomize_kstack_offset=on/off", and this config chooses the default
-- 
2.36.1.203.g1bcf4f6271

