From b19d38cbcc7b74e1b793e1790ad2fa1655af926d Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Sun, 4 Nov 2018 18:48:53 +0100
Subject: [PATCH 096/104] enable protected_{fifos,regular} by default

---
 fs/namei.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/fs/namei.c b/fs/namei.c
index 0b550f909..c21917fd0 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -1022,8 +1022,8 @@ static inline void put_link(struct nameidata *nd)
 
 static int sysctl_protected_symlinks __read_mostly = 1;
 static int sysctl_protected_hardlinks __read_mostly = 1;
-static int sysctl_protected_fifos __read_mostly;
-static int sysctl_protected_regular __read_mostly;
+static int sysctl_protected_fifos __read_mostly = 2;
+static int sysctl_protected_regular __read_mostly = 2;
 
 #ifdef CONFIG_SYSCTL
 static struct ctl_table namei_sysctls[] = {
-- 
2.36.1.203.g1bcf4f6271

