From 72eaab6f879cc67749772fd281fef21296e3c82f Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 19:43:38 -0400
Subject: [PATCH 009/103] enable SLAB_FREELIST_RANDOM by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index b19e2eeaa..835ab4f99 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1953,6 +1953,7 @@ config SLAB_MERGE_DEFAULT
 config SLAB_FREELIST_RANDOM
 	bool "Randomize slab freelist"
 	depends on SLAB || SLUB
+	default y
 	help
 	  Randomizes the freelist order used on creating new pages. This
 	  security feature reduces the predictability of the kernel slab
-- 
2.36.1.203.g1bcf4f6271

