From 7e94953dc7654de315211e91fcfed293a6404004 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 18:15:52 -0400
Subject: [PATCH 017/103] stop hiding MODIFY_LDT_SYSCALL behind EXPERT

---
 arch/x86/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index c5aab5c16..7d96b71ba 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2413,7 +2413,7 @@ config CMDLINE_OVERRIDE
 	  be set to 'N' under normal conditions.
 
 config MODIFY_LDT_SYSCALL
-	bool "Enable the LDT (local descriptor table)" if EXPERT
+	bool "Enable the LDT (local descriptor table)"
 	default y
 	help
 	  Linux can allow user programs to install a per-process x86
-- 
2.36.1.203.g1bcf4f6271

