From 97d2bc743aaf2f099f415444ee3e438868f5ad36 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 7 May 2017 18:28:33 -0400
Subject: [PATCH 030/103] disable COMPAT_BRK by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index f2936fee0..db78a8a72 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1887,7 +1887,6 @@ config SLUB_DEBUG
 
 config COMPAT_BRK
 	bool "Disable heap randomization"
-	default y
 	help
 	  Randomizing heap placement makes heap exploits harder, but it
 	  also breaks ancient binaries (including anything libc5 based).
-- 
2.36.1.203.g1bcf4f6271

