From 93eb246e85e7e84d1549b882be36b08bc39e9522 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Thu, 19 Sep 2019 19:03:01 +0200
Subject: [PATCH 040/103] enable INIT_ON_FREE_DEFAULT_ON by default

---
 security/Kconfig.hardening | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/Kconfig.hardening b/security/Kconfig.hardening
index 49b59ebbb..82e586293 100644
--- a/security/Kconfig.hardening
+++ b/security/Kconfig.hardening
@@ -231,6 +231,7 @@ config INIT_ON_ALLOC_DEFAULT_ON
 
 config INIT_ON_FREE_DEFAULT_ON
 	bool "Enable heap memory zeroing on free by default"
+	default yes
 	help
 	  This has the effect of setting "init_on_free=1" on the kernel
 	  command line. This can be disabled with "init_on_free=0".
-- 
2.36.1.203.g1bcf4f6271

