From 81cc8745e52bae5e45cbe4ddfd3b624d9982bfe8 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 14:45:59 -0400
Subject: [PATCH 065/103] enable SECURITY_PERF_EVENTS_RESTRICT by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 security/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/security/Kconfig b/security/Kconfig
index c2a58216f..eddea71f6 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -22,6 +22,7 @@ config SECURITY_DMESG_RESTRICT
 config SECURITY_PERF_EVENTS_RESTRICT
 	bool "Restrict unprivileged use of performance events"
 	depends on PERF_EVENTS
+	default y
 	help
 	  If you say Y here, the kernel.perf_event_paranoid sysctl
 	  will be set to 3 by default, and no unprivileged use of the
-- 
2.36.1.203.g1bcf4f6271

