From 3ea8b25c49e56441d55a92245dd2b99d1cb5ef90 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 23:36:14 -0400
Subject: [PATCH 094/103] enable SECURITY_TIOCSTI_RESTRICT by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 security/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/security/Kconfig b/security/Kconfig
index 122bff693..537d686a4 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -31,7 +31,7 @@ config SECURITY_PERF_EVENTS_RESTRICT
 
 config SECURITY_TIOCSTI_RESTRICT
 	bool "Restrict unprivileged use of tiocsti command injection"
-	default n
+	default y
 	help
 	  This enforces restrictions on unprivileged users injecting commands
 	  into other processes which share a tty session using the TIOCSTI
-- 
2.36.1.203.g1bcf4f6271

