From b00b0f31fca62915ac5c8d1669c45e0f3dc33730 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 12 Feb 2021 14:45:19 +0100
Subject: [PATCH 6/7] init: enable FUTEX2

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 679c1c4ea..bef11fac0 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1535,7 +1535,7 @@ config FUTEX
 config FUTEX2
 	bool "Enable futex2 support" if EXPERT
 	depends on FUTEX
-	default n
+	default y
 	help
 	  Experimental support for futex2 interface.
 
-- 
2.30.0.452.gfb7fa4a1fd

