From b0b5c4ec2b36657ac4948d8b5aa8ce0874e792d7 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 21 Jun 2019 14:02:03 +0800
Subject: [PATCH 009/106] bmq: Use TASK_ON_RQ_QUEUED for on_rq instead of 1.

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 70e1156ef1b0..4f59df3e332b 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -928,7 +928,7 @@ static void activate_task(struct task_struct *p, struct rq *rq)
 	if (task_contributes_to_load(p))
 		rq->nr_uninterruptible--;
 	enqueue_task(p, rq, ENQUEUE_WAKEUP);
-	p->on_rq = 1;
+	p->on_rq = TASK_ON_RQ_QUEUED;
 	cpufreq_update_util(rq, 0);
 }
 
-- 
2.27.0.rc2

