From 44289928b3ac502ddb971a5cc7dfced7fe52b0da Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 9 Sep 2019 10:32:17 +0800
Subject: [PATCH 027/106] Tag BMQ v1.00

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index c406bfbcfdeb..10c91a6e9552 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.99 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 1.00 by Alfred Chen.\n");
 }
 
 /**
-- 
2.27.0.rc2

