From f0d779b29ada34a19c2aee4fe2e02c20d629b3cc Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 28 Apr 2020 14:41:33 +0800
Subject: [PATCH 098/106] bmq: Set time slice for idle tasks.

---
 kernel/sched/bmq.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 0db417604dda..0f449a8e6d69 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -5545,6 +5545,9 @@ static void sched_init_topology_cpumask(void)
 	cpumask_t *chk;
 
 	for_each_online_cpu(cpu) {
+		/* take chance to reset time slice for idle tasks */
+		cpu_rq(cpu)->idle->time_slice = sched_timeslice_ns;
+
 		chk = &(per_cpu(sched_cpu_affinity_masks, cpu)[0]);
 
 		cpumask_complement(chk, cpumask_of(cpu));
@@ -5581,6 +5584,7 @@ void __init sched_init_smp(void)
 #else
 void __init sched_init_smp(void)
 {
+	cpu_rq(0)->idle->time_slice = sched_timeslice_ns;
 }
 #endif /* CONFIG_SMP */
 
-- 
2.27.0.rc2

