From 59cd4a6110ce319b6939fcf9c9bc275df0e66881 Mon Sep 17 00:00:00 2001
From: torvic9 <torvic9@mailbox.org>
Date: Fri, 2 Apr 2021 16:00:26 +0200
Subject: [PATCH 3/3] allow setting zstd compression level for modules

Signed-off-by: torvic9 <torvic9@mailbox.org>
Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile     | 2 +-
 init/Kconfig | 8 ++++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 0c88ec16f..1e133cfc0 100644
--- a/Makefile
+++ b/Makefile
@@ -1057,7 +1057,7 @@ ifdef CONFIG_MODULE_COMPRESS
     mod_compress_cmd = $(XZ) -f
   endif # CONFIG_MODULE_COMPRESS_XZ
   ifdef CONFIG_MODULE_COMPRESS_ZSTD
-    mod_compress_cmd = $(ZSTD) -T0 --rm -f -q -19 --zstd=wlog=21
+    mod_compress_cmd = $(ZSTD) -T0 -$(CONFIG_MODULE_COMPRESS_ZSTD_LEVEL) --zstd=wlog=21 --rm -f -q
   endif # CONFIG_MODULE_COMPRESS_ZSTD
 endif # CONFIG_MODULE_COMPRESS
 export mod_compress_cmd
diff --git a/init/Kconfig b/init/Kconfig
index d71a08145..c3c79c5be 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2281,6 +2281,14 @@ config MODULE_COMPRESS_ZSTD
 
 endchoice
 
+config MODULE_COMPRESS_ZSTD_LEVEL
+    int "Compression level (1-19)"
+    depends on MODULE_COMPRESS_ZSTD
+    range 1 19
+    default 19
+    help
+        Compression level used by zstd for compressing modules.
+
 config MODULE_ALLOW_MISSING_NAMESPACE_IMPORTS
 	bool "Allow loading of modules with missing namespace imports"
 	help
-- 
2.31.0.97.g1424303384

