From c5259a349b9df0d4be00404360fcd2693c2f85b1 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 5 May 2025 11:06:24 +0800
Subject: [PATCH 57/63] apple-bce: Use timer_delete_sync

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/staging/apple-bce/mailbox.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/staging/apple-bce/mailbox.c b/drivers/staging/apple-bce/mailbox.c
index e24bd3521..af1a59cb6 100644
--- a/drivers/staging/apple-bce/mailbox.c
+++ b/drivers/staging/apple-bce/mailbox.c
@@ -121,7 +121,7 @@ void bce_timestamp_stop(struct bce_timestamp *ts)
     spin_lock_irqsave(&ts->stop_sl, flags);
     ts->stopped = true;
     spin_unlock_irqrestore(&ts->stop_sl, flags);
-    del_timer_sync(&ts->timer);
+    timer_delete_sync(&ts->timer);
 
     iowrite32((u32) -2, regb + 2);
     iowrite32((u32) -1, regb);
@@ -148,4 +148,4 @@ static void bc_send_timestamp(struct timer_list *tl)
         mod_timer(&ts->timer, jiffies + msecs_to_jiffies(150));
     spin_unlock(&ts->stop_sl);
     local_irq_restore(flags);
-}
\ No newline at end of file
+}
-- 
2.49.0

