From 3b006cb076b029c8299012e9f6567b077957cf65 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 31 Dec 2021 10:55:16 +0100
Subject: [PATCH 39/47] bloc, bfq: update comments

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-cgroup.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/block/bfq-cgroup.c b/block/bfq-cgroup.c
index c32319dde..52af755fb 100644
--- a/block/bfq-cgroup.c
+++ b/block/bfq-cgroup.c
@@ -647,7 +647,10 @@ void bfq_bfqq_move(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 	struct bfq_entity *entity = &bfqq->entity;
 	struct bfq_group *old_parent = bfqq_group(bfqq);
 
-	/* No point to move bfqq to the same group */
+	/*
+	* No point to move bfqq to the same group, which can happen when
+	* root group is offlined
+	*/
 	if (old_parent == bfqg)
 		return;
 
-- 
2.34.1.75.gabe6bb3905

