From ed623b15d39f9685fefb0d9fd02063251764ee9f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 15 Jul 2024 14:27:48 +0200
Subject: [PATCH 3/3] bore-6.10: compat mode with sched-ext

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/fair.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 267ff68e0..e2333783b 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -13564,3 +13564,16 @@ __init void init_sched_fair_class(void)
 #endif /* SMP */
 
 }
+
+#ifdef CONFIG_SCHED_BORE
+void reweight_task(struct task_struct *p, int prio)
+{
+	struct sched_entity *se = &p->se;
+	struct cfs_rq *cfs_rq = cfs_rq_of(se);
+	struct load_weight *load = &se->load;
+	unsigned long weight = scale_load(sched_prio_to_weight[prio]);
+
+	reweight_entity(cfs_rq, se, weight);
+	load->inv_weight = sched_prio_to_wmult[prio];
+}
+#endif // CONFIG_SCHED_BORE
-- 
2.45.2.606.g9005149a4a

