From bd6bb25637db345a2a2c133d21dd318e412c59d4 Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Tue, 7 Mar 2023 21:19:49 +0000
Subject: [PATCH 53/59] fixup: mm: create new codetag references during page
 splitting

Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 include/linux/pgalloc_tag.h | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/include/linux/pgalloc_tag.h b/include/linux/pgalloc_tag.h
index d04d7f89d..c2c7b4e1d 100644
--- a/include/linux/pgalloc_tag.h
+++ b/include/linux/pgalloc_tag.h
@@ -61,7 +61,7 @@ static inline void pgalloc_tag_split(struct page *page, unsigned int nr)
 
 	ref = codetag_ref_from_page_ext(page_ext);
 	if (!ref->ct)
-		return;
+		goto out;
 
 	tag = is_codetag_ctx_ref(ref) ? ctc_to_alloc_tag(ref->ctx->ctc)
 				      : ct_to_alloc_tag(ref->ct);
@@ -71,6 +71,7 @@ static inline void pgalloc_tag_split(struct page *page, unsigned int nr)
 		alloc_tag_add(codetag_ref_from_page_ext(page_ext), tag, 0);
 		page_ext = page_ext_next(page_ext);
 	}
+out:
 	page_ext_put(page_ext);
 }
 
-- 
2.40.0.71.g950264636c

