From 489ac990e458b0bd9dd4e539cc13a32f0da650aa Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 13 Mar 2023 15:40:05 -0400
Subject: [PATCH 55/59] fixup! lib: code tagging module support

---
 lib/codetag.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/codetag.c b/lib/codetag.c
index a936d2988..0c7890626 100644
--- a/lib/codetag.c
+++ b/lib/codetag.c
@@ -244,15 +244,20 @@ static inline size_t range_size(const struct codetag_type *cttype,
 static void *get_symbol(struct module *mod, const char *prefix, const char *name)
 {
 	char buf[64];
+	void *ret;
 	int res;
 
 	res = snprintf(buf, sizeof(buf), "%s%s", prefix, name);
 	if (WARN_ON(res < 1 || res > sizeof(buf)))
 		return NULL;
 
-	return mod ?
+	preempt_disable();
+	ret = mod ?
 		(void *)find_kallsyms_symbol_value(mod, buf) :
 		(void *)kallsyms_lookup_name(buf);
+	preempt_enable();
+
+	return ret;
 }
 
 static struct codetag_range get_section_range(struct module *mod,
-- 
2.40.0.71.g950264636c

