From b7886a4245e018aed9cd0b4e59443b15fea1ab54 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 8 Jul 2017 02:38:54 -0400
Subject: [PATCH 11/99] disable SLAB_MERGE_DEFAULT by default

---
 init/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 6d706007e..44c1d4a61 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1932,7 +1932,6 @@ endchoice
 
 config SLAB_MERGE_DEFAULT
 	bool "Allow slab caches to be merged"
-	default y
 	depends on SLAB || SLUB
 	help
 	  For reduced kernel memory fragmentation, slab caches can be
-- 
2.36.1.203.g1bcf4f6271

