From 353e268ea3765d87627a3c49aebeeff396a32423 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Thu, 4 May 2017 18:11:52 -0400
Subject: [PATCH 16/99] disable X86_16BIT by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 arch/x86/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 7b21b2e30..177cb61c8 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1205,7 +1205,6 @@ config VM86
 
 config X86_16BIT
 	bool "Enable support for 16-bit segments"
-	default y
 	depends on MODIFY_LDT_SYSCALL
 	help
 	  This option is required by programs like Wine to run 16-bit
-- 
2.36.1.203.g1bcf4f6271

