From 0c2d9010459439ea55b5bd4802a1ab89d967c0a6 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:28:10 -0400
Subject: [PATCH 23/99] disable DEVPORT by default

---
 drivers/char/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/char/Kconfig b/drivers/char/Kconfig
index 55f48375e..d83f351c8 100644
--- a/drivers/char/Kconfig
+++ b/drivers/char/Kconfig
@@ -347,7 +347,6 @@ config NVRAM
 config DEVPORT
 	bool "/dev/port character device"
 	depends on ISA || PCI
-	default y
 	help
 	  Say Y here if you want to support the /dev/port device. The /dev/port
 	  device is similar to /dev/mem, but for I/O ports.
-- 
2.36.1.203.g1bcf4f6271

