From 2714f4ac5e4b3605514cd7f61e6966e413787be0 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sat, 27 May 2017 07:29:45 -0400
Subject: [PATCH 24/99] disable PROC_VMCORE by default

---
 fs/proc/Kconfig | 1 -
 1 file changed, 1 deletion(-)

diff --git a/fs/proc/Kconfig b/fs/proc/Kconfig
index c93000105..6a0a51b3f 100644
--- a/fs/proc/Kconfig
+++ b/fs/proc/Kconfig
@@ -41,7 +41,6 @@ config PROC_KCORE
 config PROC_VMCORE
 	bool "/proc/vmcore support"
 	depends on PROC_FS && CRASH_DUMP
-	default y
 	help
 	  Exports the dump image of crashed kernel in ELF format.
 
-- 
2.36.1.203.g1bcf4f6271

