From 1554d39c38e15eab55dbd0c00a45a1250d0e6cab Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Sun, 25 Feb 2018 02:15:24 -0500
Subject: [PATCH 36/99] enable AUDIT by default

---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index 506b81a8b..e91653cba 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -447,6 +447,7 @@ config USELIB
 config AUDIT
 	bool "Auditing support"
 	depends on NET
+	default y
 	help
 	  Enable auditing infrastructure that can be used with another
 	  kernel subsystem, such as SELinux (which requires this for
-- 
2.36.1.203.g1bcf4f6271

