From 5d1b9ef0fd53db7dcc0b86914a8b5fb4987a9043 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:36:54 +0100
Subject: [PATCH 43/99] disable SYSFS_SYSCALL by default

---
 init/Kconfig | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index 408bb7247..51cce0ace 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1497,13 +1497,12 @@ config SGETMASK_SYSCALL
 
 config SYSFS_SYSCALL
 	bool "Sysfs syscall support"
-	default y
 	help
 	  sys_sysfs is an obsolete system call no longer supported in libc.
 	  Note that disabling this option is more secure but might break
 	  compatibility with some systems.
 
-	  If unsure say Y here.
+	  If unsure say N here.
 
 config FHANDLE
 	bool "open by fhandle syscalls" if EXPERT
-- 
2.36.1.203.g1bcf4f6271

