From cbb03abbeb77f53744e158fb560b9eea581436b6 Mon Sep 17 00:00:00 2001
From: Levente Polyak <levente@leventepolyak.net>
Date: Tue, 22 Dec 2020 23:40:09 +0100
Subject: [PATCH 44/99] stop hiding UID16 behind EXPERT

---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 51cce0ace..07c49a29a 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1465,7 +1465,7 @@ menuconfig EXPERT
 	  Only use this if you really know what you are doing.
 
 config UID16
-	bool "Enable 16-bit UID system calls" if EXPERT
+	bool "Enable 16-bit UID system calls"
 	depends on HAVE_UID16 && MULTIUSER
 	default y
 	help
-- 
2.36.1.203.g1bcf4f6271

