From e19e450d74168ce49f07bf7bd167c143c5c9600b Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Tue, 11 Jun 2019 20:28:36 +0200
Subject: [PATCH] bmq-5.1: resync with upstream

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 kernel/sched/bmq.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 6ae640f5c72e..8e3983d1b480 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -157,8 +157,7 @@ ____cacheline_aligned_in_smp;
 
 #if (bmq_BITS <= BITS_PER_LONG) && (WM_BITS <= BITS_PER_LONG)
 #define bmq_find_first_bit(bm, size)		((bm[0])? __ffs((bm[0])):(size))
-#define bmq_find_next_bit(bm, size, start)	(\
-{\
+#define bmq_find_next_bit(bm, size, start)	({\
 	unsigned long tmp = (bm[0] & BITMAP_FIRST_WORD_MASK(start));\
 	(tmp)? __ffs(tmp):(size);\
 })
-- 
2.22.0.rc3

