From d886edf4fcc4ed8af632b7f35e943659842adc01 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 19 Nov 2021 17:30:52 +0100
Subject: [PATCH 15/29] LL: Disable TT by default

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index ccbe7a0ed..8696a5dfc 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -112,7 +112,7 @@ config TT_SCHED
 config TT_ACCOUNTING_STATS
 	bool "TT include all accounting and statistics"
 	depends on TT_SCHED
-	default y
+	default n
 	help
 	  This will include all CFS tasks' load accounting and statistics.
 	  If you are using 'performance' governor and do not depend/care
-- 
2.34.1.75.gabe6bb3905

