From b96304ee20f34e6b1e7139dd416d1f39069b8b07 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 2 Dec 2019 15:05:30 +0100
Subject: [PATCH 02/13] Makefile: add -bfq version suffix

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 6a01b073915e..ce9fdfc7231e 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 5
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION = -bfq
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.25.0.191.gde93cc14ab.dirty

