From 9e2f9b0be7dfcad0e4a242d9e7578e5be5b74f5f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 29 Jul 2019 15:57:58 +0200
Subject: [PATCH] bmq: Set yield_type to 2

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 5791e606c..c54c03739 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -65,7 +65,7 @@ static inline void print_scheduler_version(void)
  * 1: Deboost and requeue task. (default)
  * 2: Set rq skip task.
  */
-int sched_yield_type __read_mostly = 1;
+int sched_yield_type __read_mostly = 2;
 
 #define rq_switch_time(rq)	((rq)->clock - (rq)->last_ts_switch)
 #define boost_threshold(p)	(SCHED_TIMESLICE_NS >>\
-- 
2.22.0.545.g9c9b961d7e

