From 340cd5a59aacf17314ec1842fc3f0ff75f2c2212 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 5 Sep 2022 19:28:36 +0200
Subject: [PATCH] cachyos-5.19: mm/swap: Disable swap-in readahead

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 mm/swap.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/mm/swap.c b/mm/swap.c
index 034bb2487..b424f9eea 100644
--- a/mm/swap.c
+++ b/mm/swap.c
@@ -1132,11 +1132,16 @@ void __init swap_setup(void)
 {
 	unsigned long megs = totalram_pages() >> (20 - PAGE_SHIFT);
 
+#ifdef CONFIG_CACHY
+	/* Only swap-in pages requested, avoid readahead */
+	page_cluster = 0
+#else
 	/* Use a smaller cluster for small-memory machines */
 	if (megs < 16)
 		page_cluster = 2;
 	else
 		page_cluster = 3;
+#endif
 	/*
 	 * Right now other parts of the system means that we
 	 * _really_ don't want to cluster much more
-- 
2.37.1.377.g679aad9e82

