From 52a6a501f6f6abfd80209173364a5128d1f01b49 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 2 Feb 2020 09:46:03 +0800
Subject: [PATCH 3/3] Tag BMQ v5.5-r1

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 5b59f735f662..496af2c57092 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -82,7 +82,7 @@ early_param("bmq.timeslice", sched_timeslice);
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.5-r0 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.5-r1 by Alfred Chen.\n");
 }
 
 /**
-- 
2.25.0.114.g5b0ca878e0.dirty

