From e662a9cdb6175e45eba3fbc1cfd24cbf174653d5 Mon Sep 17 00:00:00 2001
From: Alexander Potapenko <glider@google.com>
Date: Wed, 28 Sep 2022 14:23:57 +0200
Subject: [PATCH 48/49] MAINTAINERS: kmsan: add arch/x86/mm/kmsan_shadow.c

---
 MAINTAINERS | 1 +
 1 file changed, 1 insertion(+)

diff --git a/MAINTAINERS b/MAINTAINERS
index 23de81bf5..7aa2109b7 100644
--- a/MAINTAINERS
+++ b/MAINTAINERS
@@ -11382,6 +11382,7 @@ L:	kasan-dev@googlegroups.com
 S:	Maintained
 F:	Documentation/dev-tools/kmsan.rst
 F:	arch/*/include/asm/kmsan.h
+F:	arch/*/mm/kmsan_*
 F:	include/linux/kmsan*.h
 F:	lib/Kconfig.kmsan
 F:	mm/kmsan/
-- 
2.38.0.rc1.6.g4fd6c5e444

