From d08d2e4ae691b1632d5d46008a54eb39af90bf49 Mon Sep 17 00:00:00 2001
From: Alexander Potapenko <glider@google.com>
Date: Wed, 28 Sep 2022 14:24:21 +0200
Subject: [PATCH 49/49] x86: kmsan: fix comment in kmsan_shadow.c

---
 arch/x86/mm/kmsan_shadow.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/mm/kmsan_shadow.c b/arch/x86/mm/kmsan_shadow.c
index 4c8213f81..bee2ec4a3 100644
--- a/arch/x86/mm/kmsan_shadow.c
+++ b/arch/x86/mm/kmsan_shadow.c
@@ -15,6 +15,6 @@
  * handling.
  * arch_kmsan_get_meta_or_null() (declared in the header) maps the addresses in
  * CPU entry area to addresses in cpu_entry_area_shadow/cpu_entry_area_origin.
-*/
+ */
 DEFINE_PER_CPU(char[CPU_ENTRY_AREA_SIZE], cpu_entry_area_shadow);
 DEFINE_PER_CPU(char[CPU_ENTRY_AREA_SIZE], cpu_entry_area_origin);
-- 
2.38.0.rc1.6.g4fd6c5e444

