From 0289a55e1d2c6161b9fcdb06811c46cbea13a959 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@redhat.com>
Date: Thu, 18 Jun 2020 17:34:37 +0200
Subject: [PATCH 2/2] mt76-5.7: pick mt76_for_each_q_rx

Signed-off-by: Oleksandr Natalenko <oleksandr@redhat.com>
---
 drivers/net/wireless/mediatek/mt76/mt76.h | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/net/wireless/mediatek/mt76/mt76.h b/drivers/net/wireless/mediatek/mt76/mt76.h
index 8e4759bc8..778abb203 100644
--- a/drivers/net/wireless/mediatek/mt76/mt76.h
+++ b/drivers/net/wireless/mediatek/mt76/mt76.h
@@ -654,6 +654,10 @@ static inline u16 mt76_rev(struct mt76_dev *dev)
 #define mt76_queue_tx_cleanup(dev, ...)	(dev)->mt76.queue_ops->tx_cleanup(&((dev)->mt76), __VA_ARGS__)
 #define mt76_queue_kick(dev, ...)	(dev)->mt76.queue_ops->kick(&((dev)->mt76), __VA_ARGS__)
 
+#define mt76_for_each_q_rx(dev, i)			\
+	for (i = 0; i < ARRAY_SIZE((dev)->q_rx) &&	\
+			(dev)->q_rx[i].ndesc; i++)
+
 struct mt76_dev *mt76_alloc_device(struct device *pdev, unsigned int size,
 				   const struct ieee80211_ops *ops,
 				   const struct mt76_driver_ops *drv_ops);
-- 
2.27.0.90.geebb51ba8c

