From 08752e5f920f1841a993a244993d555e9239b2f4 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Wed, 5 Apr 2023 14:16:27 +0200
Subject: [PATCH 6/6] zstd-6.2: allocations.h shall import zstd.h properly

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 lib/zstd/common/allocations.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/zstd/common/allocations.h b/lib/zstd/common/allocations.h
index a3153c4ba..630c489e0 100644
--- a/lib/zstd/common/allocations.h
+++ b/lib/zstd/common/allocations.h
@@ -16,7 +16,7 @@
 
 #include "mem.h" /* MEM_STATIC */
 #define ZSTD_STATIC_LINKING_ONLY
-#include "../zstd.h" /* ZSTD_customMem */
+#include <linux/zstd.h> /* ZSTD_customMem */
 
 #ifndef ZSTD_ALLOCATIONS_H
 #define ZSTD_ALLOCATIONS_H
-- 
2.40.0.71.g950264636c

