From aebec264135f9a372c97fc61f4d2bdd5bcdaf068 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 12 Nov 2021 21:51:58 +0100
Subject: [PATCH 03/20] LL: Disable TT by default

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 7eecfea95..f53210480 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -107,7 +107,7 @@ config THREAD_INFO_IN_TASK
 
 config TT_SCHED
 	bool "TT Scheduler"
-	default y
+	default n
 
 
 menu "General setup"
-- 
2.34.1.75.gabe6bb3905

