From 1fd474a9f6cc42d7d2dbab939de351b1604cc91e Mon Sep 17 00:00:00 2001
From: Hamad Al Marri <hamad.s.almarri@gmail.com>
Date: Tue, 7 Dec 2021 16:35:01 +0300
Subject: [PATCH 17/20] GRQ: return smp_processor_id() instead of new_cpu

---
 kernel/sched/bs.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bs.c b/kernel/sched/bs.c
index ded8ea4a5..4c711467e 100644
--- a/kernel/sched/bs.c
+++ b/kernel/sched/bs.c
@@ -1063,7 +1063,7 @@ select_task_rq_fair(struct task_struct *p, int prev_cpu, int wake_flags)
 		}
 
 		if (IS_GRQ_BL_ENABLED)
-			return new_cpu;
+			return smp_processor_id();
 
 		if (cpu_rq(cpu)->nr_running < min) {
 			new_cpu = cpu;
-- 
2.34.1.75.gabe6bb3905

