From 09b520dc5c7fd50c483074f579ec61239843b96a Mon Sep 17 00:00:00 2001
From: Hamad Al Marri <hamad.s.almarri@gmail.com>
Date: Wed, 8 Dec 2021 14:12:30 +0300
Subject: [PATCH 18/20] cleanup

---
 kernel/sched/bs.c | 9 +++------
 1 file changed, 3 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/bs.c b/kernel/sched/bs.c
index 4c711467e..9777d3592 100644
--- a/kernel/sched/bs.c
+++ b/kernel/sched/bs.c
@@ -307,14 +307,11 @@ static void __dequeue_entity_port(struct tt_node **port, struct sched_entity *se
 	struct tt_node *ttn = &se->tt_node;
 
 	// if only one se in rq
-	if ((*port)->next == NULL) {
+	if ((*port)->next == NULL)
 		(*port) = NULL;
-	}
 	// if it is the head
-	else if (ttn == (*port)) {
-		(*port)		= (*port)->next;
-		(*port)->prev	= NULL; // ??
-	}
+	else if (ttn == (*port))
+		(*port) = (*port)->next;
 }
 
 static void __enqueue_entity(struct cfs_rq *cfs_rq, struct sched_entity *se)
-- 
2.34.1.75.gabe6bb3905

