From f5216e8a2e64573cf8f4a614706b15839bf4ca0a Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 23 Jan 2020 10:45:08 +0100
Subject: [PATCH 08/17] block, bfq, DEBUG: check that an entity on a st is not
 freed

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-cgroup.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/block/bfq-cgroup.c b/block/bfq-cgroup.c
index ab80cea47..4741abd0e 100644
--- a/block/bfq-cgroup.c
+++ b/block/bfq-cgroup.c
@@ -329,8 +329,10 @@ static void bfqg_put(struct bfq_group *bfqg)
 	bfqg->ref--;
 
 	BFQ_BUG_ON(bfqg->ref < 0);
-	if (bfqg->ref == 0)
+	if (bfqg->ref == 0) {
+		BFQ_BUG_ON(bfqg->entity.on_st);
 		kfree(bfqg);
+	}
 }
 
 static void bfqg_and_blkg_get(struct bfq_group *bfqg)
-- 
2.25.0.24.gbc7a3d4dc0.dirty

