From 0665ba81203c196b649d325e4ec2491d7ff1be01 Mon Sep 17 00:00:00 2001
From: "Jan Alexander Steffens (heftig)" <jan.steffens@gmail.com>
Date: Fri, 8 Mar 2019 09:22:34 +0100
Subject: [PATCH] bfq: Avoid double definition of bfq_pid_to_str

---
 block/bfq-iosched.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.h b/block/bfq-iosched.h
index 18cc0e996abf..9339891fb40c 100644
--- a/block/bfq-iosched.h
+++ b/block/bfq-iosched.h
@@ -1027,7 +1027,7 @@ void bfq_add_bfqq_busy(struct bfq_data *bfqd, struct bfq_queue *bfqq);
 /* --------------- end of interface of B-WF2Q+ ---------------- */
 
 /* Logging facilities. */
-void bfq_pid_to_str(int pid, char *str, int len)
+static inline void bfq_pid_to_str(int pid, char *str, int len)
 {
 	if (pid != -1)
 		snprintf(str, len, "%d", pid);
-- 
2.21.0.4.g36eb1cb9cf

