From bb4f7e2b81223186a56fc32cfb17552333912863 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 6 Jul 2022 12:53:52 +0200
Subject: [PATCH 10/10] ioprio: fix EXPORT_SYMBOL

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/ioprio.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/ioprio.c b/block/ioprio.c
index 32a456b45..09af41ac6 100644
--- a/block/ioprio.c
+++ b/block/ioprio.c
@@ -162,7 +162,7 @@ int __get_task_ioprio(struct task_struct *p)
 					 task_nice_ioprio(p));
 	return prio;
 }
-EXPORT_SYMBOL_GPL(__get_task_ioprio);
+EXPORT_SYMBOL(__get_task_ioprio);
 
 static int get_task_ioprio(struct task_struct *p)
 {
-- 
2.37.0.3.g30cc8d0f14

