From eadb58cbbced02514d5eb21878a9603de67a9e13 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 30 Jun 2021 23:27:04 +0200
Subject: [PATCH 5/7] kbuild: add proper bool for MODULE_COMPRESS_ZSTD_ULTRA

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 171675c95..cbb72105d 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2271,7 +2271,7 @@ endchoice
 
 config MODULE_COMPRESS_ZSTD_ULTRA
 	depends on MODULE_COMPRESS_ZSTD
-	bool "ZSTD"
+	bool "Enable ZSTD ultra compression"
 	help
 	  Compress modules with ZSTD using the highest possible compression.
 
-- 
2.33.0

