From c29cdc30a8caeed440059e1ac98848905844cc16 Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Thu, 21 Sep 2023 19:05:34 -0700
Subject: [PATCH 10/10] fixup: build fix

Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 mm/huge_memory.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/huge_memory.c b/mm/huge_memory.c
index 48ed14dfb..557cde15c 100644
--- a/mm/huge_memory.c
+++ b/mm/huge_memory.c
@@ -2034,7 +2034,7 @@ int remap_pages_huge_pmd(struct mm_struct *dst_mm, struct mm_struct *src_mm,
 	WRITE_ONCE(src_folio->index, linear_page_index(dst_vma, dst_addr));
 
 	src_pmdval = pmdp_huge_clear_flush(src_vma, src_addr, src_pmd);
-	_dst_pmd = mk_huge_pmd(src_folio->page, dst_vma->vm_page_prot);
+	_dst_pmd = mk_huge_pmd(&src_folio->page, dst_vma->vm_page_prot);
 	_dst_pmd = maybe_pmd_mkwrite(pmd_mkdirty(_dst_pmd), dst_vma);
 	set_pmd_at(dst_mm, dst_addr, dst_pmd, _dst_pmd);
 
-- 
2.42.0

