From c19887e710a2bc7c6d7304145566a0af105a80ed Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Sun, 29 Mar 2026 19:25:13 +0800
Subject: [PATCH 57/66] applesmc-t2: Only load for T2 Macbooks

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/hwmon/applesmc-t2.c | 79 ++++++++++++++++---------------------
 1 file changed, 34 insertions(+), 45 deletions(-)

diff --git a/drivers/hwmon/applesmc-t2.c b/drivers/hwmon/applesmc-t2.c
index 616d2c1e8..3b5d71d71 100644
--- a/drivers/hwmon/applesmc-t2.c
+++ b/drivers/hwmon/applesmc-t2.c
@@ -1133,7 +1133,7 @@ static const struct dev_pm_ops applesmc_pm_ops = {
 };
 
 static struct acpi_driver applesmc_driver = {
-	.name = "applesmc",
+	.name = "applesmc-t2",
 	.class = "applesmc",
 	.ids = applesmc_ids,
 	.ops = {
@@ -1811,51 +1811,40 @@ static void applesmc_release_key_backlight(struct applesmc_device *smc)
 	destroy_workqueue(smc->backlight_wq);
 }
 
-static int applesmc_dmi_match(const struct dmi_system_id *id)
-{
-	return 1;
-}
+#ifndef T2_MAC
+#define T2_MAC(vendor, product) \
+		 .matches = { \
+			DMI_MATCH(DMI_BOARD_VENDOR, vendor), \
+			DMI_MATCH(DMI_PRODUCT_NAME, product), \
+		},
+#endif
 
-/*
- * Note that DMI_MATCH(...,"MacBook") will match "MacBookPro1,1".
- * So we need to put "Apple MacBook Pro" before "Apple MacBook".
- */
-static const struct dmi_system_id applesmc_whitelist[] __initconst = {
-	{ applesmc_dmi_match, "Apple MacBook Air", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "MacBookAir") },
-	},
-	{ applesmc_dmi_match, "Apple MacBook Pro", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "MacBookPro") },
-	},
-	{ applesmc_dmi_match, "Apple MacBook", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "MacBook") },
-	},
-	{ applesmc_dmi_match, "Apple Macmini", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "Macmini") },
-	},
-	{ applesmc_dmi_match, "Apple iMacPro", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "iMacPro") },
-	},
-	{ applesmc_dmi_match, "Apple MacPro", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "MacPro") },
-	},
-	{ applesmc_dmi_match, "Apple iMac", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "iMac") },
-	},
-	{ applesmc_dmi_match, "Apple Xserve", {
-	  DMI_MATCH(DMI_BOARD_VENDOR, "Apple"),
-	  DMI_MATCH(DMI_PRODUCT_NAME, "Xserve") },
-	},
-	{ .ident = NULL }
+static const struct dmi_system_id t2_mac_tbl[] = {
+	{ T2_MAC("Apple Inc.", "MacBookPro15,1") },
+	{ T2_MAC("Apple Inc.", "MacBookPro15,2") },
+	{ T2_MAC("Apple Inc.", "MacBookPro15,3") },
+	{ T2_MAC("Apple Inc.", "MacBookPro15,4") },
+	{ T2_MAC("Apple Inc.", "MacBookPro16,1") },
+	{ T2_MAC("Apple Inc.", "MacBookPro16,2") },
+	{ T2_MAC("Apple Inc.", "MacBookPro16,3") },
+	{ T2_MAC("Apple Inc.", "MacBookPro16,4") },
+	{ T2_MAC("Apple Inc.", "MacBookAir8,1") },
+	{ T2_MAC("Apple Inc.", "MacBookAir8,2") },
+	{ T2_MAC("Apple Inc.", "MacBookAir9,1") },
+	{ T2_MAC("Apple Inc.", "Macmini8,1") },
+	{ T2_MAC("Apple Inc.", "MacPro7,1") },
+	{ T2_MAC("Apple Inc.", "iMac20,1") },
+	{ T2_MAC("Apple Inc.", "iMac20,2") },
+	{ T2_MAC("Apple Inc.", "iMacPro1,1") },
+	{ }
 };
 
+bool apple_is_t2_mac(void)
+{
+	return dmi_check_system(t2_mac_tbl);
+}
+EXPORT_SYMBOL_GPL(apple_is_t2_mac)
+
 static int applesmc_create_modules(struct applesmc_device *smc)
 {
 	int ret;
@@ -1929,7 +1918,7 @@ static int __init applesmc_init(void)
 {
 	int ret;
 
-	if (!dmi_check_system(applesmc_whitelist)) {
+	if (!apple_is_t2_mac()) {
 		pr_warn("supported laptop not found!\n");
 		ret = -ENODEV;
 		goto out;
@@ -1958,4 +1947,4 @@ MODULE_AUTHOR("Nicolas Boichat");
 MODULE_AUTHOR("Paul Pawlowski");
 MODULE_DESCRIPTION("Apple SMC");
 MODULE_LICENSE("GPL v2");
-MODULE_DEVICE_TABLE(dmi, applesmc_whitelist);
+MODULE_DEVICE_TABLE(dmi, t2_mac_tbl);
-- 
2.54.0

