From a76dfd2dc843d78936a379d3f381143f000ab5ba Mon Sep 17 00:00:00 2001
From: Luke Jones <luke@ljones.dev>
Date: Fri, 25 Apr 2025 15:51:57 +1200
Subject: [PATCH 37/52] zone: fix-6.15+: rename del_timer to timer_delete

Signed-off-by: Luke Jones <luke@ljones.dev>
---
 drivers/platform/x86/zotac-zone-platform.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/zotac-zone-platform.c b/drivers/platform/x86/zotac-zone-platform.c
index 5f1123de4..d097f7b9c 100644
--- a/drivers/platform/x86/zotac-zone-platform.c
+++ b/drivers/platform/x86/zotac-zone-platform.c
@@ -359,7 +359,7 @@ static int set_pwm_enable(struct device *dev, u8 mode)
 		data->curve_enabled = false;
 		if (data->curve_enabled) {
 			data->curve_enabled = false;
-			del_timer(&fan_curve_timer);
+			timer_delete(&fan_curve_timer);
 		}
 		break;
 	case PWM_ENABLE_AUTO:
@@ -368,7 +368,7 @@ static int set_pwm_enable(struct device *dev, u8 mode)
 		data->curve_enabled = false;
 		if (data->curve_enabled) {
 			data->curve_enabled = false;
-			del_timer(&fan_curve_timer);
+			timer_delete(&fan_curve_timer);
 		}
 		break;
 	case PWM_ENABLE_CURVE:
@@ -1106,7 +1106,7 @@ static int __init zotac_platform_init(void)
 
 static void __exit zotac_platform_exit(void)
 {
-	del_timer_sync(&fan_curve_timer);
+	timer_delete_sync(&fan_curve_timer);
 
 	platform_driver_unregister(&zotac_platform_driver);
 	platform_device_unregister(zotac_platform_device);
-- 
2.51.2

