From c7c37c1e3dc25df4f19f0952d18f0b3aef3e1c59 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 3 Jan 2019 14:03:13 +0800
Subject: [PATCH 2/5] pds: Optimize pds_skiplist_random_level().

---
 kernel/sched/pds.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 2110c3fcd4d5..c4221b35dbf1 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -824,10 +824,9 @@ static inline int pds_skiplist_random_level(const struct task_struct *p)
 	 * so mask out ~microseconds as a factor of the random seed for skiplist
 	 * insertion.
 	 * 2. Use address of task structure pointer as another factor of the
-	 * random seed for task burst forking scenario. Shift right 9 bits to
-	 * remove the aligned zero bits in the task structure address.
+	 * random seed for task burst forking scenario.
 	 */
-	randseed = (task_rq(p)->clock >> 10) ^ ((long unsigned int)p >> 9);
+	randseed = (task_rq(p)->clock ^ (long unsigned int)p) >> 10;
 
 	return find_first_bit(&randseed, NUM_SKIPLIST_LEVEL - 1);
 }
-- 
2.20.1.98.gecbdaf0899

