From 5a951733dc4f15e791a675a283fece4c33eecb6f Mon Sep 17 00:00:00 2001
From: Luke Jones <luke@ljones.dev>
Date: Wed, 5 Feb 2025 12:37:47 +1300
Subject: [PATCH 46/51] backport: fix fw_attr use

Signed-off-by: Luke Jones <luke@ljones.dev>
---
 drivers/platform/x86/asus-armoury.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/drivers/platform/x86/asus-armoury.c b/drivers/platform/x86/asus-armoury.c
index 5768997ea..a9c18a0ca 100644
--- a/drivers/platform/x86/asus-armoury.c
+++ b/drivers/platform/x86/asus-armoury.c
@@ -94,6 +94,8 @@ struct asus_armoury_priv {
 	struct mutex mutex;
 };
 
+static const struct class *fw_attr_class;
+
 static struct asus_armoury_priv asus_armoury = {
 	.mutex = __MUTEX_INITIALIZER(asus_armoury.mutex)
 };
@@ -812,7 +814,11 @@ static int asus_fw_attr_add(void)
 	const char *name;
 	int err, i;
 
-	asus_armoury.fw_attr_dev = device_create(&firmware_attributes_class, NULL, MKDEV(0, 0),
+	err = fw_attributes_class_get(&fw_attr_class);
+	if (err)
+		return err;
+
+	asus_armoury.fw_attr_dev = device_create(fw_attr_class, NULL, MKDEV(0, 0),
 						NULL, "%s", DRIVER_NAME);
 	if (IS_ERR(asus_armoury.fw_attr_dev)) {
 		err = PTR_ERR(asus_armoury.fw_attr_dev);
@@ -928,8 +934,9 @@ static int asus_fw_attr_add(void)
 err_destroy_kset:
 	kset_unregister(asus_armoury.fw_attr_kset);
 err_destroy_classdev:
+	device_destroy(fw_attr_class, MKDEV(0, 0));
 fail_class_get:
-	device_destroy(&firmware_attributes_class, MKDEV(0, 0));
+		fw_attributes_class_put();
 	return err;
 }
 
@@ -1060,7 +1067,8 @@ static void __exit asus_fw_exit(void)
 
 	sysfs_remove_file(&asus_armoury.fw_attr_kset->kobj, &pending_reboot.attr);
 	kset_unregister(asus_armoury.fw_attr_kset);
-	device_destroy(&firmware_attributes_class, MKDEV(0, 0));
+	device_destroy(fw_attr_class, MKDEV(0, 0));
+	fw_attributes_class_put();
 
 	mutex_unlock(&asus_armoury.mutex);
 }
-- 
2.48.1

