From 6585d5fd470cf18dda8c21b06bc1393ece2b273f Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 5 May 2025 10:11:52 +0800
Subject: [PATCH 56/57] hid: Makefile: Post-patch fixing

---
 drivers/hid/Makefile | 8 +++-----
 1 file changed, 3 insertions(+), 5 deletions(-)

diff --git a/drivers/hid/Makefile b/drivers/hid/Makefile
index 396d1ebdc..a0f584754 100644
--- a/drivers/hid/Makefile
+++ b/drivers/hid/Makefile
@@ -174,12 +174,10 @@ obj-$(CONFIG_AMD_SFH_HID)       += amd-sfh-hid/
 
 obj-$(CONFIG_ASUS_ALLY_HID)  += asus-ally-hid/
 
-obj-$(CONFIG_HID_DOCKCHANNEL)   += dockchannel-hid/
+obj-$(CONFIG_SURFACE_HID_CORE)  += surface-hid/
+
+obj-$(CONFIG_INTEL_THC_HID)     += intel-thc-hid/
 
 obj-$(CONFIG_SPI_HID_APPLE_CORE)	+= spi-hid/
 
 obj-$(CONFIG_HID_DOCKCHANNEL)   += dockchannel-hid/
-
-obj-$(CONFIG_SURFACE_HID_CORE)  += surface-hid/
-
-obj-$(CONFIG_INTEL_THC_HID)     += intel-thc-hid/
-- 
2.49.0

