From 57fe4392c473535e5f38f9df20975163055a148a Mon Sep 17 00:00:00 2001
From: Luke Jones <luke@ljones.dev>
Date: Tue, 11 Feb 2025 13:24:02 +1300
Subject: [PATCH 28/28] hid-asus-ally: set default backface btns to F14/15

Signed-off-by: Luke Jones <luke@ljones.dev>
---
 drivers/hid/hid-asus-ally.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/hid/hid-asus-ally.c b/drivers/hid/hid-asus-ally.c
index 4f1e205f8..6184af19a 100644
--- a/drivers/hid/hid-asus-ally.c
+++ b/drivers/hid/hid-asus-ally.c
@@ -1356,6 +1356,8 @@ static struct ally_gamepad_cfg *ally_gamepad_cfg_create(struct hid_device *hdev)
 	input_dev->uniq = hdev->uniq;
 	input_dev->name = "ASUS ROG Ally Config";
 	input_set_capability(input_dev, EV_KEY, KEY_PROG1);
+	input_set_capability(input_dev, EV_KEY, KEY_F14);
+	input_set_capability(input_dev, EV_KEY, KEY_F15);
 	input_set_capability(input_dev, EV_KEY, KEY_F16);
 	input_set_capability(input_dev, EV_KEY, KEY_F17);
 	input_set_capability(input_dev, EV_KEY, KEY_F18);
@@ -1368,8 +1370,8 @@ static struct ally_gamepad_cfg *ally_gamepad_cfg_create(struct hid_device *hdev)
 
 	/* ignore all errors for this as they are related to USB HID I/O */
 	_gamepad_set_xpad_default(ally_cfg);
-	ally_cfg->key_mapping[ally_cfg->mode - 1].btn_m1.button = BTN_KB_M1;
-	ally_cfg->key_mapping[ally_cfg->mode - 1].btn_m2.button = BTN_KB_M2;
+	ally_cfg->key_mapping[ally_cfg->mode - 1].btn_m1.button = BTN_KB_F15;
+	ally_cfg->key_mapping[ally_cfg->mode - 1].btn_m2.button = BTN_KB_F14;
 	_gamepad_apply_btn_pair(hdev, ally_cfg, btn_pair_m1_m2);
 	gamepad_get_calibration(hdev);
 
@@ -1596,6 +1598,8 @@ static struct input_dev *ally_x_setup_input(struct hid_device *hdev)
 	input_set_capability(input, EV_KEY, BTN_THUMBR);
 
 	input_set_capability(input, EV_KEY, KEY_PROG1);
+	input_set_capability(input, EV_KEY, KEY_F14);
+	input_set_capability(input, EV_KEY, KEY_F15);
 	input_set_capability(input, EV_KEY, KEY_F16);
 	input_set_capability(input, EV_KEY, KEY_F17);
 	input_set_capability(input, EV_KEY, KEY_F18);
-- 
2.48.1

