From 0765cc87c3cf52559690b18c437b386ffacdf040 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Thu, 14 Jan 2021 10:28:03 +0100
Subject: [PATCH 12/56] block, bfq, DEBUG: print weights on idle-busy switch

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index dfb253358..85224d40b 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -1812,6 +1812,10 @@ static void bfq_bfqq_handle_idle_busy_switch(struct bfq_data *bfqd,
 		     bfq_bfqq_non_blocking_wait_rq(bfqq),
 		     old_wr_coeff);
 
+	bfq_log_bfqq(bfqd, bfqq,
+		     "new_weight %d weight %d",
+		     bfqq->entity.new_weight, bfqq->entity.weight);
+
 	BFQ_BUG_ON(bfqq->entity.budget < bfqq->entity.service);
 
 	BFQ_BUG_ON(bfqq == bfqd->in_service_queue);
-- 
2.30.0.155.g66e871b664

