From 658e829a7a8880597d59fc32dea854317290b38f Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Tue, 16 Dec 2025 01:27:50 +0800
Subject: [PATCH 10/10] x86/vmscape: Depend on HAVE_STATIC_CALL instead of
 selecting it

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 arch/x86/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 1e92e530f..e03e35a2a 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -2710,7 +2710,7 @@ config MITIGATION_TSA
 config MITIGATION_VMSCAPE
 	bool "Mitigate VMSCAPE"
 	depends on KVM
-	select HAVE_STATIC_CALL
+	depends on HAVE_STATIC_CALL
 	default y
 	help
 	  Enable mitigation for VMSCAPE attacks. VMSCAPE is a hardware security
-- 
2.52.0

