From adae5f252da47dbff87a0bf361e64ac109f08bb2 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 6 Dec 2019 16:50:39 +0100
Subject: [PATCH 6/6] block: Set bfq-dev instead of bfq

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index d8c426abb267..f9ca0ca78b7f 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7631,7 +7631,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ I/O-scheduler: v11";
+	char msg[60] = "BFQ-dev I/O-scheduler: v11";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.25.0.rc0.1.g0a76bd7381

