From 4b2b008693b000406eaf1495d4964119302ed479 Mon Sep 17 00:00:00 2001
From: Username404-59 <w.iron.zombie@gmail.com>
Date: Thu, 12 Dec 2024 21:25:40 +0100
Subject: [PATCH 2/4] kbuild: Update old polly option

Signed-off-by: Username404-59 <w.iron.zombie@gmail.com>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 1026a3571..6aed3c33c 100644
--- a/Makefile
+++ b/Makefile
@@ -906,7 +906,7 @@ ifdef CONFIG_POLLY_CLANG
 KBUILD_CFLAGS	+= -mllvm -polly \
 		   -mllvm -polly-ast-use-context \
 		   -mllvm -polly-invariant-load-hoisting \
-		   -mllvm -polly-opt-fusion=max \
+		   -mllvm -polly-loopfusion-greedy \
 		   -mllvm -polly-run-inliner \
 		   -mllvm -polly-vectorizer=stripmine
 # Polly may optimise loops with dead paths beyound what the linker
-- 
2.53.0

