From 9f610e2785fe841c6283874685019038bc4b1d7c Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 21 Sep 2020 17:26:14 +0200
Subject: [PATCH 2/2] init, Kconfig: Restore original PDS description

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c9b17ae25..6c4a1f3c1 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -887,7 +887,9 @@ config SCHED_PDS
 	bool "PDS CPU scheduler"
 	help
 	  The Priority and Deadline based Skip list multiple queue CPU
-	  Scheduler.
+	  scheduler for excellent interactivity and responsiveness on the
+	  desktop and solid scalability on normal hardware and commodity
+	  servers.
 
 endchoice
 
-- 
2.35.1.354.g715d08a9e5

