From 870e0b275c681391ba069f0a010898e472363f57 Mon Sep 17 00:00:00 2001
From: Tejun Heo <tj@kernel.org>
Date: Sat, 25 Nov 2023 10:01:32 -1000
Subject: [PATCH 38/67] scx: Fix a straggling atomic64_set

---
 kernel/sched/core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 5cc759ce2..dc042671f 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -4561,7 +4561,7 @@ static void __sched_fork(unsigned long clone_flags, struct task_struct *p)
 	p->scx.sticky_cpu	= -1;
 	p->scx.holding_cpu	= -1;
 	p->scx.kf_mask		= 0;
-	atomic64_set(&p->scx.ops_state, 0);
+	atomic_long_set(&p->scx.ops_state, 0);
 	p->scx.runnable_at	= INITIAL_JIFFIES;
 	p->scx.slice		= SCX_SLICE_DFL;
 #endif
-- 
2.43.0.rc2

