From 7b3fb4bea7029f582b7b1d7bca292a0e3d2f2111 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 27 May 2024 16:39:09 +0200
Subject: [PATCH 2/2] media: v4l2-core: bump v4l2loopback driver version to
 0.13.2

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/media/v4l2-core/v4l2loopback.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/media/v4l2-core/v4l2loopback.h b/drivers/media/v4l2-core/v4l2loopback.h
index 1bc7e6b74..48b9d9883 100644
--- a/drivers/media/v4l2-core/v4l2loopback.h
+++ b/drivers/media/v4l2-core/v4l2loopback.h
@@ -12,7 +12,7 @@
 
 #define V4L2LOOPBACK_VERSION_MAJOR 0
 #define V4L2LOOPBACK_VERSION_MINOR 13
-#define V4L2LOOPBACK_VERSION_BUGFIX 1
+#define V4L2LOOPBACK_VERSION_BUGFIX 2
 
 /* /dev/v4l2loopback interface */
 
-- 
2.45.1.145.g83f1add914

