From 5086c54925799176ad4b21e7193decc105af2fd1 Mon Sep 17 00:00:00 2001
From: Antheas Kapenekakis <lkml@antheas.dev>
Date: Wed, 10 Sep 2025 17:45:42 +0200
Subject: [PATCH 090/104] add suiplay

Signed-off-by: Antheas Kapenekakis <lkml@antheas.dev>
---
 drivers/platform/x86/ayaneo-ec.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/platform/x86/ayaneo-ec.c b/drivers/platform/x86/ayaneo-ec.c
index 41a24e091..28cb09fe8 100644
--- a/drivers/platform/x86/ayaneo-ec.c
+++ b/drivers/platform/x86/ayaneo-ec.c
@@ -139,6 +139,13 @@ static const struct dmi_system_id dmi_table[] = {
 		},
 		.driver_data = (void *)&quirk_ayaneo3,
 	},
+	{
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Mysten Labs, Inc."),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "SuiPlay0X1"),
+		},
+		.driver_data = (void *)&quirk_fan,
+	},
 	{},
 };
 
-- 
2.52.0

