From f65bac920953f4bc959716a348d5ebb91aaa0c87 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 29 Sep 2018 15:09:55 +0800
Subject: [PATCH 07/70] pds: [Sync] bbb62c0b024a sched/core: Remove the rt_avg
 code

---
 kernel/sched/pds.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 0925a5ac8840..02b59d6a4b19 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -6117,10 +6117,6 @@ static void sched_rq_cpu_starting(unsigned int cpu)
 
 int sched_cpu_starting(unsigned int cpu)
 {
-	/*
-	 * PDS doesn't have rq start time record
-	 * set_cpu_rq_start_time(cpu);
-	 */
 	sched_rq_cpu_starting(cpu);
 	sched_tick_start(cpu);
 	return 0;
-- 
2.21.0

