From 5f3598b5e8bbc93b4bdb69cd4fb445e5b4d3ddf0 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 22 Oct 2018 13:58:51 +0000
Subject: [PATCH 10/70] Tag PDS 0.99b

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 0a3daab9e55a..ce0f7682f0ef 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -90,7 +90,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99a by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99b by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.21.0

