From ca7948d77408dc9510fb60dcf77f3af67d33686b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 12 Nov 2018 10:20:18 +0000
Subject: [PATCH 19/70] Tag PDS 0.99d

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index a2ff24905914..12b15a0a2d55 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99c by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99d by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.21.0

