From 9bd5206ec10b97f3ed14c230c0faf8ab5c522ae1 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 7 Nov 2018 15:46:34 +0000
Subject: [PATCH 22/70] pds: Remove cpumask_weight() in best_mask_cpu().

---
 kernel/sched/pds.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 07d8fddd7257..0c52506cb3d0 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -1561,9 +1561,6 @@ static inline int best_mask_cpu(const int cpu, cpumask_t *cpumask)
 {
 	cpumask_t tmp, *mask;
 
-	if (cpumask_weight(cpumask) == 1)
-		return cpumask_first(cpumask);
-
 	if (cpumask_test_cpu(cpu, cpumask))
 		return cpu;
 
-- 
2.21.0

