From 64248cc2392ad3f934d27aeb9be5c251a7c471d7 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 30 Nov 2018 16:39:37 +0800
Subject: [PATCH 27/70] Tag PDS 0.99f

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 2ea20d62191a..0421a050d61c 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99e by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99f by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.21.0

