From 8f79ea509e516f422f67f3eb6119bd055a3b27c0 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 9 Dec 2018 17:41:18 +0800
Subject: [PATCH 30/70] Tag PDS 0.99g

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 5a5ad448f23c..3066135a5430 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99f by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99g by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.21.0

