From ee25e7a36c720d8d02608a6d6164143daa08400c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 18 Dec 2018 14:40:36 +0800
Subject: [PATCH 31/70] pds: Fix cpu hot-plug Oops.

---
 kernel/sched/pds.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 3066135a5430..5c0c088d6d62 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2875,6 +2875,10 @@ static inline void pds_sg_balance_check(const struct rq *rq)
 	cpumask_t chk;
 	int i;
 
+	/* Only online cpu will do sg balance checking */
+	if (unlikely(!rq->online))
+		return;
+
 	/* Only cpu in slibing idle group will do the checking */
 	if (!cpumask_test_cpu(cpu_of(rq), &sched_cpu_sg_idle_mask))
 		return;
-- 
2.21.0

