From 71b46af5472434a8c3e562f0a4cff3b4c258f066 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 19 Dec 2018 18:10:07 +0800
Subject: [PATCH 34/70] pds: Fix NORMAL_PRIO mess up with ISO_PRIO.

---
 include/linux/sched/prio.h | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/include/linux/sched/prio.h b/include/linux/sched/prio.h
index 162c2e649d0f..fba04bb91492 100644
--- a/include/linux/sched/prio.h
+++ b/include/linux/sched/prio.h
@@ -26,8 +26,8 @@
 
 #define MAX_RT_PRIO		((MAX_USER_RT_PRIO) + 1)
 
-#define NORMAL_PRIO		(MAX_USER_RT_PRIO)
-#define IDLE_PRIO		((MAX_USER_RT_PRIO) + 1)
+#define NORMAL_PRIO		(MAX_RT_PRIO)
+#define IDLE_PRIO		((MAX_RT_PRIO) + 1)
 #define PRIO_LIMIT		((IDLE_PRIO) + 1)
 #else /* !CONFIG_SCHED_PDS */
 #define MAX_RT_PRIO		MAX_USER_RT_PRIO
-- 
2.21.0

