From 966abf112db955109f6f1d357d1e622275b10645 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 14 Dec 2018 15:44:25 +0800
Subject: [PATCH 46/70] pds: Optimize best_mask_cpu().

---
 kernel/sched/pds.c | 10 ++++------
 1 file changed, 4 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index a2aabc939da4..b1ba09917ff7 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -1622,13 +1622,11 @@ static inline int best_mask_cpu(int cpu, cpumask_t *cpumask)
 	if (cpumask_test_cpu(cpu, cpumask))
 		return cpu;
 
-	for (mask = &(per_cpu(sched_cpu_affinity_chk_masks, cpu)[0]);
-	     mask < per_cpu(sched_cpu_affinity_chk_end_masks, cpu); mask++)
-		if ((cpu = cpumask_any_and(cpumask, mask)) < nr_cpu_ids)
-			return cpu;
+	mask = &(per_cpu(sched_cpu_affinity_chk_masks, cpu)[0]);
+	while ((cpu = cpumask_any_and(cpumask, mask)) >= nr_cpu_ids)
+		mask++;
 
-	/* Safe fallback, should never come here */
-	return cpumask_first(cpumask);
+	return cpu;
 }
 
 /*
-- 
2.21.0

