From 52b11730bbc0367d2eee1351d8206c71c8fe0b30 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 27 Dec 2018 11:22:27 +0800
Subject: [PATCH 49/70] pds: Fix fail to build on ppc64 arch.

---
 kernel/sched/pds.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 4f1b27baad10..dd6bce14a850 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -6012,6 +6012,7 @@ static void sched_init_topology_cpumask(void)
 			       cpu, (chk++)->bits[0]);
 		}
 #endif
+#ifdef CONFIG_SCHED_MC
 		cpumask_setall(chk);
 		cpumask_clear_cpu(cpu, chk);
 		if (cpumask_and(chk, chk, cpu_coregroup_mask(cpu))) {
@@ -6026,7 +6027,15 @@ static void sched_init_topology_cpumask(void)
 		 */
 		per_cpu(sd_llc_id, cpu) =
 			cpumask_first(cpu_coregroup_mask(cpu));
+#else
+		per_cpu(sd_llc_id, cpu) =
+			cpumask_first(topology_core_cpumask(cpu));
 
+		per_cpu(sched_cpu_llc_start_mask, cpu) = chk;
+
+		cpumask_setall(chk);
+		cpumask_clear_cpu(cpu, chk);
+#endif /* NOT CONFIG_SCHED_MC */
 		if (cpumask_and(chk, chk, topology_core_cpumask(cpu)))
 			printk(KERN_INFO "pds: cpu #%d affinity check mask - core 0x%08lx",
 			       cpu, (chk++)->bits[0]);
-- 
2.21.0

