From 3fd1de11cb880df418adde442ff0da369680f05e Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 13 Feb 2019 15:09:07 +0800
Subject: [PATCH 59/70] Tag PDS 0.99m

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 4fe8e20fd4c0..23db1e4be7a6 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99l by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99m by Alfred Chen.\n");
 }
 
 /*
-- 
2.21.0

