From fafbc63862fcbdc2aabf8b151c2596b806023a9b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 21 Feb 2019 13:52:39 +0800
Subject: [PATCH 60/70] pds: Fix hrtick_start() too short for new tasks.

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 23db1e4be7a6..f5a8cb0b8fad 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2186,7 +2186,7 @@ int sched_fork(unsigned long __maybe_unused clone_flags, struct task_struct *p)
 	rq->curr->time_slice /= 2;
 	p->time_slice = rq->curr->time_slice;
 #ifdef CONFIG_SCHED_HRTICK
-	hrtick_start(rq, rq->curr->time_slice);
+	hrtick_start(rq, US_TO_NS(rq->curr->time_slice));
 #endif
 
 	if (p->time_slice < RESCHED_US) {
-- 
2.21.0

