From 7cb56757d0c84e7992ac2d6a7a6fee565c5211b6 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 1 Mar 2019 16:53:42 +0800
Subject: [PATCH 61/70] Tag PDS 0.99n

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index f5a8cb0b8fad..20fb77b2ef17 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99m by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99n by Alfred Chen.\n");
 }
 
 /*
-- 
2.21.0

