From 1485eb0d065e7e15e6f7aa0379334524cfcb1012 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 19 Feb 2019 10:06:21 +0800
Subject: [PATCH 65/70] pds: [Sync] dfcb245e2848 sched: Fix various typos in
 comments

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 4f59eb7b8114..98b0e50243c3 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2701,7 +2701,7 @@ unsigned long nr_running(void)
  * preemption, thus the result might have a time-of-check-to-time-of-use
  * race.  The caller is responsible to use it correctly, for example:
  *
- * - from a non-preemptable section (of course)
+ * - from a non-preemptible section (of course)
  *
  * - from a thread that is bound to a single CPU
  *
-- 
2.21.0

