From 3d3a303d2eb75a5d958e1ccc808a6a63611ee721 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 19 Feb 2019 10:11:02 +0800
Subject: [PATCH 67/70] pds: [Sync] e6018c0f5c99 sched/wake_q: Document
 wake_q_add()

---
 kernel/sched/pds.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 902c4d858a58..1fb420dd49f3 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -5445,6 +5445,18 @@ void resched_cpu(int cpu)
 	raw_spin_unlock_irqrestore(&rq->lock, flags);
 }
 
+/**
+ * wake_q_add() - queue a wakeup for 'later' waking.
+ * @head: the wake_q_head to add @task to
+ * @task: the task to queue for 'later' wakeup
+ *
+ * Queue a task for later wakeup, most likely by the wake_up_q() call in the
+ * same context, _HOWEVER_ this is not guaranteed, the wakeup can come
+ * instantly.
+ *
+ * This function must be used as-if it were wake_up_process(); IOW the task
+ * must be ready to be woken at this location.
+ */
 void wake_q_add(struct wake_q_head *head, struct task_struct *task)
 {
 	struct wake_q_node *node = &task->wake_q;
-- 
2.21.0

