From 4f25989abfba7f9674281d7dc01b042d56762253 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 1 Mar 2019 17:02:27 +0800
Subject: [PATCH 70/70] Tag PDS 0.99o

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 21d15416604d..cc2512eafb79 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99n by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99o by Alfred Chen.\n");
 }
 
 /*
-- 
2.21.0

