From 5bcfac853dd79e600822d0ffcc87dde3ff8a0f6f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Sat, 10 Nov 2018 11:21:49 +0100
Subject: [PATCH 5/5] LL: Add .ll version

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/Makefile b/Makefile
index d4d36c619..05d2a9540 100644
--- a/Makefile
+++ b/Makefile
@@ -11,6 +11,10 @@ NAME = Kleptomaniac Octopus
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLNAME = linux-lucjan
+LLVERSION = .ll55
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
-- 
2.24.0.155.gd9f6f3b619

