From deb98049d27e7a7d32ad9a69e50dbe68754e4ce0 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 11 Aug 2020 21:58:03 +0800
Subject: [PATCH 030/269] sched/alt: Fix UP compilation issue.

---
 kernel/sched/alt_core.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 2f67a87b4b35..5580b600db67 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -141,7 +141,9 @@ DEFINE_PER_CPU_SHARED_ALIGNED(struct rq, runqueues);
 
 #define IDLE_WM	(IDLE_TASK_SCHED_PRIO)
 
+#ifdef CONFIG_SCHED_SMT
 static cpumask_t sched_sg_idle_mask ____cacheline_aligned_in_smp;
+#endif
 static cpumask_t sched_rq_watermark[SCHED_BITS] ____cacheline_aligned_in_smp;
 
 static inline void update_sched_rq_watermark(struct rq *rq)
@@ -5839,12 +5841,12 @@ void __init sched_init(void)
 #ifdef CONFIG_SCHED_SMT
 		rq->active_balance = 0;
 #endif
-#endif
-		rq->nr_switches = 0;
 
 #ifdef CONFIG_NO_HZ_COMMON
 		rq_csd_init(rq, &rq->nohz_csd, nohz_csd_func);
 #endif
+#endif /* CONFIG_SMP */
+		rq->nr_switches = 0;
 
 		hrtick_rq_init(rq);
 		atomic_set(&rq->nr_iowait, 0);
-- 
2.35.1.677.gabf474a5dd

