From df5eb2d34c8166aff8541e7a22fb2056a5fd1c1d Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 5 May 2025 10:05:06 +0800
Subject: [PATCH 01/62] hid: Makefile: Add asus-hid-ally to resolve conflict

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/hid/Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/hid/Makefile b/drivers/hid/Makefile
index 10ae5dedb..85af5331e 100644
--- a/drivers/hid/Makefile
+++ b/drivers/hid/Makefile
@@ -172,6 +172,8 @@ obj-$(CONFIG_INTEL_ISH_HID)	+= intel-ish-hid/
 
 obj-$(CONFIG_AMD_SFH_HID)       += amd-sfh-hid/
 
+obj-$(CONFIG_ASUS_ALLY_HID)  += asus-ally-hid/
+
 obj-$(CONFIG_SURFACE_HID_CORE)  += surface-hid/
 
 obj-$(CONFIG_INTEL_THC_HID)     += intel-thc-hid/
-- 
2.49.0

