From c9bcfabd5f64bff53f25107cbd876673bd8632f8 Mon Sep 17 00:00:00 2001
From: Eric Naim <dnaim@cachyos.org>
Date: Mon, 12 Jan 2026 13:33:47 +0700
Subject: [PATCH 37/37] hid: Makefile: Post-patch fixing

Signed-off-by: Eric Naim <dnaim@cachyos.org>
---
 drivers/hid/Makefile | 2 --
 1 file changed, 2 deletions(-)

diff --git a/drivers/hid/Makefile b/drivers/hid/Makefile
index 2ed494282..397b6cf5b 100644
--- a/drivers/hid/Makefile
+++ b/drivers/hid/Makefile
@@ -173,8 +173,6 @@ obj-$(CONFIG_INTEL_ISH_HID)	+= intel-ish-hid/
 
 obj-$(CONFIG_AMD_SFH_HID)       += amd-sfh-hid/
 
-obj-$(CONFIG_HID_DOCKCHANNEL)   += dockchannel-hid/
-
 obj-$(CONFIG_SPI_HID_APPLE_CORE)	+= spi-hid/
 
 obj-$(CONFIG_HID_DOCKCHANNEL)   += dockchannel-hid/
-- 
2.52.0

