From b9b109a9d63ce8cb574fb2de1cd46c413c3a49a5 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 19 Jan 2019 20:47:17 +0800
Subject: [PATCH 1/2] pds: Fix wrong task priority issue.

---
 kernel/sched/pds.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 5030a377d60e..96dd5be7981e 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -4233,6 +4233,7 @@ static void __setscheduler(struct rq *rq, struct task_struct *p,
 	 * Keep a potential priority boosting if called from
 	 * sched_setscheduler().
 	 */
+	p->prio = normal_prio(p);
 	if (keep_boost)
 		p->prio = rt_effective_prio(p, p->prio);
 	update_task_priodl(p);
-- 
2.20.1.142.g77556354bb

