From 68def16f78de91e42db8885427139ccc18ac8e35 Mon Sep 17 00:00:00 2001
From: Alexandre Frade <kernel@xanmod.org>
Date: Sun, 29 May 2022 00:57:40 +0000
Subject: [PATCH 08/12] XANMOD: scripts/setlocalversion: remove "+" tag for git
 repo short version

Signed-off-by: Alexandre Frade <kernel@xanmod.org>
---
 scripts/setlocalversion | 1 -
 1 file changed, 1 deletion(-)

diff --git a/scripts/setlocalversion b/scripts/setlocalversion
index af4754a35..3edda03c2 100755
--- a/scripts/setlocalversion
+++ b/scripts/setlocalversion
@@ -54,7 +54,6 @@ scm_version()
 			# If only the short version is requested, don't bother
 			# running further git commands
 			if $short; then
-				echo "+"
 				return
 			fi
 			# If we are past a tagged commit (like
-- 
2.38.1.385.g3b08839926

