From 0c0c7e1c5fc409b96e374db3cd3f56277fcbfe89 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Thu, 2 Feb 2023 12:37:50 -0500
Subject: [PATCH 53/55] alloc tagging: alloc_tag_sub_noalloc() (squash)

---
 include/linux/alloc_tag.h | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/include/linux/alloc_tag.h b/include/linux/alloc_tag.h
index 157984a21..ce4a618ab 100644
--- a/include/linux/alloc_tag.h
+++ b/include/linux/alloc_tag.h
@@ -72,7 +72,8 @@ static inline void set_codetag_empty(union codetag_ref *ref) {}
 #endif /* CONFIG_ALLOC_TAGGING_DEBUG */
 
 
-static inline void alloc_tag_sub(union codetag_ref *ref, size_t bytes)
+static inline void __alloc_tag_sub(union codetag_ref *ref, size_t bytes,
+				   bool may_allocate)
 {
 	struct alloc_tag *tag;
 
@@ -94,10 +95,24 @@ static inline void alloc_tag_sub(union codetag_ref *ref, size_t bytes)
 		alloc_tag_free_ctx(ref->ctx, &tag);
 	else
 		tag = ct_to_alloc_tag(ref->ct);
-	lazy_percpu_counter_add(&tag->bytes_allocated, -bytes);
+
+	if (may_allocate)
+		lazy_percpu_counter_add(&tag->bytes_allocated, -bytes);
+	else
+		lazy_percpu_counter_add_noupgrade(&tag->bytes_allocated, -bytes);
 	ref->ct = NULL;
 }
 
+static inline void alloc_tag_sub(union codetag_ref *ref, size_t bytes)
+{
+	__alloc_tag_sub(ref, bytes, true);
+}
+
+static inline void alloc_tag_sub_noalloc(union codetag_ref *ref, size_t bytes)
+{
+	__alloc_tag_sub(ref, bytes, false);
+}
+
 static inline void alloc_tag_add(union codetag_ref *ref, struct alloc_tag *tag, size_t bytes)
 {
 	if (!alloc_tagging_enabled())
@@ -125,6 +140,7 @@ static inline void alloc_tag_add(union codetag_ref *ref, struct alloc_tag *tag,
 #define DEFINE_ALLOC_TAG(_alloc_tag, _old)
 static inline void set_codetag_empty(union codetag_ref *ref) {}
 static inline void alloc_tag_sub(union codetag_ref *ref, size_t bytes) {}
+static inline void alloc_tag_sub_noalloc(union codetag_ref *ref, size_t bytes) {}
 static inline void alloc_tag_add(union codetag_ref *ref, struct alloc_tag *tag,
 				 size_t bytes) {}
 
-- 
2.39.0.rc2.1.gbd5df96b79

