From 1e67ab9332f7c8afcd66f998779f9443fd5222ae Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Fri, 20 Jan 2023 11:17:56 -0500
Subject: [PATCH 54/55] fixup! mm: percpu: Alloc tagging

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 mm/percpu.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/percpu.c b/mm/percpu.c
index 377870ff2..f9ec3b0a5 100644
--- a/mm/percpu.c
+++ b/mm/percpu.c
@@ -1713,7 +1713,7 @@ static void pcpu_alloc_tag_alloc_hook(struct pcpu_chunk *chunk, int off,
 static void pcpu_alloc_tag_free_hook(struct pcpu_chunk *chunk, int off, size_t size)
 {
 	if (likely(chunk->obj_exts))
-		alloc_tag_sub(&chunk->obj_exts[off >> PCPU_MIN_ALLOC_SHIFT].tag, size);
+		alloc_tag_sub_noalloc(&chunk->obj_exts[off >> PCPU_MIN_ALLOC_SHIFT].tag, size);
 }
 #else
 static void pcpu_alloc_tag_alloc_hook(struct pcpu_chunk *chunk, int off,
-- 
2.39.0.rc2.1.gbd5df96b79

