From 269e2dc961c84913dec956491b5494ea69d4fc65 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Thu, 2 Feb 2023 13:31:30 -0500
Subject: [PATCH 55/55] fixup! lib: introduce page allocation tagging

---
 include/linux/pgalloc_tag.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/linux/pgalloc_tag.h b/include/linux/pgalloc_tag.h
index d1bcdb903..d24741049 100644
--- a/include/linux/pgalloc_tag.h
+++ b/include/linux/pgalloc_tag.h
@@ -60,6 +60,9 @@ static inline void pgalloc_tag_split(struct page *page, unsigned int nr)
 		return;
 
 	ref = codetag_ref_from_page_ext(page_ext);
+	if (!ref->ct)
+		return;
+
 	tag = is_codetag_ctx_ref(ref) ? ctc_to_alloc_tag(ref->ctx->ctc)
 				      : ct_to_alloc_tag(ref->ct);
 	page_ext = page_ext_next(page_ext);
-- 
2.39.0.rc2.1.gbd5df96b79

