From 430fd179940600fd49ab29cd14b3334fca49fa78 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 16 May 2022 17:13:59 +0200
Subject: [PATCH 11/28] give rdrand some credit

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 drivers/char/random.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/char/random.c b/drivers/char/random.c
index 4c9adb4f3..5926c9e11 100644
--- a/drivers/char/random.c
+++ b/drivers/char/random.c
@@ -994,6 +994,8 @@ int __init rand_initialize(void)
 		    !arch_get_random_long_early(&rv)) {
 			rv = random_get_entropy();
 			arch_init = false;
+		else
+			credit_entropy_bits(1);
 		}
 		_mix_pool_bytes(&rv, sizeof(rv));
 	}
-- 
2.36.1.74.g277cf0bc36

