From 6e9bd47b5c03ee302ed426f82a58a1f036f05665 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 8 May 2023 01:18:58 +0200
Subject: [PATCH 02/28] ddcci-6.4: adopt 48380368de

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/char/ddcci.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/char/ddcci.c b/drivers/char/ddcci.c
index 4d71d26ad..129aede43 100644
--- a/drivers/char/ddcci.c
+++ b/drivers/char/ddcci.c
@@ -39,7 +39,7 @@ static int autoprobe_addr_count = 5;
 static dev_t ddcci_cdev_first;
 static dev_t ddcci_cdev_next;
 static dev_t ddcci_cdev_end;
-static DEFINE_SEMAPHORE(core_lock);
+static DEFINE_SEMAPHORE(core_lock, 1);
 
 struct bus_type ddcci_bus_type;
 EXPORT_SYMBOL_GPL(ddcci_bus_type);
-- 
2.41.0.159.g0bfa463d37

