From 953008e0a4c2e6543e92114f0878801ee13be90c Mon Sep 17 00:00:00 2001
From: Tejun Heo <tj@kernel.org>
Date: Tue, 5 Dec 2023 14:12:51 -1000
Subject: [PATCH 55/93] scx: Remove now unused $rb_node from first_local_task()

---
 kernel/sched/ext.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/kernel/sched/ext.c b/kernel/sched/ext.c
index 1095d494c..53ee906aa 100644
--- a/kernel/sched/ext.c
+++ b/kernel/sched/ext.c
@@ -1779,8 +1779,6 @@ static void put_prev_task_scx(struct rq *rq, struct task_struct *p)
 
 static struct task_struct *first_local_task(struct rq *rq)
 {
-	struct rb_node *rb_node;
-
 	WARN_ON_ONCE(rb_first_cached(&rq->scx.local_dsq.priq));
 	return list_first_entry_or_null(&rq->scx.local_dsq.fifo,
 					struct task_struct, scx.dsq_node.fifo);
-- 
2.43.0.232.ge79552d197

