From 2a467f297cadd92de08bfe714b8e58855e4a7aef Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Thu, 19 May 2022 10:47:50 +0200
Subject: [PATCH 10/10] sched/alt: [Tentative] Add lockdep to select_task_rq()

---
 kernel/sched/alt_core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 00b3144d7..d0d121d59 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -1975,6 +1975,8 @@ static int select_fallback_rq(int cpu, struct task_struct *p)
 
 static inline int select_task_rq(struct task_struct *p)
 {
+	lockdep_assert_held(&p->pi_lock);
+
 	cpumask_t chk_mask, tmp;
 
 	if (unlikely(!cpumask_and(&chk_mask, p->cpus_ptr, cpu_active_mask)))
-- 
2.36.1.74.g277cf0bc36

