From 8f588f3b292c69b8a9d0ba61015d7c9ad961b11e Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Tue, 22 Oct 2019 18:33:26 +0200
Subject: [PATCH] mm/ksm, proc: bail out on kthreads

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 fs/proc/base.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/fs/proc/base.c b/fs/proc/base.c
index c462e1872aad..5eee439434a8 100644
--- a/fs/proc/base.c
+++ b/fs/proc/base.c
@@ -2999,7 +2999,11 @@ static int ksm_open(struct inode *inode, struct file *file)
 	struct mm_struct *mm = ERR_PTR(-ESRCH);
 
 	if (task) {
-		mm = mm_access(task, PTRACE_MODE_ATTACH_FSCREDS);
+		if (unlikely(task->flags & PF_KTHREAD))
+			mm = ERR_PTR(-EINVAL);
+		else
+			mm = mm_access(task, PTRACE_MODE_ATTACH_FSCREDS);
+
 		put_task_struct(task);
 
 		if (!IS_ERR_OR_NULL(mm)) {
-- 
2.24.0.rc0

