From c3909dd1d0813a041751394425caebf5ecc6bcd1 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 11 Dec 2023 15:53:14 -0500
Subject: [PATCH 71/74] squash! shm: Kill bogus dependency on list.h

---
 include/linux/shm.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/shm.h b/include/linux/shm.h
index 681a3eda7..c55bef053 100644
--- a/include/linux/shm.h
+++ b/include/linux/shm.h
@@ -4,10 +4,10 @@
 
 #include <linux/types.h>
 #include <asm/page.h>
-#include <uapi/linux/shm.h>
 #include <asm/shmparam.h>
 
 struct file;
+struct task_struct;
 
 #ifdef CONFIG_SYSVIPC
 struct sysv_shm {
-- 
2.43.0.232.ge79552d197

