From b68063ef5dc0a2e0fb39c349dbb23e5b78ae8cd0 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 4 Sep 2019 05:27:03 +0800
Subject: [PATCH 07/27] bmq: [Sync] ucamp

BMQ doesn't support ucamp.
---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index 2179feea4e4f..566ee17ed90e 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -735,6 +735,7 @@ menu "Scheduler features"
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
+	depends on !SCHED_BMQ
 	help
 	  This feature enables the scheduler to track the clamped utilization
 	  of each CPU based on RUNNABLE tasks scheduled on that CPU.
-- 
2.23.0.162.gf1d4a28250

