diff -Nadrup a/Makefile b/Makefile
--- a/Makefile	2018-11-04 14:50:54.000000000 +0100
+++ b/Makefile	2018-11-09 20:08:53.729996453 +0100
@@ -15,6 +15,10 @@ NAME = "People's Front"
 PHONY := _all
 _all:
 
+LLVERSION = .ll3
+LLNAME = linux-lucjan
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # o Do not use make's built-in rules and variables
 #   (this increases performance and avoids hard-to-debug behaviour);
 # o Look for make include files relative to root of kernel src
