diff -Narup a/Makefile b/Makefile
--- a/Makefile	2018-11-04 14:50:54.000000000 +0100
+++ b/Makefile	2018-11-12 22:31:40.749994681 +0100
@@ -11,6 +11,10 @@ NAME = "People's Front"
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLVERSION = .ll4
+LLNAME = linux-lucjan
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
