diff -Narup a/Makefile b/Makefile
--- a/Makefile	2018-11-13 20:09:00.000000000 +0100
+++ b/Makefile	2018-11-15 12:18:50.790145057 +0100
@@ -11,6 +11,10 @@ NAME = "People's Front"
 # Comments in this file are targeted only to the developer, do not
 # expect to learn how to build the kernel reading this file.
 
+LLNAME = linux-lucjan
+LLVERSION = .ll18
+EXTRAVERSION := $(EXTRAVERSION)$(LLVERSION)
+
 # That's our default target when none is given on the command line
 PHONY := _all
 _all:
