From 507e2546b08d9b966bfb28f307bc0c8887f25aa3 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Sun, 27 Oct 2019 22:32:24 +0100
Subject: [PATCH 3/3] mt76: add missing ASPM include

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/net/wireless/mediatek/mt76/pci.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/net/wireless/mediatek/mt76/pci.c b/drivers/net/wireless/mediatek/mt76/pci.c
index 04c5a692bc85..073e237e4728 100644
--- a/drivers/net/wireless/mediatek/mt76/pci.c
+++ b/drivers/net/wireless/mediatek/mt76/pci.c
@@ -4,6 +4,7 @@
  */
 
 #include <linux/pci.h>
+#include <linux/pci-aspm.h>
 
 void mt76_pci_disable_aspm(struct pci_dev *pdev)
 {
-- 
2.24.0.rc1

