From 8d64a3bc3c80e9c9677e3d51391663a7c90b6b6f Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Wed, 21 Aug 2019 15:51:34 +0200
Subject: [PATCH 3/6] block, bfq, DEBUG: log pid of queue chosen for injection

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 9f9db21a3d4b..06a4fd69a151 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -4973,6 +4973,10 @@ static struct bfq_queue *bfq_select_queue(struct bfq_data *bfqd)
 				     !bfq_bfqq_has_short_ttime(bfqq));
 			new_bfqq = bfq_choose_bfqq_for_injection(bfqd);
 			BUG_ON(new_bfqq == bfqq);
+			if (new_bfqq)
+				bfq_log_bfqq(bfqd, bfqq,
+					"chosen the queue %d for injection",
+					new_bfqq->pid);
 			bfqq = new_bfqq;
 		} else {
 			bfqq = NULL;
-- 
2.23.0

