From 8acc1ff4f6bffbc471c8cb3a227dfbd3cc7f9b4a Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Wed, 21 Aug 2019 15:38:27 +0200
Subject: [PATCH 6/6] block, bfq: raise bfq version to v11

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index a3b865c6a2b5..89f589089475 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7591,7 +7591,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ I/O-scheduler: v10";
+	char msg[60] = "BFQ I/O-scheduler: v11";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.23.0

