From 38a05c77cc59d0a39ead3db769c2496f760bbf1c Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Wed, 4 Sep 2019 05:27:03 +0800
Subject: [PATCH 07/53] bmq: [Sync] ucamp

BMQ doesn't support ucamp.
---
 init/Kconfig | 1 +
 1 file changed, 1 insertion(+)

diff --git a/init/Kconfig b/init/Kconfig
index a5adad6b1c17..85e26e3130f9 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -740,6 +740,7 @@ menu "Scheduler features"
 config UCLAMP_TASK
 	bool "Enable utilization clamping for RT/FAIR tasks"
 	depends on CPU_FREQ_GOV_SCHEDUTIL
+	depends on !SCHED_BMQ
 	help
 	  This feature enables the scheduler to track the clamped utilization
 	  of each CPU based on RUNNABLE tasks scheduled on that CPU.
-- 
2.24.0.155.gd9f6f3b619

