From 67a304a4f52145abfbbe56ec18e0a86656f7cfe2 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 4 Jul 2019 19:15:37 +0800
Subject: [PATCH 12/53] bmq: Remove PDS comments.

---
 kernel/sched/bmq.c | 7 +------
 1 file changed, 1 insertion(+), 6 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 3e3c89d4cac7..3e11373e7e73 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -601,11 +601,7 @@ static inline void sched_update_tick_dependency(struct rq *rq) { }
 #endif
 
 /*
- * Removing from the runqueue. Deleting a task from the skip list is done
- * via the stored node reference in the task struct and does not require a full
- * look up. Thus it occurs in O(k) time where k is the "level" of the list the
- * task was stored at - usually < 4, max 16.
- *
+ * Removing from the runqueue.
  * Context: rq->lock
  */
 static inline void dequeue_task(struct task_struct *p, struct rq *rq, int flags)
@@ -634,7 +630,6 @@ static inline void dequeue_task(struct task_struct *p, struct rq *rq, int flags)
 
 /*
  * Adding task to the runqueue.
- *
  * Context: rq->lock
  */
 static inline void enqueue_task(struct task_struct *p, struct rq *rq, int flags)
-- 
2.24.0.155.gd9f6f3b619

