From 4ec175850650aa2a617ca4705956a027a13dd601 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 18 Jul 2019 13:36:51 +0800
Subject: [PATCH 16/53] bmq: Instroduce mainline ttwu_queue().

---
 kernel/sched/bmq.c | 24 +++++++++++++-----------
 1 file changed, 13 insertions(+), 11 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index face8875d58a..f682322d0398 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -1461,6 +1461,17 @@ ttwu_do_activate(struct rq *rq, struct task_struct *p, int wake_flags)
 	ttwu_do_wakeup(rq, p, 0);
 }
 
+static inline void ttwu_queue(struct task_struct *p, int cpu, int wake_flags)
+{
+	struct rq *rq = cpu_rq(cpu);
+
+	raw_spin_lock(&rq->lock);
+	update_rq_clock(rq);
+	ttwu_do_activate(rq, p, wake_flags);
+	check_preempt_curr(rq, p);
+	raw_spin_unlock(&rq->lock);
+}
+
 static int ttwu_remote(struct task_struct *p, int wake_flags)
 {
 	struct rq *rq;
@@ -1587,7 +1598,6 @@ static int try_to_wake_up(struct task_struct *p, unsigned int state,
 			  int wake_flags)
 {
 	unsigned long flags;
-	struct rq *rq;
 	int cpu, success = 0;
 
 	preempt_disable();
@@ -1708,17 +1718,9 @@ static int try_to_wake_up(struct task_struct *p, unsigned int state,
 		delayacct_blkio_end(p);
 		atomic_dec(&task_rq(p)->nr_iowait);
 	}
-#endif
-
-	rq = cpu_rq(cpu);
-	raw_spin_lock(&rq->lock);
-
-	update_rq_clock(rq);
-	ttwu_do_activate(rq, p, wake_flags);
-	check_preempt_curr(rq, p);
-
-	raw_spin_unlock(&rq->lock);
+#endif /* CONFIG_SMP */
 
+	ttwu_queue(p, cpu, wake_flags);
 unlock:
 	raw_spin_unlock_irqrestore(&p->pi_lock, flags);
 out:
-- 
2.24.0.155.gd9f6f3b619

