From f8b7e69516f0afe029c765d22a1ecb07a483a919 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 1 Aug 2019 22:51:34 +0800
Subject: [PATCH 21/53] bmq: Consider llc in ttwu code path.

---
 kernel/sched/bmq.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index c5ed01de2b79..4737772051b2 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -1360,15 +1360,24 @@ static inline int select_task_rq(struct task_struct *p)
 	if (unlikely(!cpumask_and(&chk_mask, p->cpus_ptr, cpu_online_mask)))
 		return select_fallback_rq(task_cpu(p), p);
 
-	preempt_level = SCHED_PRIO2WATERMARK(task_sched_prio(p));
 	level = bmq_find_first_bit(sched_rq_watermark_bitmap, WM_BITS);
-	while (level < preempt_level) {
+	while (level < 2) {
 		if (cpumask_and(&tmp, &chk_mask, &sched_rq_watermark[level]))
 			return best_mask_cpu(task_cpu(p), &tmp);
-
 		level = bmq_find_next_bit(sched_rq_watermark_bitmap, WM_BITS,
 					  level + 1);
 	}
+	preempt_level = SCHED_PRIO2WATERMARK(task_sched_prio(p));
+	if (level < preempt_level) {
+		cpumask_clear(&tmp);
+		do {
+			cpumask_or(&tmp, &tmp, &sched_rq_watermark[level]);
+			level = bmq_find_next_bit(sched_rq_watermark_bitmap,
+						  WM_BITS, level + 1);
+		} while (level < preempt_level);
+		if (cpumask_and(&tmp, &tmp, &chk_mask))
+			return best_mask_cpu(task_cpu(p), &tmp);
+	}
 
 	return best_mask_cpu(task_cpu(p), &chk_mask);
 }
@@ -2819,8 +2828,7 @@ static inline int take_other_rq_tasks(struct rq *rq, int cpu)
 	affinity_mask = per_cpu(sched_cpu_llc_start_mask, cpu);
 	end_mask = per_cpu(sched_cpu_affinity_chk_end_masks, cpu);
 	do {
-		int i, tried = 0;
-
+		int i;
 		for_each_cpu_and(i, &sched_rq_pending_mask, affinity_mask) {
 			int nr_migrated;
 			struct rq *src_rq;
@@ -2840,9 +2848,6 @@ static inline int take_other_rq_tasks(struct rq *rq, int cpu)
 				cpufreq_update_util(rq, 0);
 				return 1;
 			}
-			if (tried)
-				return 0;
-			tried++;
 		}
 	} while (++affinity_mask < end_mask);
 
-- 
2.24.0.155.gd9f6f3b619

