From d67e9df31028c2d4ac1a57a8c9c70708e98a84ae Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 15 Aug 2019 23:33:58 +0800
Subject: [PATCH 23/53] Tag BMQ v0.99

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 81a50e851dd2..b1cc2518157f 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -56,7 +56,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.98 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 0.99 by Alfred Chen.\n");
 }
 
 /**
-- 
2.24.0.155.gd9f6f3b619

