From 5056b08b6185088b312aeb1d4cef5ef0e179d662 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 30 Sep 2019 17:48:13 +0800
Subject: [PATCH 30/53] Tag BMQ v5.3.1

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 2ed179a833f4..034a5fa1859d 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 1.00 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.3.1 by Alfred Chen.\n");
 }
 
 /**
-- 
2.24.0.155.gd9f6f3b619

