From 89cf2aab7249939cef0f077afb121eee3ce1f87d Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 22 Oct 2019 15:35:50 +0800
Subject: [PATCH 36/53] Tag BMQ v5.3-r2

---
 kernel/sched/bmq.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index a61d5263c144..4302cfe1048c 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -68,7 +68,7 @@
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.3.1 by Alfred Chen.\n");
+	printk(KERN_INFO "bmq: BMQ CPU Scheduler 5.3-r2 by Alfred Chen.\n");
 }
 
 /**
-- 
2.24.0.155.gd9f6f3b619

