From 298c47b42d6f0d5dc4beca7b8670323a70332f54 Mon Sep 17 00:00:00 2001
From: Steven Barrett <steven@liquorix.net>
Date: Mon, 13 Sep 2021 20:19:08 -0500
Subject: [PATCH 2/4] ZEN: Disable CacULE by default

---
 init/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index ceab944b6..e134fede2 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -839,13 +839,13 @@ endmenu
 
 config CACULE_SCHED
 	bool "CacULE CPU scheduler"
-	default y
+	default n
 	help
 	  The CacULE CPU scheduler is based on interactivity score mechanism.
 	  The interactivity score is inspired by the ULE scheduler (FreeBSD
 	  scheduler).
 
-	  If unsure, say Y here.
+	  If unsure, say N here.
 
 config CACULE_RDB
 	bool "RDB (Response Driven Balancer)"
-- 
2.33.1.711.g9d530dc002

