From 5f39219c872a844b9b31592f53bda7500eb5f8a0 Mon Sep 17 00:00:00 2001
From: Steven Barrett <steven@liquorix.net>
Date: Mon, 13 Sep 2021 20:19:08 -0500
Subject: [PATCH 2/4] ZEN: Disable CacULE by default

---
 init/Kconfig | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index b6db195db..fbb1aa393 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -867,13 +867,13 @@ endmenu
 
 config CACULE_SCHED
 	bool "CacULE CPU scheduler"
-	default y
+	default n
 	help
 	  The CacULE CPU scheduler is based on interactivity score mechanism.
 	  The interactivity score is inspired by the ULE scheduler (FreeBSD
 	  scheduler).
 
-	  If unsure, say Y here.
+	  If unsure, say N here.
 
 config CACULE_RDB
 	bool "RDB (Response Driven Balancer)"
-- 
2.34.1.75.gabe6bb3905

