From 11e97785558c110d610c11b87e561fdcb53ca683 Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Wed, 20 Sep 2023 14:56:45 -0700
Subject: [PATCH 8/8] fixups: mmu_notifier fixes

Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 mm/userfaultfd.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/mm/userfaultfd.c b/mm/userfaultfd.c
index 716867a9e..7b0353fe3 100644
--- a/mm/userfaultfd.c
+++ b/mm/userfaultfd.c
@@ -904,6 +904,9 @@ static int remap_pages_pte(struct mm_struct *dst_mm,
 	struct mmu_notifier_range range;
 	int err = 0;
 
+	mmu_notifier_range_init(&range, MMU_NOTIFY_CLEAR, 0, src_mm,
+				src_addr, src_addr + PAGE_SIZE);
+	mmu_notifier_invalidate_range_start(&range);
 retry:
 	dst_pte = pte_offset_map_nolock(dst_mm, dst_pmd, dst_addr, &dst_ptl);
 
@@ -1008,10 +1011,6 @@ static int remap_pages_pte(struct mm_struct *dst_mm,
 			}
 		}
 
-		mmu_notifier_range_init(&range, MMU_NOTIFY_CLEAR, 0, src_mm,
-					src_addr, src_addr + PAGE_SIZE);
-		mmu_notifier_invalidate_range_start_nonblock(&range);
-
 		double_pt_lock(dst_ptl, src_ptl);
 
 		if (!pte_same(*src_pte, orig_src_pte) ||
@@ -1050,7 +1049,6 @@ static int remap_pages_pte(struct mm_struct *dst_mm,
 		double_pt_unlock(dst_ptl, src_ptl);
 
 		anon_vma_unlock_write(src_anon_vma);
-		mmu_notifier_invalidate_range_end(&range);
 		put_anon_vma(src_anon_vma);
 		src_anon_vma = NULL;
 
@@ -1130,6 +1128,7 @@ static int remap_pages_pte(struct mm_struct *dst_mm,
 		pte_unmap(dst_pte);
 	if (src_pte)
 		pte_unmap(src_pte);
+	mmu_notifier_invalidate_range_end(&range);
 
 	return err;
 }
-- 
2.42.0

