From 15206fc674b9397e5f07b260a6ddcb6abd338caf Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 4 Dec 2023 13:28:16 +0100
Subject: [PATCH 70/88] scx_nest: More include path and build updates

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 tools/sched_ext/scx_nest.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/tools/sched_ext/scx_nest.c b/tools/sched_ext/scx_nest.c
index 4ad0279f7..87c831df2 100644
--- a/tools/sched_ext/scx_nest.c
+++ b/tools/sched_ext/scx_nest.c
@@ -9,7 +9,6 @@
 #include <signal.h>
 #include <libgen.h>
 #include <bpf/bpf.h>
-#include "user_exit_info.h"
 #include "scx_nest.bpf.skel.h"
 #include <scx/common.h>
 #include "scx_nest.h"
-- 
2.43.0.232.ge79552d197

