From e29764319767d0a0678a9e4ed33833c1edbde8e4 Mon Sep 17 00:00:00 2001
From: Andrew Morton <akpm@linux-foundation.org>
Date: Thu, 26 Jan 2023 15:36:39 -0800
Subject: [PATCH 356/356] mm-add-memcpy_from_file_folio-fix

Cc: "Fabio M. De Francesco" <fmdefrancesco@gmail.com>
Cc: Ira Weiny <ira.weiny@intel.com>
Cc: "Matthew Wilcox (Oracle)" <willy@infradead.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---
 include/linux/highmem.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/linux/highmem.h b/include/linux/highmem.h
index 82017c822..344cfc23b 100644
--- a/include/linux/highmem.h
+++ b/include/linux/highmem.h
@@ -433,7 +433,7 @@ static inline size_t memcpy_from_file_folio(char *to, struct folio *folio,
 	char *from = kmap_local_folio(folio, offset);
 
 	if (folio_test_highmem(folio))
-		len = min(len, PAGE_SIZE - offset);
+		len = min_t(size_t, len, PAGE_SIZE - offset);
 	else
 		len = min(len, folio_size(folio) - offset);
 
-- 
2.39.0.rc2.1.gbd5df96b79

