From 361884c5282c96f6f04c93f331ed1cb28dd9f240 Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Sun, 18 Feb 2024 23:03:07 -0500
Subject: [PATCH 37/37] fixup! lib: add memory allocations report in show_mem()

---
 mm/show_mem.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/show_mem.c b/mm/show_mem.c
index 3bf74129a..511ceaed4 100644
--- a/mm/show_mem.c
+++ b/mm/show_mem.c
@@ -433,7 +433,7 @@ void __show_mem(unsigned int filter, nodemask_t *nodemask, int max_zone_idx)
 			printk("Memory allocations:\n");
 			seq_buf_init(&s, buf, 4096);
 			alloc_tags_show_mem_report(&s);
-			printk("%s", buf);
+			printk(KERN_NOTICE "%s", seq_buf_str(&s));
 			kfree(buf);
 		}
 	}
-- 
2.43.0.232.ge79552d197

