From e77727d23858174eac588306df7c53741c7db4db Mon Sep 17 00:00:00 2001
From: Wyes Karny <wyes.karny@amd.com>
Date: Tue, 20 Dec 2022 06:11:44 +0000
Subject: [PATCH 4/4] amd_pstate: Set auto select register

Signed-off-by: Wyes Karny <wyes.karny@amd.com>
---
 drivers/cpufreq/amd-pstate.c | 29 +++++++++++++++++++++++++++++
 1 file changed, 29 insertions(+)

diff --git a/drivers/cpufreq/amd-pstate.c b/drivers/cpufreq/amd-pstate.c
index 50f5fa84b..ae6889f10 100644
--- a/drivers/cpufreq/amd-pstate.c
+++ b/drivers/cpufreq/amd-pstate.c
@@ -164,6 +164,22 @@ static int cppc_init_perf(struct amd_cpudata *cpudata)
 		   cppc_perf.lowest_nonlinear_perf);
 	WRITE_ONCE(cpudata->lowest_perf, cppc_perf.lowest_perf);
 
+	ret = cppc_get_auto_sel_caps(cpudata->cpu, &cppc_perf);
+	if (ret) {
+		pr_warn("WYES_DEBUG: failed to get auto_sel caps\n");
+		return 0;
+	}
+
+	pr_info("WYES_DEBUG: auto_sel: %d\n", cppc_perf.auto_sel);
+
+	if (cppc_state == CPPC_PASSIVE)
+		ret = cppc_set_auto_sel(cpudata->cpu, 0);
+	else if (cppc_state == CPPC_GUIDED)
+		ret = cppc_set_auto_sel(cpudata->cpu, 1);
+
+	if (ret)
+		pr_warn("WYES_DEBUG: failed to set auto_sel caps\n");
+
 	return 0;
 }
 
@@ -663,7 +679,20 @@ static int amd_pstate_unregister_driver(int dummy)
 
 static int amd_pstate_change_driver_mode(int mode)
 {
+	int cpu = 0;
+
 	cppc_state = mode;
+	if (!boot_cpu_has(X86_FEATURE_CPPC)) {
+		if (cppc_state == CPPC_PASSIVE) {
+			for_each_present_cpu(cpu) {
+				cppc_set_auto_sel(cpu, 0);
+			}
+		} else if (cppc_state == CPPC_GUIDED) {
+			for_each_present_cpu(cpu) {
+				cppc_set_auto_sel(cpu, 1);
+			}
+		}
+	}
 	return 0;
 }
 
-- 
2.39.0.rc2.1.gbd5df96b79

