From ad195bf12b73fd8ef8a289b1b3905008529dd9be Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 6 Mar 2023 10:03:06 +0100
Subject: [PATCH 4/5] ddcci-6.3: adopt 162736b0d71a9 and a9b12f8b4e330

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/char/ddcci.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/char/ddcci.c b/drivers/char/ddcci.c
index 77fee6e03..a8b9c4ed4 100644
--- a/drivers/char/ddcci.c
+++ b/drivers/char/ddcci.c
@@ -931,7 +931,7 @@ ATTRIBUTE_GROUPS(ddcci_char_device);
 
 /* DDC/CI bus */
 
-static int ddcci_device_uevent(struct device *dev, struct kobj_uevent_env *env)
+static int ddcci_device_uevent(const struct device *dev, struct kobj_uevent_env *env)
 {
 	struct ddcci_device	*device = to_ddcci_device(dev);
 	char model[ARRAY_SIZE(device->model)];
@@ -1011,7 +1011,7 @@ static void ddcci_device_release(struct device *dev)
 	kfree(device);
 }
 
-static char *ddcci_devnode(struct device *dev,
+static char *ddcci_devnode(const struct device *dev,
 			 umode_t *mode, kuid_t *uid, kgid_t *gid)
 {
 	struct ddcci_device *device;
@@ -1021,7 +1021,7 @@ static char *ddcci_devnode(struct device *dev,
 			 device->i2c_client->adapter->nr);
 }
 
-static char *ddcci_dependent_devnode(struct device *dev,
+static char *ddcci_dependent_devnode(const struct device *dev,
 			 umode_t *mode, kuid_t *uid, kgid_t *gid)
 {
 	struct ddcci_device *device;
-- 
2.39.2.501.gd9d677b2d8

