From 493233d9e48e78cd9dff3ca0819ddab4cbb4969b Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 6 Mar 2023 10:09:25 +0100
Subject: [PATCH 5/5] ddcci-6.3: adopt d2bf38c088e0

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/char/ddcci.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/char/ddcci.c b/drivers/char/ddcci.c
index a8b9c4ed4..158143fda 100644
--- a/drivers/char/ddcci.c
+++ b/drivers/char/ddcci.c
@@ -43,6 +43,7 @@ static DEFINE_SEMAPHORE(core_lock);
 
 struct bus_type ddcci_bus_type;
 EXPORT_SYMBOL_GPL(ddcci_bus_type);
+static bool ddcci_bus_registered;
 
 /* Assert neccessary string array sizes  */
 #ifndef sizeof_field
@@ -1100,7 +1101,7 @@ int ddcci_register_driver(struct module *owner, struct ddcci_driver *driver)
 	int ret;
 
 	/* Can't register until after driver model init */
-	if (unlikely(WARN_ON(!ddcci_bus_type.p)))
+	if (unlikely(WARN_ON(!ddcci_bus_registered)))
 		return -EAGAIN;
 
 	pr_debug("registering driver [%s]\n", driver->driver.name);
@@ -1849,6 +1850,7 @@ static int __init ddcci_module_init(void)
 		pr_err("failed to register bus 'ddcci'\n");
 		goto err_busreg;
 	}
+	ddcci_bus_registered = true;
 
 	/* Register I2C driver */
 	ret = i2c_add_driver(&ddcci_driver);
-- 
2.39.2.501.gd9d677b2d8

