From f6b661ffdaa212ef58647036dfc00218d238613f Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Fri, 14 Jul 2023 12:53:06 -0400
Subject: [PATCH 47/47] fixup! mm: vmalloc: Enable memory allocation profiling

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 include/linux/vmalloc.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/vmalloc.h b/include/linux/vmalloc.h
index 8a622faed..106d78e75 100644
--- a/include/linux/vmalloc.h
+++ b/include/linux/vmalloc.h
@@ -2,6 +2,8 @@
 #ifndef _LINUX_VMALLOC_H
 #define _LINUX_VMALLOC_H
 
+#include <linux/alloc_tag.h>
+#include <linux/sched.h>
 #include <linux/spinlock.h>
 #include <linux/init.h>
 #include <linux/list.h>
-- 
2.41.0.159.g0bfa463d37

