From 5aee8bde2c30125b9e644bb817ee5a1f15086e52 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Thu, 27 Sep 2018 14:55:16 +0800
Subject: [PATCH 02/42] pds: [Sync] f0b752168d70 audit: convert sessionid unset
 to a macro

---
 init/init_task.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/init_task.c b/init/init_task.c
index 9f7bc70c9339..9faaf8257186 100644
--- a/init/init_task.c
+++ b/init/init_task.c
@@ -124,7 +124,7 @@ struct task_struct init_task
 	.thread_node	= LIST_HEAD_INIT(init_signals.thread_head),
 #ifdef CONFIG_AUDITSYSCALL
 	.loginuid	= INVALID_UID,
-	.sessionid	= (unsigned int)-1,
+	.sessionid	= AUDIT_SID_UNSET,
 #endif
 #ifdef CONFIG_PERF_EVENTS
 	.perf_event_mutex = __MUTEX_INITIALIZER(init_task.perf_event_mutex),
-- 
2.20.1.2.gb21ebb671b

