From 933acede401ea8290e8909dd62ba54dfe08a421b Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 28 Sep 2018 10:34:11 +0800
Subject: [PATCH 05/42] pds: [Sync] 088fe47ce952 signal: Add
 calculate_sigpending()

---
 kernel/sched/pds.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 9603b9fd2368..8e25595b73b2 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -2582,6 +2582,8 @@ asmlinkage __visible void schedule_tail(struct task_struct *prev)
 
 	if (current->set_child_tid)
 		put_user(task_pid_vnr(current), current->set_child_tid);
+
+	calculate_sigpending();
 }
 
 /*
-- 
2.20.1.2.gb21ebb671b

