From 48cb1f616baae90c971942bd20aa86733bea81f3 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 29 Oct 2018 15:32:23 +0800
Subject: [PATCH 15/42] pds: Optimize ISO branch in normal_prio().

---
 kernel/sched/pds.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 98b52add292e..e1fb57addbf6 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -1079,8 +1079,6 @@ static inline int normal_prio(struct task_struct *p)
 {
 	if (normal_task(p))
 		return NORMAL_PRIO;
-	if (iso_task(p))
-		return ISO_PRIO;
 	if (has_rt_policy(p))
 		return MAX_RT_PRIO - 1 - p->rt_priority;
 	return IDLE_PRIO;
-- 
2.20.1.2.gb21ebb671b

