From 550499546f43d845737f8c8e2e7cadefd120aeee Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 29 Oct 2018 15:37:04 +0800
Subject: [PATCH 16/42] Tag PDS 0.99c

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index e1fb57addbf6..f4bf4cfd8af9 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -96,7 +96,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99b by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99c by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.20.1.2.gb21ebb671b

