From 6f649391f56f6d51297caf62030cba77c4340b96 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Mon, 12 Nov 2018 10:17:39 +0000
Subject: [PATCH 18/42] pds: UP compilation fix.

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 7b08d3eeb566..a2ff24905914 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -319,7 +319,7 @@ static void update_rq_clock_task(struct rq *rq, s64 delta)
  * In theory, the compile should just see 0 here, and optimize out the call
  * to sched_rt_avg_update. But I don't trust it...
  */
-#ifdef HAVE_SCHED_AVG_IRQ
+#if defined(CONFIG_IRQ_TIME_ACCOUNTING) || defined(CONFIG_PARAVIRT_TIME_ACCOUNTING)
 	s64 steal = 0, irq_delta = 0;
 #endif
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
-- 
2.20.1.2.gb21ebb671b

