From 90b961fe1cc4a3db8a872dea708e88f05032d822 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Fri, 16 Nov 2018 14:01:32 +0000
Subject: [PATCH 23/42] Tag PDS 0.99e

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 0c52506cb3d0..35a88c457238 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -84,7 +84,7 @@ enum {
 
 static inline void print_scheduler_version(void)
 {
-	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99d by Alfred Chen.\n");
+	printk(KERN_INFO "pds: PDS-mq CPU Scheduler 0.99e by Alfred Chen.\n");
 }
 
 /* task_struct::on_rq states: */
-- 
2.20.1.2.gb21ebb671b

