From 9409b6b0726dc5d452ea6b792d862cc8acb6b352 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 13 Nov 2018 09:59:18 +0000
Subject: [PATCH 25/42] pds: Default 4ms rr_interal.

---
 kernel/sched/pds.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/pds.c b/kernel/sched/pds.c
index 6e6ce87fe9a8..cb9e1ef1d90f 100644
--- a/kernel/sched/pds.c
+++ b/kernel/sched/pds.c
@@ -106,7 +106,7 @@ static inline int task_on_rq_migrating(struct task_struct *p)
  * Value is in ms and set to a minimum of 6ms. Scales with number of cpus.
  * Tunable via /proc interface.
  */
-#define SCHED_DEFAULT_RR (6)
+#define SCHED_DEFAULT_RR (4)
 int rr_interval __read_mostly = SCHED_DEFAULT_RR;
 
 static int __init rr_interval_set(char *str)
-- 
2.20.1.2.gb21ebb671b

