From efda065d1e69e43d5ada0741bbaeb42a29836c44 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 30 Mar 2020 18:03:26 +0200
Subject: [PATCH 10/36] block: Set bfq-dev instead of bfq

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 671314afa..3a1f2aa6f 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -7693,7 +7693,7 @@ MODULE_ALIAS("bfq-iosched");
 static int __init bfq_init(void)
 {
 	int ret;
-	char msg[60] = "BFQ I/O-scheduler: v11";
+	char msg[60] = "BFQ-dev I/O-scheduler: v11";
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
 	ret = blkcg_policy_register(&blkcg_policy_bfq);
-- 
2.30.0.81.g72c4083ddf

