From 8b20b8087d330fd882c5f2d12916a3b6a6590ad2 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 2 Dec 2019 15:05:30 +0100
Subject: [PATCH 2/7] Makefile: add -bfq version suffix

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index d4d36c61940b..e9d95b2f8e89 100644
--- a/Makefile
+++ b/Makefile
@@ -2,7 +2,7 @@
 VERSION = 5
 PATCHLEVEL = 4
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION = -bfq
 NAME = Kleptomaniac Octopus
 
 # *DOCUMENTATION*
-- 
2.24.0.155.gd9f6f3b619

