From 153c69d4856658ac11bdc5c0f319b343ca4e296e Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Mon, 2 Dec 2019 16:24:20 +0100
Subject: [PATCH 04/18] block, bfq, DEBUG: check proc refs in case of requeue

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index b7cfd36ec0a8..ee58e19cf4ca 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -3897,6 +3897,8 @@ static bool __bfq_bfqq_expire(struct bfq_data *bfqd, struct bfq_queue *bfqq,
 
 		bfq_del_bfqq_busy(bfqd, bfqq, true);
 	} else {
+		BFQ_BUG_ON(RB_EMPTY_ROOT(&bfqq->sort_list) &&
+			   !bfqq_process_refs(bfqq));
 		bfq_requeue_bfqq(bfqd, bfqq, true);
 		/*
 		 * Resort priority tree of potential close cooperators.
-- 
2.25.0.114.g5b0ca878e0.dirty

