From 819106ea6a7cc24a464207a87c6a641775317430 Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Wed, 3 May 2017 12:05:15 -0400
Subject: [PATCH 003/103] enable SECURITY_DMESG_RESTRICT by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 security/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/security/Kconfig b/security/Kconfig
index 4065df6b9..01cad4dd5 100644
--- a/security/Kconfig
+++ b/security/Kconfig
@@ -9,7 +9,7 @@ source "security/keys/Kconfig"
 
 config SECURITY_DMESG_RESTRICT
 	bool "Restrict unprivileged access to the kernel syslog"
-	default n
+	default y
 	help
 	  This enforces restrictions on unprivileged users reading the kernel
 	  syslog via dmesg(8).
-- 
2.36.1.203.g1bcf4f6271

