From c7df286bca28b328602784aaaa221f73e4b42c7a Mon Sep 17 00:00:00 2001
From: Daniel Micay <danielmicay@gmail.com>
Date: Mon, 29 May 2017 12:11:11 -0400
Subject: [PATCH 026/103] enable DEBUG_WX by default

Signed-off-by: Daniel Micay <danielmicay@gmail.com>
---
 mm/Kconfig.debug | 1 +
 1 file changed, 1 insertion(+)

diff --git a/mm/Kconfig.debug b/mm/Kconfig.debug
index 5bd5bb097..c9ed59dd4 100644
--- a/mm/Kconfig.debug
+++ b/mm/Kconfig.debug
@@ -130,6 +130,7 @@ config DEBUG_WX
 	depends on ARCH_HAS_DEBUG_WX
 	depends on MMU
 	select PTDUMP_CORE
+	default y
 	help
 	  Generate a warning if any W+X mappings are found at boot.
 
-- 
2.36.1.203.g1bcf4f6271

