From 73689675f047a66073cdf25e56266f11bd4014c7 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Fri, 16 Feb 2024 16:30:21 +0100
Subject: [PATCH 3/3] v4l2loopback-6.7: update to latest git HEAD

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/media/v4l2-core/v4l2loopback.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/media/v4l2-core/v4l2loopback.c b/drivers/media/v4l2-core/v4l2loopback.c
index 8529007ac..5861995ec 100644
--- a/drivers/media/v4l2-core/v4l2loopback.c
+++ b/drivers/media/v4l2-core/v4l2loopback.c
@@ -188,8 +188,8 @@ MODULE_PARM_DESC(
 		V4L2LOOPBACK_DEFAULT_EXCLUSIVECAPS) "]");
 
 /* format specifications */
-#define V4L2LOOPBACK_SIZE_MIN_WIDTH 48
-#define V4L2LOOPBACK_SIZE_MIN_HEIGHT 32
+#define V4L2LOOPBACK_SIZE_MIN_WIDTH 2
+#define V4L2LOOPBACK_SIZE_MIN_HEIGHT 1
 #define V4L2LOOPBACK_SIZE_DEFAULT_MAX_WIDTH 8192
 #define V4L2LOOPBACK_SIZE_DEFAULT_MAX_HEIGHT 8192
 
-- 
2.43.0.232.ge79552d197

