From b6aab287fa98e458a876a21cff59133b868df59a Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 29 Apr 2021 02:53:36 +0200
Subject: [PATCH 1/4] sched/alt: Add MG-LRU changes through ifdef macro

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/alt_core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 918d22237..2ad3d8c5f 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -3656,6 +3656,9 @@ context_switch(struct rq *rq, struct task_struct *prev,
 		 * finish_task_switch()'s mmdrop().
 		 */
 		switch_mm_irqs_off(prev->active_mm, next->mm, next);
+#ifdef CONFIG_LRU_GEN
+		lru_gen_use_mm(next->mm);
+#endif
 
 		if (!prev->mm) {                        // from kernel
 			/* will mmdrop() in finish_task_switch(). */
-- 
2.35.1.354.g715d08a9e5


From c38d60fb89f7f5d5e258a215319709e58ef207a7 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 28 Sep 2020 20:20:29 +0200
Subject: [PATCH 2/4] sched/alt: Export can_nice symbol for Android Binder IPC

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 kernel/sched/alt_core.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 2ad3d8c5f..9a5fea3f6 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -5275,6 +5275,7 @@ int can_nice(const struct task_struct *p, const int nice)
 	return (nice_rlim <= task_rlimit(p, RLIMIT_NICE) ||
 		capable(CAP_SYS_NICE));
 }
+EXPORT_SYMBOL_GPL(can_nice);
 
 #ifdef __ARCH_WANT_SYS_NICE
 
-- 
2.35.1.354.g715d08a9e5


From 69572f3f8b8e7dc7279d9dbf796fee6fe9d9ad15 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 17 Sep 2020 01:38:47 +0200
Subject: [PATCH 3/4] init, Kconfig: set default value of SCHED_PDS

Prefer PDS to BMQ by default due to better CPU utilisation & higher throughput.

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index 2213c3060..c9b17ae25 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -874,7 +874,7 @@ if SCHED_ALT
 
 choice
 	prompt "Alternative CPU Scheduler"
-	default SCHED_BMQ
+	default SCHED_PDS
 
 config SCHED_BMQ
 	bool "BMQ CPU scheduler"
-- 
2.35.1.354.g715d08a9e5


From 4c1291d090d54502746d11ddf2961cd364b92c63 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 21 Sep 2020 17:26:14 +0200
Subject: [PATCH 4/4] init, Kconfig: Restore original PDS description

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 init/Kconfig | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/init/Kconfig b/init/Kconfig
index c9b17ae25..6c4a1f3c1 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -887,7 +887,9 @@ config SCHED_PDS
 	bool "PDS CPU scheduler"
 	help
 	  The Priority and Deadline based Skip list multiple queue CPU
-	  Scheduler.
+	  scheduler for excellent interactivity and responsiveness on the
+	  desktop and solid scalability on normal hardware and commodity
+	  servers.
 
 endchoice
 
-- 
2.35.1.354.g715d08a9e5

