From 7e14f56ced4cc469f0f193c2ddd5e14bfda22b80 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Mon, 1 Nov 2021 13:19:33 +0100
Subject: [PATCH 3/9] port: initialize ata before graphics

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 Makefile.rej | 12 ++++++++++++
 1 file changed, 12 insertions(+)
 create mode 100644 Makefile.rej

diff --git a/Makefile.rej b/Makefile.rej
new file mode 100644
index 000000000..d1d42f53d
--- /dev/null
+++ b/Makefile.rej
@@ -0,0 +1,12 @@
+--- Makefile
++++ Makefile
+@@ -1020,9 +1020,6 @@ KBUILD_CFLAGS	+= -fno-strict-overflow
+ # Make sure -fstack-check isn't enabled (like gentoo apparently did)
+ KBUILD_CFLAGS  += -fno-stack-check
+ 
+-# conserve stack if available
+-KBUILD_CFLAGS   += $(call cc-option,-fconserve-stack)
+-
+ # Prohibit date/time macros, which would make the build non-deterministic
+ KBUILD_CFLAGS   += -Werror=date-time
+ 
-- 
2.33.1.711.g9d530dc002

