From bbbb8430ba9663c2caa1624e52a4b9602a477db8 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Tue, 27 Sep 2022 17:28:40 +0200
Subject: [PATCH 14/14] mm: fix pmadv_ksm system call

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 mm/madvise.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/madvise.c b/mm/madvise.c
index 2633d207b..1ce0e468a 100644
--- a/mm/madvise.c
+++ b/mm/madvise.c
@@ -1512,6 +1512,7 @@ SYSCALL_DEFINE3(pmadv_ksm, int, pidfd, int, behaviour, unsigned int, flags)
 	struct mm_struct *mm;
 	unsigned int f_flags;
 	struct vm_area_struct *vma;
+	VMA_ITERATOR(vmi, mm, 0);
 
 	if (flags != 0) {
 		ret = -EINVAL;
@@ -1558,7 +1559,7 @@ SYSCALL_DEFINE3(pmadv_ksm, int, pidfd, int, behaviour, unsigned int, flags)
 		goto release_mm;
 	}
 
-	for (vma = mm->mmap; vma; vma = vma->vm_next) {
+	for_each_vma(vmi, vma) {
 		switch (behaviour) {
 			case MADV_MERGEABLE:
 				ret = ksm_madvise_merge(vma->vm_mm, vma, &vma->vm_flags);
-- 
2.38.0.rc1.6.g4fd6c5e444

