From 785a20f74c17c37f14644edbf65084070ab48238 Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 19 Sep 2020 10:00:49 +0800
Subject: [PATCH 056/273] Project-C v5.9-r0

---
 kernel/sched/alt_core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 3c5eba046ed6..f36264fea75c 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -45,7 +45,7 @@
 #define CREATE_TRACE_POINTS
 #include <trace/events/sched.h>
 
-#define ALT_SCHED_VERSION "v5.8-r3"
+#define ALT_SCHED_VERSION "v5.9-r0"
 
 /* rt_prio(prio) defined in include/linux/sched/rt.h */
 #define rt_task(p)		rt_prio((p)->prio)
-- 
2.36.1.74.g277cf0bc36

