From 0146736a4ef9d1762505298014bfe71186a762bf Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 15 Dec 2020 11:19:05 +0800
Subject: [PATCH 079/273] sched/bmq: Fix compilation issue.

---
 kernel/sched/bmq_imp.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/bmq_imp.h b/kernel/sched/bmq_imp.h
index 83c2d019c446..3faba5f9bb69 100644
--- a/kernel/sched/bmq_imp.h
+++ b/kernel/sched/bmq_imp.h
@@ -62,7 +62,7 @@ static inline void time_slice_expired(struct task_struct *p, struct rq *rq)
 	}
 }
 
-static inline int task_running_nice(struct task_struct *p)
+inline int task_running_nice(struct task_struct *p)
 {
 	return (p->prio + p->boost_prio > DEFAULT_PRIO + MAX_PRIORITY_ADJ);
 }
-- 
2.36.1.74.g277cf0bc36

