From d26cb50d14aa7c8db9bf1b4adf947c22f4f03175 Mon Sep 17 00:00:00 2001
From: Tejun Heo <tj@kernel.org>
Date: Wed, 7 Feb 2024 09:08:19 -1000
Subject: [PATCH 125/132] scx: Actually allocate rq->scx.cpus_to_kick_if_idle

To avoid crashing the kernel when CONFIG_CPUMASK_OFFSTACK is set.

Signed-off-by: Tejun Heo <tj@kernel.org>
---
 kernel/sched/ext.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/kernel/sched/ext.c b/kernel/sched/ext.c
index b84450ebe..ec963a5fc 100644
--- a/kernel/sched/ext.c
+++ b/kernel/sched/ext.c
@@ -4382,6 +4382,7 @@ void __init init_sched_ext_class(void)
 		INIT_LIST_HEAD(&rq->scx.runnable_list);
 
 		BUG_ON(!zalloc_cpumask_var(&rq->scx.cpus_to_kick, GFP_KERNEL));
+		BUG_ON(!zalloc_cpumask_var(&rq->scx.cpus_to_kick_if_idle, GFP_KERNEL));
 		BUG_ON(!zalloc_cpumask_var(&rq->scx.cpus_to_preempt, GFP_KERNEL));
 		BUG_ON(!zalloc_cpumask_var(&rq->scx.cpus_to_wait, GFP_KERNEL));
 		init_irq_work(&rq->scx.kick_cpus_irq_work, kick_cpus_irq_workfn);
-- 
2.43.0.232.ge79552d197

