From 0d4c360bc5c539c6a1ccfe0326f85f2f43ec5e6f Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Tue, 15 Sep 2020 11:07:43 +0800
Subject: [PATCH 045/270] Project-C v5.8-r3

---
 kernel/sched/alt_core.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index 1dd4972beda4..b469c9488d18 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -45,7 +45,7 @@
 #define CREATE_TRACE_POINTS
 #include <trace/events/sched.h>
 
-#define ALT_SCHED_VERSION "v5.8-r2"
+#define ALT_SCHED_VERSION "v5.8-r3"
 
 /* rt_prio(prio) defined in include/linux/sched/rt.h */
 #define rt_task(p)		rt_prio((p)->prio)
-- 
2.35.1.677.gabf474a5dd

