From 631ae894c8f766f23bf748502e2aa36d14f7673a Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 13 Mar 2023 09:40:56 +0100
Subject: [PATCH 6/6] ddcci-6.3: adopt 03c835f498b5

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/char/ddcci.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/char/ddcci.c b/drivers/char/ddcci.c
index 158143fda..4d71d26ad 100644
--- a/drivers/char/ddcci.c
+++ b/drivers/char/ddcci.c
@@ -1673,8 +1673,9 @@ static int ddcci_detect(struct i2c_client *client, struct i2c_board_info *info)
 }
 
 /* I2C probe function */
-static int ddcci_probe(struct i2c_client *client, const struct i2c_device_id *id)
+static int ddcci_probe(struct i2c_client *client)
 {
+	const struct i2c_device_id *id = i2c_client_get_device_id(client);
 	int i, ret = -ENODEV, tmp;
 	unsigned char main_addr, addr;
 	struct ddcci_bus_drv_data *drv_data;
-- 
2.40.0.71.g950264636c

