From 36983447c20eb7c380c8d8d5a0f4bb8ab05e323d Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sun, 4 Apr 2021 08:54:29 +0800
Subject: [PATCH 133/269] sched/alt: Code clean up

---
 kernel/sched/alt_core.c | 11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/kernel/sched/alt_core.c b/kernel/sched/alt_core.c
index ae37764ced1e..3e79fdc14152 100644
--- a/kernel/sched/alt_core.c
+++ b/kernel/sched/alt_core.c
@@ -1138,20 +1138,11 @@ void set_task_cpu(struct task_struct *p, unsigned int new_cpu)
 	__set_task_cpu(p, new_cpu);
 }
 
-static inline bool is_per_cpu_kthread(struct task_struct *p)
-{
-	return ((p->flags & PF_KTHREAD) && (1 == p->nr_cpus_allowed));
-}
-
 #define MDF_FORCE_ENABLED	0x80
 
 static void
 __do_set_cpus_allowed(struct task_struct *p, const struct cpumask *new_mask, u32 flags);
 
-static int __set_cpus_allowed_ptr(struct task_struct *p,
-				  const struct cpumask *new_mask,
-				  u32 flags);
-
 void migrate_disable(void)
 {
 	struct task_struct *p = current;
@@ -1753,7 +1744,7 @@ static int __set_cpus_allowed_ptr(struct task_struct *p,
 		goto out;
 
 	if (p->migration_disabled) {
-		if (p->cpus_ptr != &p->cpus_mask)
+		if (likely(p->cpus_ptr != &p->cpus_mask))
 			__do_set_cpus_allowed(p, &p->cpus_mask, SCA_MIGRATE_ENABLE);
 		p->migration_disabled = 0;
 		p->migration_flags |= MDF_FORCE_ENABLED;
-- 
2.35.1.677.gabf474a5dd

