From 4e113c7e8988c755d4224fa02944d5e55378c679 Mon Sep 17 00:00:00 2001
From: "J. R. Okajima" <hooanon05g@gmail.com>
Date: Tue, 28 May 2019 16:42:30 +0900
Subject: [PATCH 4/8] aufs5.1 20190603

Signed-off-by: J. R. Okajima <hooanon05g@gmail.com>
---
 include/uapi/linux/aufs_type.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/uapi/linux/aufs_type.h b/include/uapi/linux/aufs_type.h
index f1607cdbcb6f..7ab67b4c3d78 100644
--- a/include/uapi/linux/aufs_type.h
+++ b/include/uapi/linux/aufs_type.h
@@ -40,7 +40,7 @@
 
 #include <linux/limits.h>
 
-#define AUFS_VERSION	"5.1-20190520"
+#define AUFS_VERSION	"5.1-20190603"
 
 /* todo? move this to linux-2.6.19/include/magic.h */
 #define AUFS_SUPER_MAGIC	('a' << 24 | 'u' << 16 | 'f' << 8 | 's')
-- 
2.22.0.214.g8dca754b1e

