From 99c57b6399caaf26c43d4546d787e6b55831c6fd Mon Sep 17 00:00:00 2001
From: "J. R. Okajima" <hooanon05g@gmail.com>
Date: Sun, 9 Jun 2019 01:39:48 +0900
Subject: [PATCH 8/8] aufs5.1 20190610

Signed-off-by: J. R. Okajima <hooanon05g@gmail.com>
---
 include/uapi/linux/aufs_type.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/uapi/linux/aufs_type.h b/include/uapi/linux/aufs_type.h
index 7ab67b4c3d78..a19cbb220391 100644
--- a/include/uapi/linux/aufs_type.h
+++ b/include/uapi/linux/aufs_type.h
@@ -40,7 +40,7 @@
 
 #include <linux/limits.h>
 
-#define AUFS_VERSION	"5.1-20190603"
+#define AUFS_VERSION	"5.1-20190610"
 
 /* todo? move this to linux-2.6.19/include/magic.h */
 #define AUFS_SUPER_MAGIC	('a' << 24 | 'u' << 16 | 'f' << 8 | 's')
-- 
2.22.0.214.g8dca754b1e

