From e050816b6a805a772f0317581a8601515bb4b6c2 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Sat, 22 Jun 2019 10:40:32 +0200
Subject: [PATCH 19/23] block, bfq, DEBUG: log per-queue rq completion times

Signed-off-by: Paolo Valente <paolo.valente@linaro.org>
---
 block/bfq-iosched.c | 4 ++++
 block/bfq-iosched.h | 4 ++++
 2 files changed, 8 insertions(+)

diff --git a/block/bfq-iosched.c b/block/bfq-iosched.c
index 21e1aa6148c0..52ef4596bace 100644
--- a/block/bfq-iosched.c
+++ b/block/bfq-iosched.c
@@ -5197,6 +5197,7 @@ static struct request *__bfq_dispatch_request(struct blk_mq_hw_ctx *hctx)
 	       bfq_bfqq_wait_request(bfqq));
 
 	rq = bfq_dispatch_rq_from_bfqq(bfqd, bfqq);
+	bfqq->last_dispatch = ktime_get_ns();
 
 	BFQ_BUG_ON(bfqq->entity.budget < bfqq->entity.service);
 
@@ -6161,6 +6162,9 @@ static void bfq_completed_request(struct bfq_queue *bfqq, struct bfq_data *bfqd)
 
 	now_ns = ktime_get_ns();
 
+	bfq_log_bfqq(bfqd, bfqq, "rq completion time: %llu us",
+		     div_u64(now_ns - bfqq->last_dispatch, NSEC_PER_USEC));
+
 	bfqq->ttime.last_end_request = now_ns;
 
 	/*
diff --git a/block/bfq-iosched.h b/block/bfq-iosched.h
index c154b232afe9..b97726ab616f 100644
--- a/block/bfq-iosched.h
+++ b/block/bfq-iosched.h
@@ -375,6 +375,10 @@ struct bfq_queue {
 	 * the woken queues when this queue exits.
 	 */
 	struct hlist_head woken_list;
+
+	/* time of last rq dispatch (ns) */
+	u64 last_dispatch;
+
 };
 
 /**
-- 
2.22.0.214.g8dca754b1e

