From 6aeda55680402889601936d56d071e10f85084de Mon Sep 17 00:00:00 2001
From: Alfred Chen <cchalpha@gmail.com>
Date: Sat, 8 Jun 2019 18:46:06 +0200
Subject: [PATCH] bmq-5.1: try to fix lockup

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 kernel/sched/bmq.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/bmq.c b/kernel/sched/bmq.c
index 13d1192e6074..6ae640f5c72e 100644
--- a/kernel/sched/bmq.c
+++ b/kernel/sched/bmq.c
@@ -156,8 +156,12 @@ ____cacheline_aligned_in_smp;
 #define TASK_SCHED_WATERMARK(p) (SCHED_PRIO2WATERMARK((p)->bmq_idx))
 
 #if (bmq_BITS <= BITS_PER_LONG) && (WM_BITS <= BITS_PER_LONG)
-#define bmq_find_first_bit(bm, size)		__ffs((bm[0]))
-#define bmq_find_next_bit(bm, size, start)	__ffs((bm[0] & BITMAP_FIRST_WORD_MASK(start)))
+#define bmq_find_first_bit(bm, size)		((bm[0])? __ffs((bm[0])):(size))
+#define bmq_find_next_bit(bm, size, start)	(\
+{\
+	unsigned long tmp = (bm[0] & BITMAP_FIRST_WORD_MASK(start));\
+	(tmp)? __ffs(tmp):(size);\
+})
 #else
 #define bmq_find_first_bit(bm, size)		find_first_bit((bm), (size))
 #define bmq_find_next_bit(bm, size, start)	find_next_bit(bm, size, start)
-- 
2.22.0.rc3

