From 9c0f3ebcb3bbd0a81f7533af83f822b7a205c04f Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Thu, 14 Jan 2021 16:57:36 +0100
Subject: [PATCH 35/38] port: bfq: don't check active group if bfq.weight is
 not changed

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 block/bfq-iosched.h | 21 +++++++++++----------
 1 file changed, 11 insertions(+), 10 deletions(-)

diff --git a/block/bfq-iosched.h b/block/bfq-iosched.h
index 551064beb..bc4897c2a 100644
--- a/block/bfq-iosched.h
+++ b/block/bfq-iosched.h
@@ -1009,6 +1009,17 @@ extern struct blkcg_policy blkcg_policy_bfq;
 #define for_each_entity(entity)	\
 	for (; entity ; entity = entity->parent)
 
+static inline void bfqd_enable_active_group_check(struct bfq_data *bfqd)
+{
+	cmpxchg_relaxed(&bfqd->check_active_group, false, true);
+}
+
+static inline bool bfqd_has_active_group(struct bfq_data *bfqd)
+{
+	return bfqd->check_active_group &&
+	       bfqd->num_groups_with_pending_reqs > 0;
+}
+
 /*
  * For each iteration, compute parent in advance, so as to be safe if
  * entity is deallocated during the iteration. Such a deallocation may
@@ -1084,16 +1095,6 @@ static const char *checked_dev_name(const struct device *dev)
 }
 
 #ifdef CONFIG_BFQ_GROUP_IOSCHED
-static inline void bfqd_enable_active_group_check(struct bfq_data *bfqd)
-{
-	cmpxchg_relaxed(&bfqd->check_active_group, false, true);
-}
-
-static inline bool bfqd_has_active_group(struct bfq_data *bfqd)
-{
-	return bfqd->check_active_group &&
-	       bfqd->num_groups_with_pending_reqs > 0;
-}
 
 #define bfq_log_bfqq(bfqd, bfqq, fmt, args...)  do {		\
 	char pid_str[MAX_PID_STR_LENGTH];			\
-- 
2.30.0.335.ge6362826a0

