From 67f9cb9ca027aca60c7b11fac5a68f68e84cf5b0 Mon Sep 17 00:00:00 2001
From: torvic9 <torvic9@mailbox.org>
Date: Fri, 2 Apr 2021 16:00:26 +0200
Subject: [PATCH 3/7] allow setting zstd compression level for modules

Signed-off-by: torvic9 <torvic9@mailbox.org>
---
 Makefile     | 2 +-
 init/Kconfig | 8 ++++++++
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 74c1a4ed6..2a7caf7a1 100644
--- a/Makefile
+++ b/Makefile
@@ -1086,7 +1086,7 @@ ifdef CONFIG_MODULE_COMPRESS
     mod_compress_cmd = $(XZ) --lzma2=dict=2MiB -f
   endif # CONFIG_MODULE_COMPRESS_XZ
   ifdef CONFIG_MODULE_COMPRESS_ZSTD
-    mod_compress_cmd = $(ZSTD) -T0 --rm -f -q
+    mod_compress_cmd = $(ZSTD) -T0 -$(CONFIG_MODULE_COMPRESS_ZSTD_LEVEL) --rm -f -q
   endif # CONFIG_MODULE_COMPRESS_ZSTD
 endif # CONFIG_MODULE_COMPRESS
 export mod_compress_cmd
diff --git a/init/Kconfig b/init/Kconfig
index f9c850072..a16e83ce6 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -2272,6 +2272,14 @@ config MODULE_COMPRESS_ZSTD
 
 endchoice
 
+config MODULE_COMPRESS_ZSTD_LEVEL
+    int "Compression level (1-19)"
+    depends on MODULE_COMPRESS_ZSTD
+    range 1 19
+    default 3
+    help
+        Compression level used by zstd for compressing modules.
+
 config MODULE_ALLOW_MISSING_NAMESPACE_IMPORTS
 	bool "Allow loading of modules with missing namespace imports"
 	help
-- 
2.32.0.93.g670b81a890

