From efc8c717899149b143120030aef352ff802ac88b Mon Sep 17 00:00:00 2001
From: Andy Lavr <andy.lavr@gmail.com>
Date: Mon, 4 Mar 2019 17:46:18 +0200
Subject: [PATCH] block: loop: Small fix

Signed-off-by: Andy Lavr <andy.lavr@gmail.com>
---
 drivers/block/loop.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/block/loop.c b/drivers/block/loop.c
index ee9b11f2c4d4..e51bf39142af 100644
--- a/drivers/block/loop.c
+++ b/drivers/block/loop.c
@@ -756,8 +756,8 @@ static int loop_change_fd(struct loop_device *lo, struct block_device *bdev,
 	mutex_unlock(&loop_ctl_mutex);
 	if (file)
 		fput(file);
-		if (virt_file)
-			fput(virt_file);
+	if (virt_file)
+		fput(virt_file);
 	return error;
 }
 
-- 
2.21.0.4.g36eb1cb9cf

