From e1f02ab2d2578811a43b193c837af3707831781a Mon Sep 17 00:00:00 2001
From: Username404-59 <w.iron.zombie@gmail.com>
Date: Thu, 12 Dec 2024 21:41:01 +0100
Subject: [PATCH 3/4] kbuild: load polly llvm plugin

Signed-off-by: Username404-59 <w.iron.zombie@gmail.com>
---
 Makefile | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index fa433f1d9..e8737f5ea 100644
--- a/Makefile
+++ b/Makefile
@@ -871,7 +871,8 @@ KBUILD_RUSTFLAGS += -Cdebug-assertions=$(if $(CONFIG_RUST_DEBUG_ASSERTIONS),y,n)
 KBUILD_RUSTFLAGS += -Coverflow-checks=$(if $(CONFIG_RUST_OVERFLOW_CHECKS),y,n)
 
 ifdef CONFIG_POLLY_CLANG
-KBUILD_CFLAGS	+= -mllvm -polly \
+KBUILD_CFLAGS	+= -fplugin=LLVMPolly.so \
+		   -mllvm -polly \
 		   -mllvm -polly-ast-use-context \
 		   -mllvm -polly-invariant-load-hoisting \
 		   -mllvm -polly-loopfusion-greedy \
-- 
2.47.1

