From 21070d97a3149717d22eb213df396916665c8eab Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 7 Apr 2025 09:19:08 +0200
Subject: [PATCH 2/3] net/ovpn-dco: adopt 69c7be1b903fca

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/net/ovpn-dco/main.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ovpn-dco/main.c b/drivers/net/ovpn-dco/main.c
index 87a7aa455..c5757e3ed 100644
--- a/drivers/net/ovpn-dco/main.c
+++ b/drivers/net/ovpn-dco/main.c
@@ -163,10 +163,12 @@ static const struct nla_policy ovpn_policy[IFLA_OVPN_MAX + 1] = {
 					    __OVPN_MODE_AFTER_LAST - 1),
 };
 
-static int ovpn_newlink(struct net *src_net, struct net_device *dev, struct nlattr *tb[],
-			struct nlattr *data[], struct netlink_ext_ack *extack)
+static int ovpn_newlink(struct net_device *dev,
+			struct rtnl_newlink_params *params,
+			struct netlink_ext_ack *extack)
 {
 	struct ovpn_struct *ovpn = netdev_priv(dev);
+	struct nlattr **data = params->data;
 	int ret;
 
 	ret = security_tun_dev_create();
-- 
2.49.0

