From 52055dc75aaf0e90e79bd8aabd2455cd7d57e694 Mon Sep 17 00:00:00 2001
From: Oleksandr Natalenko <oleksandr@natalenko.name>
Date: Mon, 7 Apr 2025 09:22:32 +0200
Subject: [PATCH 3/3] net/ovpn-dco: adopt 8fa7292fee5c52

Signed-off-by: Oleksandr Natalenko <oleksandr@natalenko.name>
---
 drivers/net/ovpn-dco/peer.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ovpn-dco/peer.c b/drivers/net/ovpn-dco/peer.c
index 849d6f558..35c930596 100644
--- a/drivers/net/ovpn-dco/peer.c
+++ b/drivers/net/ovpn-dco/peer.c
@@ -191,8 +191,8 @@ void ovpn_peer_float(struct ovpn_peer *peer, struct sk_buff *skb)
 
 static void ovpn_peer_timer_delete_all(struct ovpn_peer *peer)
 {
-	del_timer_sync(&peer->keepalive_xmit);
-	del_timer_sync(&peer->keepalive_recv);
+	timer_delete_sync(&peer->keepalive_xmit);
+	timer_delete_sync(&peer->keepalive_recv);
 }
 
 static void ovpn_peer_free(struct ovpn_peer *peer)
@@ -303,7 +303,7 @@ void ovpn_peer_keepalive_set(struct ovpn_peer *peer, u32 interval, u32 timeout)
 		delta = msecs_to_jiffies(interval * MSEC_PER_SEC);
 		mod_timer(&peer->keepalive_xmit, jiffies + delta);
 	} else {
-		del_timer(&peer->keepalive_xmit);
+		timer_delete(&peer->keepalive_xmit);
 	}
 
 	peer->keepalive_timeout = timeout;
@@ -311,7 +311,7 @@ void ovpn_peer_keepalive_set(struct ovpn_peer *peer, u32 interval, u32 timeout)
 		delta = msecs_to_jiffies(timeout * MSEC_PER_SEC);
 		mod_timer(&peer->keepalive_recv, jiffies + delta);
 	} else {
-		del_timer(&peer->keepalive_recv);
+		timer_delete(&peer->keepalive_recv);
 	}
 }
 
-- 
2.49.0

