From 61714cacb29e8a78320b3b53ec4e6b4ca673f01b Mon Sep 17 00:00:00 2001
From: Andy Lavr <andy.lavr@gmail.com>
Date: Mon, 24 Dec 2018 13:18:45 +0200
Subject: [PATCH] block: Fix enables BFQ-MQ multi-queue IO scheduling by
 default

Signed-off-by: Andy Lavr <andy.lavr@gmail.com>
---
 block/Kconfig.iosched | 18 ++++++++++++++++--
 1 file changed, 16 insertions(+), 2 deletions(-)

diff --git a/block/Kconfig.iosched b/block/Kconfig.iosched
index 030e6fb84d5d..cb8080fc56ff 100644
--- a/block/Kconfig.iosched
+++ b/block/Kconfig.iosched
@@ -41,7 +41,7 @@ config CFQ_GROUP_IOSCHED
 
 config IOSCHED_BFQ_SQ
 	tristate "BFQ-SQ I/O scheduler"
-	default n
+	default y
 	---help---
 	The BFQ-SQ I/O scheduler (for legacy blk: SQ stands for
 	SingleQueue) distributes bandwidth among all processes
@@ -62,7 +62,7 @@ config BFQ_SQ_GROUP_IOSCHED
 choice
 
 	prompt "Default I/O scheduler"
-	default DEFAULT_CFQ
+	default DEFAULT_BFQ_MQ
 	help
 	  Select the I/O scheduler which will be used by default for all
 	  block devices.
@@ -83,6 +83,16 @@ choice
 		  any workload. It also tries to guarantee low latency to
 		  interactive and soft real-time applications.
 
+	config DEFAULT_BFQ_MQ
+		bool "BFQ-MQ" if MQ_IOSCHED_BFQ=y
+		help
+		  Selects BFQ-MQ as the default I/O scheduler which will be
+		  used by default for all block devices.
+		  The BFQ-MQ I/O scheduler aims at distributing the bandwidth
+		  as desired, independently of the disk parameters and with
+		  any workload. It also tries to guarantee low latency to
+		  interactive and soft real-time applications.
+
 	config DEFAULT_NOOP
 		bool "No-op"
 
@@ -93,10 +103,14 @@ config DEFAULT_IOSCHED
 	default "deadline" if DEFAULT_DEADLINE
 	default "cfq" if DEFAULT_CFQ
 	default "bfq-sq" if DEFAULT_BFQ_SQ
+	default "bfq-mq" if DEFAULT_BFQ_MQ
 	default "noop" if DEFAULT_NOOP
 
 config MQ_IOSCHED_BFQ
 	tristate "BFQ-MQ I/O Scheduler"
+	select DM_MQ_DEFAULT
+	select SCSI_MQ_DEFAULT
+	select BLK_CGROUP_IOLATENCY
 	default y
 	---help---
 	BFQ I/O scheduler for BLK-MQ. BFQ-MQ distributes bandwidth
-- 
2.20.1.2.gb21ebb671b

