From 89ba68184be4aee167ebc842678bd12d5be696aa Mon Sep 17 00:00:00 2001
From: Andy Lavr <andy.lavr@gmail.com>
Date: Mon, 24 Dec 2018 13:32:07 +0200
Subject: [PATCH] block/Kconfig.iosched: Fix depends for BLK_CGROUP_IOLATENCY

Signed-off-by: Andy Lavr <andy.lavr@gmail.com>
---
 block/Kconfig.iosched | 1 +
 1 file changed, 1 insertion(+)

diff --git a/block/Kconfig.iosched b/block/Kconfig.iosched
index cb8080fc56ff..378dfb268b7f 100644
--- a/block/Kconfig.iosched
+++ b/block/Kconfig.iosched
@@ -110,6 +110,7 @@ config MQ_IOSCHED_BFQ
 	tristate "BFQ-MQ I/O Scheduler"
 	select DM_MQ_DEFAULT
 	select SCSI_MQ_DEFAULT
+	depends on BLOCK
 	select BLK_CGROUP_IOLATENCY
 	default y
 	---help---
-- 
2.20.1.2.gb21ebb671b

