From eb84d96127d5a49d83712297ec3672ea9369587c Mon Sep 17 00:00:00 2001
From: Tor Vic <torvic9@mailbox.org>
Date: Mon, 20 Feb 2023 11:31:40 +0100
Subject: [PATCH 6/7] Kconfig, zstd: Rename COMP_VAL to COMPRESSION_LVL

---
 init/Kconfig         | 2 +-
 scripts/Makefile.lib | 4 ++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/init/Kconfig b/init/Kconfig
index efdf85f03..a8eb642df 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -351,7 +351,7 @@ endchoice
 menu "ZSTD compression options"
 	depends on KERNEL_ZSTD
 
-config ZSTD_COMP_VAL
+config ZSTD_COMPRESSION_LVL
 	int "Compression level (1-22)"
 	range 1 22
 	default "22"
diff --git a/scripts/Makefile.lib b/scripts/Makefile.lib
index f34cb523b..c13a218c5 100644
--- a/scripts/Makefile.lib
+++ b/scripts/Makefile.lib
@@ -542,8 +542,8 @@ quiet_cmd_xzmisc = XZMISC  $@
 # decompression is used, like initramfs decompression, zstd22 should likely not
 # be used because it would require zstd to allocate a 128 MB buffer.
 
-ifdef CONFIG_ZSTD_COMP_VAL
-zstd_comp_val	:= $(CONFIG_ZSTD_COMP_VAL)
+ifdef CONFIG_ZSTD_COMPRESSION_LVL
+zstd_comp_val	:= $(CONFIG_ZSTD_COMPRESSION_LVL)
 ifeq ($(shell test $(zstd_comp_val) -gt 19; echo $$?),0)
 zstd_comp_val	+= --ultra
 endif
-- 
2.40.0.71.g950264636c

