From f2dadf7f207b7eec73b268d36c6dbb2ecffce27e Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Tue, 13 Jun 2023 02:26:21 +0000
Subject: [PATCH 48/50] fixup! CONFIG_MEM_ALLOC_PROFILING_ENABLED_BY_DEFAULT
 description

Signed-off-by: Suren Baghdasaryan <surenb@google.com>
---
 lib/Kconfig.debug | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/Kconfig.debug b/lib/Kconfig.debug
index 60496018f..0bed33026 100644
--- a/lib/Kconfig.debug
+++ b/lib/Kconfig.debug
@@ -975,7 +975,7 @@ config MEM_ALLOC_PROFILING
 	  memory leaks with a low performance impact.
 
 config MEM_ALLOC_PROFILING_ENABLED_BY_DEFAULT
-	bool "Disable memory allocation profiling by default"
+	bool "Enable memory allocation profiling by default"
 	default y
 	depends on MEM_ALLOC_PROFILING
 
-- 
2.41.0

